<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>GeoMXAnalysisWorkflow â€¢ GeoMXAnalysisWorkflow</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="GeoMXAnalysisWorkflow">
<meta property="og:description" content="A workshop demostrating the analysis workflow of GeoMX transcriptmoics data.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">GeoMXAnalysisWorkflow</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="articles/GeoMXAnalysisWorkflow.html">Get started</a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/DavisLaboratory/GeoMXAnalysisWorkflow" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div class="section level1">
<div class="page-header"><h1 id="analysing-geomx-dsp-dataset-with-standr">Analysing GeoMX DSP dataset with standR<a class="anchor" aria-label="anchor" href="#analysing-geomx-dsp-dataset-with-standr"></a>
</h1></div>
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>This workshop will focus on performing analysis of spatial transcriptomics data from the Nanostring GeoMX DSP platform. We will perform data QC, normalisation and batch correction using methods in the <code>standR</code> package to analyse a publicly available GeoMX WTA dataset. Following this, we will perform differential expression analysis using tools from the <code>limma</code> and <code>edgeR</code> packages. Finally, we will carry out a gene set enrichment analysis (GSEA) using <code>fry</code> from the <code>limma</code> package and visualise and interpret the GSEA results using the <code>vissE</code> package.</p>
<p>The workshop will be organised into two broad sections: * Processing and analysing the data using standR * Performing DE analysis and GSEA</p>
<p>Detailed material can be found <a href="https://davislaboratory.github.io/GeoMXAnalysisWorkflow/" class="external-link">here</a>.</p>
</div>
<div class="section level2">
<h2 id="pre-requisites">Pre-requisites<a class="anchor" aria-label="anchor" href="#pre-requisites"></a>
</h2>
<p>The workshop is aimed at researchers who are interested in analysing GeoMX transcriptomics data. Some basic R knowledge is assumed - this is not an introduction to R course. If you are not familiar with the R statistical programming language it is strongly suggested that you work through an introductory R course before you attend this workshop.</p>
</div>
<div class="section level2">
<h2 id="r-packages-used">
<em>R</em> packages used<a class="anchor" aria-label="anchor" href="#r-packages-used"></a>
</h2>
<p>The following key R packages will be used:</p>
<ul>
<li><code>standR</code></li>
<li><code>edgeR</code></li>
<li><code>limma</code></li>
<li><code>msigdb</code></li>
<li><code>GSEABase</code></li>
<li><code>igraph</code></li>
<li><code>vissE</code></li>
<li><code>SpatialExperiment</code></li>
<li><code>Scater</code></li>
</ul>
</div>
<div class="section level2">
<h2 id="time-outline">Time outline<a class="anchor" aria-label="anchor" href="#time-outline"></a>
</h2>
<table class="table">
<colgroup>
<col width="91%">
<col width="8%">
</colgroup>
<thead><tr class="header">
<th>Activity</th>
<th>Time</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Introduction &amp; setup</td>
<td>10m</td>
</tr>
<tr class="even">
<td>Part 1. Processing and analysing the data using standR</td>
<td>45m</td>
</tr>
<tr class="odd">
<td>Part 2. Performing DE analysis and GSEA</td>
<td>45m</td>
</tr>
<tr class="even">
<td>Q &amp; A</td>
<td>10m</td>
</tr>
</tbody>
</table>
</div>
<div class="section level2">
<h2 id="workshop-goals-and-objectives">Workshop goals and objectives<a class="anchor" aria-label="anchor" href="#workshop-goals-and-objectives"></a>
</h2>
<div class="section level3">
<h3 id="learning-goals">Learning goals<a class="anchor" aria-label="anchor" href="#learning-goals"></a>
</h3>
<ul>
<li>Learn how to analyse GeoMX transcriptomics data using standR.</li>
<li>Learn how to perform DE analysis with the GeoMX transcriptomics data using limma.</li>
<li>Learn how to visualize the GSEA results using vissE.</li>
</ul>
</div>
<div class="section level3">
<h3 id="learning-objectives">Learning objectives<a class="anchor" aria-label="anchor" href="#learning-objectives"></a>
</h3>
<ul>
<li>Generate QC plots for the GeoMX transcriptomics data.</li>
<li>Perform normalisation/batch correction on the GeoMX transcriptomics data.</li>
<li>Assessing the normalisation/batch correction results.</li>
<li>Perform differential expression ananlysis on the GeoMX transcriptomics data.</li>
<li>Perform gene set enrichment analysis on the DE results.</li>
<li>Apply vissE to identify higher-order phenotypes and to visualise the GSEA results.</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="workshop-package-installation">Workshop package installation<a class="anchor" aria-label="anchor" href="#workshop-package-installation"></a>
</h2>
<div class="section level3">
<h3 id="guide">Guide<a class="anchor" aria-label="anchor" href="#guide"></a>
</h3>
<p>This is necessary in order to reproduce the code shown in the workshop. The workshop is designed for R <code>4.2</code> and Bioconductor <code>3.15</code> and can be installed via GitHub with the following command.</p>
<pre><code><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">'remotes'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Install workshop package</span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"DavisLaboratory/GeoMXAnalysisWorkflow"</span>, build_vignettes <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># To view vignettes</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://davislaboratory.github.io/GeoMXAnalysisWorkflow/" class="external-link">GeoMXAnalysisWorkflow</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/browseVignettes.html" class="external-link">browseVignettes</a></span><span class="op">(</span><span class="st">"GeoMXAnalysisWorkflow"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Or go to [here](https://davislaboratory.github.io/GeoMXAnalysisWorkflow/articles/GeoMXAnalysisWorkflow.html)</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="to-use-the-resulting-image">To use the resulting image:<a class="anchor" aria-label="anchor" href="#to-use-the-resulting-image"></a>
</h2>
<div class="sourceCode" id="cb2"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="ex">docker</span> run <span class="at">-e</span> PASSWORD=<span class="op">&lt;</span>choose_a_password_for_rstudio<span class="op">&gt;</span> -p 8787:8787 ghcr.io/ningbioinfo/geomxanalysisworkflow:latest</span></code></pre></div>
<p>Once running, navigate to <a href="http://localhost:8787/" class="external-link uri">http://localhost:8787/</a> and then login with Username:rstudio and Password:password.</p>
<p>You should see the Rmarkdown file with all the workshop code which you can run.</p>
<p><em>NOTE</em>: Running docker that uses the password in plain text like above exposes the password to others in a multi-user system (like a shared workstation or compute node). In practice, consider using an environment variable instead of plain text to pass along passwords and other secrets in docker command lines.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/DavisLaboratory/GeoMXAnalysisWorkflow/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/DavisLaboratory/GeoMXAnalysisWorkflow/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small>GPL (&gt;= 3) + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing GeoMXAnalysisWorkflow</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Ning Liu <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0002-9487-9305" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Chin Wee Tan <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0001-9695-7218" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Melissa J. Davis <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0003-4864-7033" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>



  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Ning Liu, Chin Wee Tan, Melissa J. Davis.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
