<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Analysing Nanostring's GeoMX transcriptomics data using standR, limma and vissE • GeoMXAnalysisWorkflow</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Analysing Nanostring's GeoMX transcriptomics data using standR, limma and vissE">
<meta property="og:description" content="GeoMXAnalysisWorkflow">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">GeoMXAnalysisWorkflow</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/GeoMXAnalysisWorkflow.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/DavisLaboratory/GeoMXAnalysisWorkflow" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="GeoMXAnalysisWorkflow_files/htmlwidgets-1.5.4/htmlwidgets.js"></script><link href="GeoMXAnalysisWorkflow_files/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="GeoMXAnalysisWorkflow_files/datatables-binding-0.23/datatables.js"></script><link href="GeoMXAnalysisWorkflow_files/dt-core-1.11.3/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="GeoMXAnalysisWorkflow_files/dt-core-1.11.3/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="GeoMXAnalysisWorkflow_files/dt-core-1.11.3/js/jquery.dataTables.min.js"></script><link href="GeoMXAnalysisWorkflow_files/crosstalk-1.2.0/css/crosstalk.min.css" rel="stylesheet">
<script src="GeoMXAnalysisWorkflow_files/crosstalk-1.2.0/js/crosstalk.min.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Analysing Nanostring’s GeoMX transcriptomics
data using standR, limma and vissE</h1>
                        <h4 data-toc-skip class="author">Ning Liu</h4>
            <address class="author_afil">
      Bioinformatics Division, Walter and Eliza Hall Institute of
Medical Research, Parkville, VIC 3052, AustraliaDepartment of Medical
Biology, University of Melbourne, Parkville, VIC 3010,
Australia<br><a class="author_email" href="mailto:#"></a><a href="mailto:liu.n@wehi.edu.au" class="email">liu.n@wehi.edu.au</a>
      </address>
                              <h4 data-toc-skip class="author">Chin Wee
Tan</h4>
            <address class="author_afil">
      Bioinformatics Division, Walter and Eliza Hall Institute of
Medical Research, Parkville, VIC 3052, AustraliaDepartment of Medical
Biology, University of Melbourne, Parkville, VIC 3010,
Australia<br><a class="author_email" href="mailto:#"></a><a href="mailto:cwtan@wehi.edu.au" class="email">cwtan@wehi.edu.au</a>
      </address>
                              <h4 data-toc-skip class="author">Melissa J
Davis</h4>
            <address class="author_afil">
      Bioinformatics Division, Walter and Eliza Hall Institute of
Medical Research, Parkville, VIC 3052, AustraliaDepartment of Medical
Biology, University of Melbourne, Parkville, VIC 3010,
AustraliaDepartment of Biochemistry and Molecular Biology, Faculty of
Medicine, Dentistry and Health Sciences, University of Melbourne,
Parkville, VIC, 3010,
Australia<br><a class="author_email" href="mailto:#"></a><a href="mailto:davis.m@wehi.edu.au" class="email">davis.m@wehi.edu.au</a>
      </address>
                  
            <h4 data-toc-skip class="date">Jul 2022</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/DavisLaboratory/GeoMXAnalysisWorkflow/blob/HEAD/vignettes/GeoMXAnalysisWorkflow.Rmd" class="external-link"><code>vignettes/GeoMXAnalysisWorkflow.Rmd</code></a></small>
      <div class="hidden name"><code>GeoMXAnalysisWorkflow.Rmd</code></div>

    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>Abstract</p>
    </div>
    
<p>
</p>
<p><strong>R version</strong>: R version 4.2.0 (2022-04-22) <br><strong>Bioconductor version</strong>: 3.16</p>

<div class="section level2">
<h2 id="background-and-introduction">Background and introduction<a class="anchor" aria-label="anchor" href="#background-and-introduction"></a>
</h2>
<div class="section level3">
<h3 id="nanostring-geomx-data">Nanostring GeoMX data<a class="anchor" aria-label="anchor" href="#nanostring-geomx-data"></a>
</h3>
<p>Nanostring’s GeoMx DSP data comes from the GeoMx DSP workflow which
integrates standard pathology and molecular profiling to obtain robust
and reproducible spatial multiomics data. DSP data typically comes from
whole tissue sections, FFPE or fresh frozen samples. These can be imaged
and stained for RNA or protein using which specific tissue compartments
or cell types to be profiled based on the biology can be selected. These
selected <code>regions</code> or <code>areas</code> of interests will
have their count expression levels quantified using either the nCounter
Analysis System or an Illumina Sequencer.</p>
<p>GeoMx RNA assays allows quantitative and spatial measurements of
transcripts (up to the whole transcriptome) from single sections of FFPE
or fixed fresh frozen tissues. Typical gene panels utilised include the
Cancer transcriptome Atlas (CTA, ~1800 genes) and Whole Transcriptome
Atlas (WTA, ~ 18000 genes).</p>
<p>Count data are generated from the DSP pipeline and technical
pre-processed using the GeoMx Data Analysis Suite (DSPDA). This includes
the removal of low performing probes and calculation of a typical QC
metrics called LOQ using the negative probes. Data normalization are
typically conducted based on recommendations from Nanostring by the
technologist/technicians with a suggested Q3 normalized data output.</p>
<p><strong>Note: Q3-normalised data is not recommended for any
bioinformatics workflow or pipeline. Rather, the technical probe
corrected counts (probeQC, which accounts for any technical machine
systematic errors during the DSP run) is recommended.</strong></p>
<div class="figure">
<img src="../inst/images/workflow.jpg" title="GeoMX data analysis" alt=""><p class="caption">flowchart</p>
</div>
</div>
<div class="section level3">
<h3 id="analysis-of-spatial-geomx-datasets">Analysis of spatial GeoMX datasets<a class="anchor" aria-label="anchor" href="#analysis-of-spatial-geomx-datasets"></a>
</h3>
<p>A typical bioinformatics analysis of a GeoMX DSP dataset often starts
with a count table (from sequencing reads of genes for each region of
interest (ROI)), and ends with either identifying differential expressed
genes or performing gene signature/gene-set scoring pathway/enrichment
analyses in various conditions or experimental designs. Nevertheless,
before performing any differential expression (DE) analysis or other
downstream analyses that are based on the gene counts, proper quality
control(QC) and normalisation of the data is essential and if not done
properly, will greatly impact the correctness and validity of the DE and
corresponding downstream analysis’ results. We therefore developed a
bioconductor package called <code>standR</code>
(<strong>S</strong>patial <strong>t</strong>ranscriptomics
<strong>a</strong>nalyzes a<strong>n</strong>d <strong>d</strong>ecoding
in <strong>R</strong>) to assist the QC, normalisation and batch
correction of the GeoMX transcriptomics data.</p>
<p>There are three major advantages of using the <code>standR</code>
package to analyse the GeoMX DSP datasets:</p>
<ol style="list-style-type: decimal">
<li><p>The package uses the <code>SpatialExperiment</code>
infrastructure to analyse the data, this infrastructure is a lineage of
the <code>SummarisedExperiment</code> family, which is highly
recommended in the bioconductor community. It is compatible and
transferable with many other well developed packages in the RNA-seq
analysis world, such as <code>scater</code>, <code>scran</code>,
<code>edgeR</code> and <code>limma</code>.</p></li>
<li><p>The package features a comprehensive route for quality control,
and provides various visualisation functions to help in the assessment
of the various quality control metrics.</p></li>
<li><p>Batch effect is a common feature in transcriptomic dataset,
especially in GeoMX DSP data due to the way the slides are typically
utilised due to experimental constrains/designs. The package currently
provides three batch correction methods that will remove the unwanted
batch effect and provides statistics for assessing the correction
process and outcomes.</p></li>
</ol>
<p>In this workshop, we will firstly use <code>standR</code> to process
and analyse a published GeoMX WTA dataset using the recommended
workflow. This will demonstrate our recommended workflow for processing
and analysing GeoMX transcriptomics datasets. Secondly, we will perform
DE analysis of the processed data using the <code>limma-voom</code>
pipeline, followed by a gene-set enrichment analysis using
<code>fry</code> and subsequent visualisation of the higher order
results using the R package <code>vissE</code>.</p>
</div>
</div>
<div class="section level2">
<h2 id="using-standr-for-process-and-analyseg-geomx-transcriptomics-data">Using standR for process and analyseg GeoMX transcriptomics
data<a class="anchor" aria-label="anchor" href="#using-standr-for-process-and-analyseg-geomx-transcriptomics-data"></a>
</h2>
<div class="section level3">
<h3 id="load-data">Load data<a class="anchor" aria-label="anchor" href="#load-data"></a>
</h3>
<p>The data we are using in this workshop is a published GeoMX whole
transcriptome atlas (WTA) dataset of diabetic kidney disease (DKD) that
was made available by Nanostring (url XXX, see if we can include a image
of the typical dataset).</p>
<p>This dataset includes data on <strong>seven slides</strong>, which
includes <strong>4 DKD and 3 healthy</strong> slides. The ROI selection
strategy employed is a regional based approach focused primarily on two
distinct structures found in the kidney, namely: <strong>tubules or
glomeruli</strong>.</p>
<p><em>Glomeruli</em> ROIs were identified and curated by a pathologist
as either <strong>relatively healthy or diseased</strong> regardless of
the tissue type (DKD or healthy).<br><em>Tubule</em> ROIs were segmented into <strong>distal (PanCK positive)
and proximal (PanCK negative)</strong> tubules. Distal and proximal
tubules are distinct in terms of their differing functions in the
kidney.</p>
<p>This GeoMX DSP data can be retrieved by querying the
<code>ExperimentHub</code> repository. The query below searches for all
objects in the hub associated with the search term “standR”.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">ExperimentHub</span><span class="op">)</span></span>
<span></span>
<span><span class="va">eh</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ExperimentHub/man/ExperimentHub-class.html" class="external-link">ExperimentHub</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/AnnotationHub/man/AnnotationHub-class.html" class="external-link">query</a></span><span class="op">(</span><span class="va">eh</span>, <span class="st">"standR"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## ExperimentHub with 3 records</span></span>
<span><span class="co">## # snapshotDate(): 2022-07-15</span></span>
<span><span class="co">## # $dataprovider: Nanostring</span></span>
<span><span class="co">## # $species: NA</span></span>
<span><span class="co">## # $rdataclass: data.frame</span></span>
<span><span class="co">## # additional mcols(): taxonomyid, genome, description,</span></span>
<span><span class="co">## #   coordinate_1_based, maintainer, rdatadateadded, preparerclass, tags,</span></span>
<span><span class="co">## #   rdatapath, sourceurl, sourcetype </span></span>
<span><span class="co">## # retrieve records with, e.g., 'object[["EH7364"]]' </span></span>
<span><span class="co">## </span></span>
<span><span class="co">##            title                   </span></span>
<span><span class="co">##   EH7364 | GeomxDKDdata_count      </span></span>
<span><span class="co">##   EH7365 | GeomxDKDdata_sampleAnno </span></span>
<span><span class="co">##   EH7366 | GeomxDKDdata_featureAnno</span></span></code></pre>
<p>The three files that are associated with the DKD GeoMX data in the
<code>standR</code> package can then be retrieved by specifying their
IDs. The respective objects can then be retrieved using the accession
IDs (“EH7364” for counts,“EH7365” for sample annotation,“EH7366” for
gene annotations)</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">countFile</span> <span class="op">&lt;-</span> <span class="va">eh</span><span class="op">[[</span><span class="st">"EH7364"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">sampleAnnoFile</span> <span class="op">&lt;-</span> <span class="va">eh</span><span class="op">[[</span><span class="st">"EH7365"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">featureAnnoFile</span> <span class="op">&lt;-</span> <span class="va">eh</span><span class="op">[[</span><span class="st">"EH7366"</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<p>We can have a first look at the format of the three files, which are
the typical files made available by NanoString.</p>
<p>The <code>countFile</code> is a tab-delimited file, it contains the
count table (features by samples) we generally see in transcriptomics
analysis. By default as provided by the Nanostring, it is required to
have the gene name column with the column name of “TargetName”.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.delim</a></span><span class="op">(</span><span class="va">countFile</span>, nrows <span class="op">=</span> <span class="fl">5</span><span class="op">)</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##   TargetName disease3_scan...001...PanCK disease3_scan...001...neg</span></span>
<span><span class="co">## 1      PADI2                          15                        35</span></span>
<span><span class="co">## 2    CYP24A1                          23                        32</span></span>
<span><span class="co">## 3    SUPT16H                          35                        66</span></span>
<span><span class="co">## 4      ZMIZ2                          51                        81</span></span>
<span><span class="co">## 5      SPAG9                          48                        65</span></span></code></pre>
<p>The <code>sampleAnnoFile</code> is a tab-delimited file, containing
all the annotation (metadata) for the samples. By default as provided by
the Nanostring, it is required to include the sample name column with
the column name of “SegmentDisplayName”.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.delim</a></span><span class="op">(</span><span class="va">sampleAnnoFile</span>, nrows <span class="op">=</span> <span class="fl">5</span><span class="op">)</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##   SlideName      ScanName ROILabel      SegmentLabel</span></span>
<span><span class="co">## 1  disease3 disease3_scan        7 Geometric Segment</span></span>
<span><span class="co">## 2  disease3 disease3_scan        8 Geometric Segment</span></span>
<span><span class="co">## 3  disease3 disease3_scan        9 Geometric Segment</span></span>
<span><span class="co">## 4  disease3 disease3_scan       10 Geometric Segment</span></span>
<span><span class="co">## 5  disease3 disease3_scan       11 Geometric Segment</span></span>
<span><span class="co">##                        SegmentDisplayName</span></span>
<span><span class="co">## 1 disease3_scan | 007 | Geometric Segment</span></span>
<span><span class="co">## 2 disease3_scan | 008 | Geometric Segment</span></span>
<span><span class="co">## 3 disease3_scan | 009 | Geometric Segment</span></span>
<span><span class="co">## 4 disease3_scan | 010 | Geometric Segment</span></span>
<span><span class="co">## 5 disease3_scan | 011 | Geometric Segment</span></span></code></pre>
<p>The <code>featureAnnoFile</code> is a tab-delimited file, containing
all the annotation (metadata) of the genes in the dataset. By default as
provided by the Nanostring, it is required to include the gene name
column with the column name of “TargetName”.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.delim</a></span><span class="op">(</span><span class="va">featureAnnoFile</span>, nrows <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##       RTS_ID TargetName ProbeID Negative</span></span>
<span><span class="co">## 1 RTS0020877        A2M   35614    FALSE</span></span>
<span><span class="co">## 2 RTS0020878       NAT2   35615    FALSE</span></span>
<span><span class="co">## 3 RTS0020879      ACADM   35616    FALSE</span></span>
<span><span class="co">## 4 RTS0020880      ACADS   35617    FALSE</span></span>
<span><span class="co">## 5 RTS0020881      ACAT1   35618    FALSE</span></span></code></pre>
<p>As described in the introduction, there are many advantages to use a
mature infrastructure throughout the analysis, such as compatibility
with other tools.</p>
<p>Therefore, the first step in the <code>standR</code> package workflow
is to construct a <code>SpatialExperiment</code> object that includes
all the information available in the data. Here we can use the function
<code>readGeoMX</code> to do so. For more information about the
<code>SpatialExperiment</code> infrastructure, see <a href="http://www.bioconductor.org/packages/release/bioc/html/SpatialExperiment.html" class="external-link">here</a>.</p>
<p><strong>Note 1: by default, the <code>readGeoMx</code> function will
look for the gene name column in both the <code>countFile</code> and
<code>featureAnnoFile</code> with the column name of “TargetName”, and
the sample name column in the <code>sampleAnnoFile</code> with the
column name of “SegmentDisplayName”, these column names are given by the
Nanostring in the default settings, if your data have been modified, you
can indicate the corresponding column names by specifying the parameter
“colnames.as.rownames” in the <code>readGeoMx</code> function when
loading the data.</strong></p>
<p><strong>Note 2: if you plan to use <code>readGeoMx</code> to
construct the <code>SpatialExperiment</code> object with your own data,
make sure that the files you use as inputs are tab-delimited
files.</strong></p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/DavisLaboratory/standR" class="external-link">standR</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">spe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/standR/man/readGeoMx.html" class="external-link">readGeoMx</a></span><span class="op">(</span><span class="va">countFile</span>, <span class="va">sampleAnnoFile</span>, featureAnnoFile <span class="op">=</span> <span class="va">featureAnnoFile</span><span class="op">)</span></span></code></pre></div>
<p>Check the basic information about the dataset by entering the object
name directly. We see that the data has measurements for approximately
18503 genes and 231 ROIs.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe</span></span></code></pre></div>
<pre><code><span><span class="co">## class: SpatialExperiment </span></span>
<span><span class="co">## dim: 18503 231 </span></span>
<span><span class="co">## metadata(1): NegProbes</span></span>
<span><span class="co">## assays(2): counts logcounts</span></span>
<span><span class="co">## rownames(18503): PADI2 CYP24A1 ... FAM166A AGTPBP1</span></span>
<span><span class="co">## rowData names(3): RTS_ID ProbeID Negative</span></span>
<span><span class="co">## colnames(231): disease3_scan | 001 | PanCK disease3_scan | 001 | neg</span></span>
<span><span class="co">##   ... disease1B_scan | 023 | Geometric Segment disease1B_scan | 024 |</span></span>
<span><span class="co">##   Geometric Segment</span></span>
<span><span class="co">## colData names(25): SlideName ScanName ... RoiReportY sample_id</span></span>
<span><span class="co">## reducedDimNames(0):</span></span>
<span><span class="co">## mainExpName: NULL</span></span>
<span><span class="co">## altExpNames(0):</span></span>
<span><span class="co">## spatialCoords names(2) : ROICoordinateX ROICoordinateY</span></span>
<span><span class="co">## imgData names(0):</span></span></code></pre>
<p>Both count-level data and logCPM measurements are stored in the
<code>spatialExperiment</code> object. Specifically, the raw count data
is stored in the <code>counts</code> assay slot, while the log-CPM
(count per million) of the data is calculated by default with the
<code>readGeoMX</code> function and stored in the <code>logcounts</code>
assay of the object.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">assayNames</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "counts"    "logcounts"</span></span></code></pre>
<p>We can have a look at the count table by using the <code>assay</code>
function and specify the table name.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">assay</span><span class="op">(</span><span class="va">spe</span>, <span class="st">"counts"</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##         disease3_scan | 001 | PanCK disease3_scan | 001 | neg</span></span>
<span><span class="co">## PADI2                            15                        35</span></span>
<span><span class="co">## CYP24A1                          23                        32</span></span>
<span><span class="co">## SUPT16H                          35                        66</span></span>
<span><span class="co">## ZMIZ2                            51                        81</span></span>
<span><span class="co">## SPAG9                            48                        65</span></span>
<span><span class="co">##         disease3_scan | 002 | PanCK disease3_scan | 002 | neg</span></span>
<span><span class="co">## PADI2                            36                        41</span></span>
<span><span class="co">## CYP24A1                          23                        41</span></span>
<span><span class="co">## SUPT16H                          46                        61</span></span>
<span><span class="co">## ZMIZ2                            87                       113</span></span>
<span><span class="co">## SPAG9                            46                        91</span></span>
<span><span class="co">##         disease3_scan | 003 | PanCK</span></span>
<span><span class="co">## PADI2                             9</span></span>
<span><span class="co">## CYP24A1                           3</span></span>
<span><span class="co">## SUPT16H                          20</span></span>
<span><span class="co">## ZMIZ2                            68</span></span>
<span><span class="co">## SPAG9                            25</span></span></code></pre>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">assay</span><span class="op">(</span><span class="va">spe</span>, <span class="st">"logcounts"</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##         disease3_scan | 001 | PanCK disease3_scan | 001 | neg</span></span>
<span><span class="co">## PADI2                      4.694722                  5.097130</span></span>
<span><span class="co">## CYP24A1                    5.205348                  4.988548</span></span>
<span><span class="co">## SUPT16H                    5.738537                  5.903808</span></span>
<span><span class="co">## ZMIZ2                      6.236242                  6.175525</span></span>
<span><span class="co">## SPAG9                      6.155076                  5.883737</span></span>
<span><span class="co">##         disease3_scan | 002 | PanCK disease3_scan | 002 | neg</span></span>
<span><span class="co">## PADI2                      5.507768                  5.054985</span></span>
<span><span class="co">## CYP24A1                    4.952999                  5.054985</span></span>
<span><span class="co">## SUPT16H                    5.824568                  5.551068</span></span>
<span><span class="co">## ZMIZ2                      6.679218                  6.363792</span></span>
<span><span class="co">## SPAG9                      5.824568                  6.073582</span></span>
<span><span class="co">##         disease3_scan | 003 | PanCK</span></span>
<span><span class="co">## PADI2                      4.573874</span></span>
<span><span class="co">## CYP24A1                    3.518078</span></span>
<span><span class="co">## SUPT16H                    5.538414</span></span>
<span><span class="co">## ZMIZ2                      7.183436</span></span>
<span><span class="co">## SPAG9                      5.827206</span></span></code></pre>
<p>Sample metadata is stored in the <code>colData</code> of the
object.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">colData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 5 rows and 5 columns</span></span>
<span><span class="co">##                               SlideName      ScanName  ROILabel SegmentLabel</span></span>
<span><span class="co">##                             &lt;character&gt;   &lt;character&gt; &lt;numeric&gt;  &lt;character&gt;</span></span>
<span><span class="co">## disease3_scan | 001 | PanCK    disease3 disease3_scan         1        PanCK</span></span>
<span><span class="co">## disease3_scan | 001 | neg      disease3 disease3_scan         1          neg</span></span>
<span><span class="co">## disease3_scan | 002 | PanCK    disease3 disease3_scan         2        PanCK</span></span>
<span><span class="co">## disease3_scan | 002 | neg      disease3 disease3_scan         2          neg</span></span>
<span><span class="co">## disease3_scan | 003 | PanCK    disease3 disease3_scan         3        PanCK</span></span>
<span><span class="co">##                                          Sample_ID</span></span>
<span><span class="co">##                                        &lt;character&gt;</span></span>
<span><span class="co">## disease3_scan | 001 | PanCK DSP-1001250007864-D-..</span></span>
<span><span class="co">## disease3_scan | 001 | neg   DSP-1001250007864-D-..</span></span>
<span><span class="co">## disease3_scan | 002 | PanCK DSP-1001250007864-D-..</span></span>
<span><span class="co">## disease3_scan | 002 | neg   DSP-1001250007864-D-..</span></span>
<span><span class="co">## disease3_scan | 003 | PanCK DSP-1001250007864-D-..</span></span></code></pre>
<p>Gene metadata are stored in the <code>rowData</code> of the
object.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">rowData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 5 rows and 3 columns</span></span>
<span><span class="co">##              RTS_ID   ProbeID  Negative</span></span>
<span><span class="co">##         &lt;character&gt; &lt;numeric&gt; &lt;logical&gt;</span></span>
<span><span class="co">## PADI2    RTS0051249     34691     FALSE</span></span>
<span><span class="co">## CYP24A1  RTS0021490     36149     FALSE</span></span>
<span><span class="co">## SUPT16H  RTS0030451     44219     FALSE</span></span>
<span><span class="co">## ZMIZ2    RTS0034986     48277     FALSE</span></span>
<span><span class="co">## SPAG9    RTS0024185     38596     FALSE</span></span></code></pre>
<p>The <code>readGeoMX</code> function has other parameters such as
<code>hasNegProbe</code> and <code>NegProbeName</code> that are opt to
deal with negative probes in the data. Then breifly talking about
negprobes in both WTA and CTA. By default, the <code>readGeoMx</code>
function will remove the negative probe, the one with name
“NegProbe-WTX”, in the count table and put it in the metadata of the
object. User can turn this off by specifying
<code>hasNegProbe = FALSE</code> in the function, just make sure no
duplicate gene names in the “TargetName” column.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">metadata</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">NegProbes</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##              disease3_scan | 001 | PanCK disease3_scan | 001 | neg</span></span>
<span><span class="co">## NegProbe-WTX                    18.08935                  34.26331</span></span>
<span><span class="co">##              disease3_scan | 002 | PanCK disease3_scan | 002 | neg</span></span>
<span><span class="co">## NegProbe-WTX                    21.55552                  40.01078</span></span>
<span><span class="co">##              disease3_scan | 003 | PanCK</span></span>
<span><span class="co">## NegProbe-WTX                    8.992909</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="import-from-dgelist-object">Import from DGEList object<a class="anchor" aria-label="anchor" href="#import-from-dgelist-object"></a>
</h3>
<p>Alternatively, <code>standR</code> provides a function to generate a
spatial experiment object from a DGEList object, which would be useful
for users who used <code>edgeR</code> package and have existing analyses
and implementations using DGEList objects to port across to the standR
workflow.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dge</span> <span class="op">&lt;-</span> <span class="fu">edgeR</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/SE2DGEList.html" class="external-link">SE2DGEList</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span></span>
<span></span>
<span><span class="va">spe2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/standR/man/readGeoMxFromDGE.html" class="external-link">readGeoMxFromDGE</a></span><span class="op">(</span><span class="va">dge</span><span class="op">)</span></span>
<span></span>
<span><span class="va">spe2</span></span></code></pre></div>
<pre><code><span><span class="co">## class: SpatialExperiment </span></span>
<span><span class="co">## dim: 18503 231 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(2): counts logcounts</span></span>
<span><span class="co">## rownames(18503): PADI2 CYP24A1 ... FAM166A AGTPBP1</span></span>
<span><span class="co">## rowData names(3): RTS_ID ProbeID Negative</span></span>
<span><span class="co">## colnames(231): disease3_scan | 001 | PanCK disease3_scan | 001 | neg</span></span>
<span><span class="co">##   ... disease1B_scan | 023 | Geometric Segment disease1B_scan | 024 |</span></span>
<span><span class="co">##   Geometric Segment</span></span>
<span><span class="co">## colData names(28): group lib.size ... RoiReportY sample_id</span></span>
<span><span class="co">## reducedDimNames(0):</span></span>
<span><span class="co">## mainExpName: NULL</span></span>
<span><span class="co">## altExpNames(0):</span></span>
<span><span class="co">## spatialCoords names(0) :</span></span>
<span><span class="co">## imgData names(0):</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="quality-control">Quality control<a class="anchor" aria-label="anchor" href="#quality-control"></a>
</h3>
<p>The recommended quality control (QC) checks for the GeoMX
transcriptome data consist of three major steps:</p>
<ol style="list-style-type: decimal">
<li><p><strong>Inspection of the sample metadata</strong>: sample
metadata can be view in<br>
tabular-like format using the <code>colData</code> function, however
here we aim to visualise the relations across the various sample
information, such as which slide did the ROIs came from, which are the
control groups and treatment groups, what are the pre-defined tissue
types etc. By doing this, we will have an overview of how<br>
the experiment was designed, the potential questions of interest, are
they clear batch effects to look out for, and the comparisons of
interest that can be established.</p></li>
<li><p><strong>Gene level QC</strong>: At the gene level, by default we
aim at removing genes that are not expressed in more than 90% of the
ROIs, and identifying any ROIs with very few genes being
expressed.<br>
This is similar to the process used in <code><a href="https://rdrr.io/pkg/edgeR/man/filterByExpr.html" class="external-link">edgeR::filterByExpr</a></code>,
as genes with consistently low counts are unlikely be identified as
significant genes. By keeping only the genes with sufficiently large
counts in the analysis, we can increase the statistical power while
reducing multiple testing burden.</p></li>
<li><p><strong>ROI level QC</strong>: At the ROI level, we aim to
identify the low-quality ROIs that have small library size (i.e. total
feature count) and low cell count. these low-quality ROIs, if not
removed would show up as isolated clusters in the dimension reduction
plots (PCA/UMAPs) and thereby affect the comparisons conducted during DE
analyses.</p></li>
</ol>
<div class="section level4">
<h4 id="sample-level-qc">Sample level QC<a class="anchor" aria-label="anchor" href="#sample-level-qc"></a>
</h4>
<p>To visualise sample metadata, we can use the
<code>plotSampleInfo</code> function. In this dataset, the following key
features are of interest for which we would like to look at: slides
(“SlideName”), diseases status (“disease_status”), tissue regions
(“region”), and the<br>
different cancer subtypes (“pathology”). These can be queries by listing
them in the function.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/standR/man/plotSampleInfo.html" class="external-link">plotSampleInfo</a></span><span class="op">(</span><span class="va">spe</span>, column2plot <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SlideName"</span>,<span class="st">"disease_status"</span>,<span class="st">"region"</span>,<span class="st">"SegmentLabel"</span>,<span class="st">"pathology"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="GeoMXAnalysisWorkflow_files/figure-html/unnamed-chunk-15-1.png" width="700"></p>
<p>Based on the known sample information, we know that all the
<em>glomerulus</em> ROIs are classified as either abnormal or healthy,
and <em>tubule</em> are classified as PanCK negative(“neg”) or positive
(“PanCK”).</p>
<p>We therefore merge the region and segment/pathology related
annotations to avoid collinearity, which can cause problems when dealing
with batch correction. Specifically:<br>
For <em>glomerulus</em>, “region” will concatenate with
“pathology”.<br>
For <em>tubule</em>, “region” will concatenate with “segmentLabel”.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">colData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">regions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">region</span>,<span class="st">"_"</span>,<span class="fu">colData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">SegmentLabel</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"_Geometric Segment"</span>,<span class="st">""</span>,<span class="va">.</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"_"</span>,<span class="fu">colData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">pathology</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"_NA"</span>,<span class="st">""</span>,<span class="va">.</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="fu">colData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">regions</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## .</span></span>
<span><span class="co">## glomerulus_abnormal  glomerulus_healthy          tubule_neg        tubule_PanCK </span></span>
<span><span class="co">##                  77                  72                  41                  41</span></span></code></pre>
<p>Re-plot the amended sample annotation to check that the changes have
been made.<br>
It is now clearer and more straightforward in terms of the comparisons
we can establish and work with.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/standR/man/plotSampleInfo.html" class="external-link">plotSampleInfo</a></span><span class="op">(</span><span class="va">spe</span>, column2plot <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SlideName"</span>,<span class="st">"disease_status"</span>,<span class="st">"regions"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="GeoMXAnalysisWorkflow_files/figure-html/unnamed-chunk-17-1.png" width="700"></p>
</div>
<div class="section level4">
<h4 id="gene-level-qc">Gene level QC<a class="anchor" aria-label="anchor" href="#gene-level-qc"></a>
</h4>
<p>Now we check on the gene level data. Using the
<code>addPerROIQC</code> function, we can add<br>
key statistics to the <code>colData</code> of the object. For the
purpose of this exercise,<br>
we will set the argument <code>rm_genes</code> to TRUE, and keeping the
default settings of<br><code>min_count = 5</code> and <code>sample_fraction = 0.9</code>. We
first calculate the expression<br>
threshold using the logCPM data (to account for library size
variations), we then<br>
filter out the genes with low-expression values that’s below the set
threshold in<br>
more than 90% of the ROIs.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/standR/man/addPerROIQC.html" class="external-link">addPerROIQC</a></span><span class="op">(</span><span class="va">spe</span>, rm_genes <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Looking at the object again, we can see that 121 genes were
removed.</p>
<p>The count matrix of the genes that were removed will be stored in the
<code>metadata</code><br>
of the object with prefix <em>genes_rm</em>, alongside the calculated
expression<br>
threshold (<em>lcpm_threshold</em>).</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 18382   231</span></span></code></pre>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">metadata</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "NegProbes"         "lcpm_threshold"    "genes_rm_rawCount"</span></span>
<span><span class="co">## [4] "genes_rm_logCPM"</span></span></code></pre>
<p>Using the <code>plotGeneQC</code> function, we can then assess the
logCPM expressions of the<br>
genes that were removed across the samples. The function also plots the
histogram<br>
/distribution of the proportion of non-expressed genes in all the ROIs
(as a<br>
percentage). By default, the top 9 genes are plotted here (ordered by
the mean expression). Users can customise the number of genes plotted
using the parameter <code>top_n</code>.</p>
<p>Moreover, users can order the samples (using parameter
<code>ordannots</code>) or color/shape<br>
the dots by specific annotation to better compare and assess for
specific biological or experimental factors which are influencing how
these genes were expressed across the samples (e.g. gene may be higly
expression in particular tissue types or under particular treatment
conditions. This genes are to be access or curated by domain experts to
ascertain or determine if any of these genes are of
biological/experimental significance. This provides a potential warning
to whether the experiments have worked as per intended.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/standR/man/plotGeneQC.html" class="external-link">plotGeneQC</a></span><span class="op">(</span><span class="va">spe</span>, ordannots <span class="op">=</span> <span class="st">"regions"</span>, col <span class="op">=</span> <span class="va">regions</span>, point_size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="GeoMXAnalysisWorkflow_files/figure-html/unnamed-chunk-20-1.png" width="700"></p>
</div>
<div class="section level4">
<h4 id="qc-on-rois">QC on ROIs<a class="anchor" aria-label="anchor" href="#qc-on-rois"></a>
</h4>
</div>
</div>
<div class="section level3">
<h3 id="roi-level-qc">ROI level QC<a class="anchor" aria-label="anchor" href="#roi-level-qc"></a>
</h3>
<p>After clecking the genes, we can now look at the ROI level data.
Using the <code>plotROIQC</code> function, we can visualise QC
statistics at the ROI level. By default, the library size and cell count
(AOINucleiCount) will be computed.</p>
<p>In the ROI level QC, we first aim to identify (if any) ROI(s) that
have relatively low library size and low cell count because they are
considered as low quality samples due to insufficient sequencing depth
or the lack of RNA in the selected region. Frequency histograms are also
provided for both library size and nuclei count to assist with assessing
any abnormal distributions of samples in the data.</p>
<p>In this case, we assess the distribution plot for library size
against the nuclei count. Looking at the scatter plot, we expect the
library sizes to be mostly positively correlated with the cell count
(i.e. nuclei count). It is not unexpected for there to be some ROIs
having relatively low library size and having a reasonable number of
cells (nuclei count). In this dataset, we see this library size vs cell
count relationship is relatively smooth with no aberrations observed
(e.g. spikes at the lower ranges)</p>
<p>To remove/filter low quality samples, we define a filtering threshold
near the lower end of the library size range, in this case at 50,000. By
stratifying (color) the points based on their slide names, we not that
the ROIs below the threshold are all from slide <em>disease1B</em> (pale
orange). This might suggest potential of some technical errors
associated with slide <em>disease1B</em>.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/standR/man/plotROIQC.html" class="external-link">plotROIQC</a></span><span class="op">(</span><span class="va">spe</span>, y_threshold <span class="op">=</span> <span class="fl">50000</span>, color <span class="op">=</span> <span class="va">SlideName</span><span class="op">)</span></span></code></pre></div>
<p><img src="GeoMXAnalysisWorkflow_files/figure-html/unnamed-chunk-21-1.png" width="700"></p>
<p>In this experiment, based on the above plot, the library size
threshold of 50,000 looks to be a reasonable cutoff. As such we subset
the spatial experiment object based<br>
on the library size using this threshold in <code>colData</code>.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">qc</span> <span class="op">&lt;-</span> <span class="fu">colData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">lib_size</span> <span class="op">&gt;</span> <span class="fl">50000</span></span>
<span><span class="va">spe</span> <span class="op">&lt;-</span> <span class="va">spe</span><span class="op">[</span>, <span class="va">qc</span><span class="op">]</span></span></code></pre></div>
<p><strong>Note: The same workflow and logic can also be applied to the
nuclei count. We will also check the gene detection rate to assess the
quality of the dataset.</strong></p>
<p>The function <code>plotROIQC</code> is looking at nuclei count and
library size of each ROI by default, user can change the x or y axis to
any other statics they want to QC by specifying the parameter
<code>x_axis</code> or <code>y_axis</code>. For example, the statistics
limit of quantification (LOQ) is summarised by the Nanostring process
pipeline and it’s provided in most/all of the GeoMX DSP dataset. This
metrics is calculated by on the distribution of the negative probes, and
it’s suggested by Nanostring that LOQ can be useful when there are large
segments (i.e. big ROI or high cell count samples), and a minimum LOQ of
2 is recommended.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/standR/man/plotROIQC.html" class="external-link">plotROIQC</a></span><span class="op">(</span><span class="va">spe</span>,  y_axis <span class="op">=</span> <span class="st">"LOQ"</span>, y_lab <span class="op">=</span> <span class="st">"LOQ"</span>, x_axis <span class="op">=</span> <span class="st">"lib_size"</span>, x_lab <span class="op">=</span> <span class="st">"Library size"</span>,  y_threshold <span class="op">=</span> <span class="fl">2</span>, col <span class="op">=</span> <span class="va">SlideName</span><span class="op">)</span></span></code></pre></div>
<p><img src="GeoMXAnalysisWorkflow_files/figure-html/unnamed-chunk-23-1.png" width="700"></p>
<div class="section level4">
<h4 id="relative-log-expression-distribution">Relative log expression distribution<a class="anchor" aria-label="anchor" href="#relative-log-expression-distribution"></a>
</h4>
<p>After filtering, we will use function <code>plotRLExpr</code> to
visualise the relative log expression (RLE) of the data to identify any
technical variation that may be present in the dataset. We looks at the
relative distance between the median of the RLE for each ROI (the dot in
the boxplot) to zero.</p>
<p>By default, we plot the RLE of the raw count, where we expect to see
majority of the variation to be contributed by differences in library
size.</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/standR/man/plotRLExpr.html" class="external-link">plotRLExpr</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span></span></code></pre></div>
<p><img src="GeoMXAnalysisWorkflow_files/figure-html/unnamed-chunk-24-1.png" width="700"></p>
<p>By using <code>assay = 2</code> to run RLE on the logCPM data, we can
see that most of the technical variations due to library size
differences are removed.</p>
<p>We can follow up by sorting the data based on the different sample
metadata annotations by specifying the <code>ordannots</code> parameter.
This can be visualised either with color or shape mapping parameters
(based on similar approaches for plotting in <code>ggplot</code>),
enabling quick assessment of the possible factors that’s contributing to
the observed technical variation(s).</p>
<p>In this case, we stratify by <em>slideName</em> using different
colors which clearly show substantial variations between the slides as
well as to a lesser extent within each slide.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/standR/man/plotRLExpr.html" class="external-link">plotRLExpr</a></span><span class="op">(</span><span class="va">spe</span>, ordannots <span class="op">=</span> <span class="st">"SlideName"</span>, assay <span class="op">=</span> <span class="fl">2</span>, color <span class="op">=</span> <span class="va">SlideName</span><span class="op">)</span></span></code></pre></div>
<p><img src="GeoMXAnalysisWorkflow_files/figure-html/unnamed-chunk-25-1.png" width="700"></p>
<p>We can also try out other factors (e.g. regions ) to see how they
influence the expression.</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/standR/man/plotRLExpr.html" class="external-link">plotRLExpr</a></span><span class="op">(</span><span class="va">spe</span>, ordannots <span class="op">=</span> <span class="st">"regions"</span>, assay <span class="op">=</span> <span class="fl">2</span>, color <span class="op">=</span> <span class="va">regions</span><span class="op">)</span></span></code></pre></div>
<p><img src="GeoMXAnalysisWorkflow_files/figure-html/unnamed-chunk-26-1.png" width="700"></p>
</div>
<div class="section level4">
<h4 id="dimension-reduction">Dimension reduction<a class="anchor" aria-label="anchor" href="#dimension-reduction"></a>
</h4>
<div class="section level5">
<h5 id="pca">PCA<a class="anchor" aria-label="anchor" href="#pca"></a>
</h5>
<p>We can also look at the data by breaking it down into the lower
dimensions.</p>
<p>Using the <code>drawPCA</code> function, we can perform principal
component analysis (PCA) on the data. The PCA can help visualise any
potential systemic variations (both biological and technical) in the
data and to identify the main factors contributing to the
variations.</p>
<p>Here we stratify the points based on slide information
(<em>SlideName</em> by color) and tissues regions (<em>regions</em>
using shapes). It is clear that tissue regions /types can be explained
on PC1, particularly between <em>glomerulus</em> and <em>tubule</em>.
ROIs group and clusters by the individual slide within each tissue
structure. Note that the subtypes in the <em>tubule</em> ROIs are are
clearly separated along PC2. This is not the case for the different
subtypes of glomerulus (which groups together).</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/standR/man/drawPCA.html" class="external-link">drawPCA</a></span><span class="op">(</span><span class="va">spe</span>, assay <span class="op">=</span> <span class="fl">2</span>, col <span class="op">=</span> <span class="va">SlideName</span>, shape <span class="op">=</span> <span class="va">regions</span><span class="op">)</span></span></code></pre></div>
<p><img src="GeoMXAnalysisWorkflow_files/figure-html/unnamed-chunk-27-1.png" width="1440"></p>
<p>Diseased tissues and control tissues are mixed (disease slides and
normal slides).</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/standR/man/drawPCA.html" class="external-link">drawPCA</a></span><span class="op">(</span><span class="va">spe</span>, assay <span class="op">=</span> <span class="fl">2</span>, col <span class="op">=</span> <span class="va">disease_status</span>, shape <span class="op">=</span> <span class="va">regions</span><span class="op">)</span></span></code></pre></div>
<p><img src="GeoMXAnalysisWorkflow_files/figure-html/unnamed-chunk-28-1.png" width="1440"></p>
<p>The <code>standR</code> package also provide other functions to
visualise the PCA, including the PCA scree plot, pair-dimension PCA plot
and PCA bi-plot.</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/standR/man/plotScreePCA.html" class="external-link">plotScreePCA</a></span><span class="op">(</span><span class="va">spe</span>, assay <span class="op">=</span> <span class="fl">2</span>, dims <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<p><img src="GeoMXAnalysisWorkflow_files/figure-html/unnamed-chunk-29-1.png" width="700"></p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/standR/man/plotPairPCA.html" class="external-link">plotPairPCA</a></span><span class="op">(</span><span class="va">spe</span>, col <span class="op">=</span> <span class="va">disease_status</span>, </span>
<span>            shape <span class="op">=</span> <span class="va">regions</span>, assay <span class="op">=</span> <span class="fl">2</span>, n_dimension <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span></code></pre></div>
<p><img src="GeoMXAnalysisWorkflow_files/figure-html/unnamed-chunk-30-1.png" width="1440"></p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/standR/man/plotPCAbiplot.html" class="external-link">plotPCAbiplot</a></span><span class="op">(</span><span class="va">spe</span>, n_loadings <span class="op">=</span> <span class="fl">10</span>, assay <span class="op">=</span> <span class="fl">2</span>, col <span class="op">=</span> <span class="va">regions</span><span class="op">)</span></span></code></pre></div>
<p><img src="GeoMXAnalysisWorkflow_files/figure-html/unnamed-chunk-31-1.png" width="700"></p>
</div>
<div class="section level5">
<h5 id="mds">MDS<a class="anchor" aria-label="anchor" href="#mds"></a>
</h5>
<p>Another way to visualise the data is to look at the Multidimensional
scaling (MDS) plots. The function <code>plotMDS</code> provides the
means to visualise the data in this way.</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">standR</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/standR/man/plotMDS.html" class="external-link">plotMDS</a></span><span class="op">(</span><span class="va">spe</span>, assay <span class="op">=</span> <span class="fl">2</span>, color <span class="op">=</span> <span class="va">disease_status</span>, shape <span class="op">=</span> <span class="va">regions</span><span class="op">)</span></span></code></pre></div>
<p><img src="GeoMXAnalysisWorkflow_files/figure-html/unnamed-chunk-32-1.png" width="700"></p>
</div>
<div class="section level5">
<h5 id="umap">UMAP<a class="anchor" aria-label="anchor" href="#umap"></a>
</h5>
<p>Furthermore, since we’re using <code>SpatialExperiment</code> as our
infrastructure, we are able to incorporate or apply other widely used
packages such as <code>scater</code>, which is commonly used in single
cell data and spatial 10x genomics visium data analyses. We also provide
the function <code>plotDR</code> to visualise any dimension reduction
results generated using <code>scater::run*</code>, such as UMAP, TSNE
and NMF. This can be done by simply specifying the <code>dimred</code>
parameter.</p>
<p>Here we plot the UMAP of our data. Similar variations can be
generated for other approaches like PCA and MDS as discussed
earlier.</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html" class="external-link">runUMAP</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/standR/man/plotDR.html" class="external-link">plotDR</a></span><span class="op">(</span><span class="va">spe</span>, dimred <span class="op">=</span> <span class="st">"UMAP"</span>, col <span class="op">=</span> <span class="va">regions</span><span class="op">)</span></span></code></pre></div>
<p><img src="GeoMXAnalysisWorkflow_files/figure-html/unnamed-chunk-33-1.png" width="700"></p>
</div>
</div>
</div>
<div class="section level3">
<h3 id="normalization">Normalization<a class="anchor" aria-label="anchor" href="#normalization"></a>
</h3>
<p>If there are observed technical variations identified in the earlier
QC steps, before proceeding with any analysis of the data, it is
necessary to appropriately perform normalisation of the data to
rectify/minimise the identified variation.</p>
<p>The <code>standR</code> package offers normalization options
including TMM, RPKM, TPM, CPM, upperquartile and sizefactor. Among
these, RPKM and TPM require gene length information (add
<code>genelength</code> column to the <code>rowData</code> of the
object). For TMM, upperquartile and sizefactor, their normalized factor
will be stored in their <code>metadata</code>.</p>
<p>Here we used TMM to normalize the data.</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe_tmm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/standR/man/geomxNorm.html" class="external-link">geomxNorm</a></span><span class="op">(</span><span class="va">spe</span>, method <span class="op">=</span> <span class="st">"TMM"</span><span class="op">)</span></span></code></pre></div>
<p>To assess how well the normalisation was able to remove unwanted
variataions, we make use of RLE and PCA plot in conjunction with the
factors of interest.</p>
<p>In this case, from the resulting RLE plot, most of the medians of RLE
are close to zero, suggesting that most of the technical variation have
been removed.</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/standR/man/plotRLExpr.html" class="external-link">plotRLExpr</a></span><span class="op">(</span><span class="va">spe_tmm</span>, assay <span class="op">=</span> <span class="fl">2</span>, color <span class="op">=</span> <span class="va">SlideName</span><span class="op">)</span> <span class="op">+</span> <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"TMM"</span><span class="op">)</span></span></code></pre></div>
<p><img src="GeoMXAnalysisWorkflow_files/figure-html/unnamed-chunk-35-1.png" width="700"></p>
<p>However, from the PCA plots, the batch effect due to the different
slides are still being observed, confounding the known biology of
interest (which is between disease and normal).</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/standR/man/plotPairPCA.html" class="external-link">plotPairPCA</a></span><span class="op">(</span><span class="va">spe_tmm</span>, assay <span class="op">=</span> <span class="fl">2</span>, color <span class="op">=</span> <span class="va">disease_status</span>, shape <span class="op">=</span> <span class="va">regions</span><span class="op">)</span></span></code></pre></div>
<p><img src="GeoMXAnalysisWorkflow_files/figure-html/unnamed-chunk-36-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="batch-correction">Batch correction<a class="anchor" aria-label="anchor" href="#batch-correction"></a>
</h3>
<p>In the Nanostring’s GeoMX DSP protocol, each slide is typically only
able to<br>
fit a handful of tissue segments (Tissue microarrays/FFPE cores), it is
common<br>
that DSP data are confounded by the batch effect introduced by the
different<br>
slides. In order to establish appropriate comparisons between the ROIs
in the downstream analyses, it is necessary to remove this batch effect
from the data.</p>
<p>In the <code>standR</code> package, we provide two approaches for
removing batch effects (RUV4 and Limma), more methods (e.g. RUVg) are
included in the development version.</p>
<div class="section level4">
<h4 id="correction-method-remove-unwanted-variation-4-ruv4">Correction method : Remove Unwanted Variation 4 (RUV4)<a class="anchor" aria-label="anchor" href="#correction-method-remove-unwanted-variation-4-ruv4"></a>
</h4>
<p>Remove Unwanted Variation 4 (RUV4) is a method developed by Terry
Speed and Johann Gagnon-Bartsch to use negative control genes to remove
unwanted variations, see the published paper <a href="https://academic.oup.com/biostatistics/article/13/3/539/248166" class="external-link">here</a>.</p>
<p>To run batch correction using RUV4, a list of “negative control genes
(NCGs)” will be required.</p>
<p><code>standR</code> provides a function <code>findNCGs</code> which
identify the NCGs from the data. In this case, since the batch effect is
mostly due to slide effects, we aim to identify NCGs across all the
slides. As such, the <code>batch_name</code> parameter was set to
“SlideName”, and the top 300 least variable genes (ranked by coefficient
of variation) across different slides were identified as NCGs. These are
stored in the object as “NCGs”.</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/standR/man/findNCGs.html" class="external-link">findNCGs</a></span><span class="op">(</span><span class="va">spe</span>, batch_name <span class="op">=</span> <span class="st">"SlideName"</span>, top_n <span class="op">=</span> <span class="fl">300</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">metadata</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "NegProbes"         "lcpm_threshold"    "genes_rm_rawCount"</span></span>
<span><span class="co">## [4] "genes_rm_logCPM"   "NCGs"</span></span></code></pre>
<p>Now we run RUV4 using the function <code>geomxBatchCorrection</code>.
By default this function will use <code>RUV4</code> to normalize the
data.</p>
<p>For RUV4 correction, the function requires 3 addition parameters
other than the input object:</p>
<ul>
<li>
<code>factors</code>: the factor of interest, i.e. the biological
variation to keep;</li>
<li>
<code>NCGs</code>: the list of negative control genes detected using
the function <code>findNCGs</code>;</li>
<li>
<code>k</code>: the number of unwanted factors to use. Based on
RUV’s documentation, it is suggest that to use the smallest k possible
where the observed technical variation is no longer observed.</li>
</ul>
<p>Choosing the optimal k is one of the most important task when
performing batch correction using RUV. The best way to do so is to test
out each k and assess the corresponding diagnostic plot (e.g. PCA). The
optimal k would be the smallest value that produces a separation of the
main biology of interest of the experiment on a PCA plot.</p>
<p>So here we run through the paired PCA plots for k between 1 and
5.</p>
<p>In this case, we create a combined factor of interest
<em>biology</em> (concatenate <em>disease_status</em> and
<em>regions</em> information) in the object. This factor will be
specified in the <code>factors</code> parameter of the function.</p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">colData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">biology</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">disease_status</span>, <span class="st">"_"</span>, <span class="fu">colData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">regions</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">5</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">spe_ruv</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/standR/man/geomxBatchCorrection.html" class="external-link">geomxBatchCorrection</a></span><span class="op">(</span><span class="va">spe</span>, factors <span class="op">=</span> <span class="st">"biology"</span>, </span>
<span>                   NCGs <span class="op">=</span> <span class="fu">metadata</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">NCGs</span>, k <span class="op">=</span> <span class="va">i</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/standR/man/plotPairPCA.html" class="external-link">plotPairPCA</a></span><span class="op">(</span><span class="va">spe_ruv</span>, assay <span class="op">=</span> <span class="fl">2</span>, n_dimension <span class="op">=</span> <span class="fl">4</span>, color <span class="op">=</span> <span class="va">disease_status</span>, shape <span class="op">=</span> <span class="va">regions</span>, title <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"k = "</span>, <span class="va">i</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>  </span>
<span><span class="op">}</span></span></code></pre></div>
<p><img src="GeoMXAnalysisWorkflow_files/figure-html/unnamed-chunk-38-1.png" width="960"><img src="GeoMXAnalysisWorkflow_files/figure-html/unnamed-chunk-38-2.png" width="960"><img src="GeoMXAnalysisWorkflow_files/figure-html/unnamed-chunk-38-3.png" width="960"><img src="GeoMXAnalysisWorkflow_files/figure-html/unnamed-chunk-38-4.png" width="960"><img src="GeoMXAnalysisWorkflow_files/figure-html/unnamed-chunk-38-5.png" width="960"></p>
<p>After assessing the generated PCA plots, we choose k = 4 to be our
best k. From the resulting PCA, we can see that the disease status are
reasonable<br>
separated within each region type.</p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe_ruv</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/standR/man/geomxBatchCorrection.html" class="external-link">geomxBatchCorrection</a></span><span class="op">(</span><span class="va">spe</span>, factors <span class="op">=</span> <span class="st">"biology"</span>, </span>
<span>                   NCGs <span class="op">=</span> <span class="fu">metadata</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">NCGs</span>, k <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/standR/man/plotPairPCA.html" class="external-link">plotPairPCA</a></span><span class="op">(</span><span class="va">spe_ruv</span>, assay <span class="op">=</span> <span class="fl">2</span>, color <span class="op">=</span> <span class="va">disease_status</span>, shape <span class="op">=</span> <span class="va">regions</span>, title <span class="op">=</span> <span class="st">"RUV4, k = 4"</span>, n_dimension <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span></code></pre></div>
<p><img src="GeoMXAnalysisWorkflow_files/figure-html/unnamed-chunk-39-1.png" width="960"></p>
</div>
<div class="section level4">
<h4 id="correction-method-limma">Correction method: limma<a class="anchor" aria-label="anchor" href="#correction-method-limma"></a>
</h4>
<p>The other available batch correction method is based on the
<code>removeBatchEffect</code> function from the bioconductor package
<code>limma</code>, more details of the method can see paper <a href="https://academic.oup.com/nar/article/43/7/e47/2414268" class="external-link">here</a>.</p>
<p>To use the limma batch correction, set the parameter
<code>method</code> to “Limma”, which uses the remove batch correction
method from <code>limma</code> package. In this mode, the function
requires 2 addition parameters other than the input object:</p>
<ul>
<li>
<code>batch</code>: a vector that indicating the batch information
for all samples;</li>
<li>
<code>design</code>: a design matrix generated from
<code>model.matrix</code>, in the design matrix, all
biologically-relevant factors should be included.</li>
</ul>
<p>In this case, the batch effect is based on the slides (SlideName) and
factors of interest includes “disease_status” and “regions”.</p>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe_lrb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/standR/man/geomxBatchCorrection.html" class="external-link">geomxBatchCorrection</a></span><span class="op">(</span><span class="va">spe</span>,</span>
<span>                       batch <span class="op">=</span> <span class="fu">colData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">SlideName</span>, method <span class="op">=</span> <span class="st">"Limma"</span>,</span>
<span>                       design <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="op">~</span> <span class="va">disease_status</span> <span class="op">+</span> <span class="va">regions</span>, </span>
<span>                                             data <span class="op">=</span> <span class="fu">colData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">)</span><span class="op">[</span>,<span class="op">-</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p>Once again, we use the respective QC plots like PCA to inspect and
assess the effectiveness of the applied batch correction process.</p>
<p>In this instance, using <code><a href="https://rdrr.io/pkg/limma/man/removeBatchEffect.html" class="external-link">limma::removeBatchEffect</a></code>
approach has effectively separated the samples based on the disease
status on PC1. Furthermore, the samples are also reasonable well
clusters and separated based on tissue regions (particular for tubules)
across PC2 and PC3</p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/standR/man/plotPairPCA.html" class="external-link">plotPairPCA</a></span><span class="op">(</span><span class="va">spe_lrb</span>, assay <span class="op">=</span> <span class="fl">2</span>, color <span class="op">=</span> <span class="va">disease_status</span>, shape <span class="op">=</span> <span class="va">regions</span>, title <span class="op">=</span> <span class="st">"Limma removeBatch"</span>, n_dimension <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span></code></pre></div>
<p><img src="GeoMXAnalysisWorkflow_files/figure-html/unnamed-chunk-41-1.png" width="700"></p>
</div>
<div class="section level4">
<h4 id="evaluation">Evaluation<a class="anchor" aria-label="anchor" href="#evaluation"></a>
</h4>
<div class="section level5">
<h5 id="summary-statistics">Summary statistics<a class="anchor" aria-label="anchor" href="#summary-statistics"></a>
</h5>
<p>The typical approach to interrogating the effectiveness of batch
correction process on the data uses dimension reduction plots like PCAs.
Here we further suggest the use of summarized statistics to assess the
effectiveness of batch correction. The 6 summarized statistics tested in
this package includes:</p>
<ul>
<li>adjusted rand index.<br>
</li>
<li>jaccard similarity coefficient.<br>
</li>
<li>silhouette coefficient.<br>
</li>
<li>Chi-squared coefficient.</li>
<li>Mirkin distance.<br>
</li>
<li>Simple matching coefficient.</li>
</ul>
<p>This assessment can be conducted by using the
<code>plotClusterEvalStats</code> function provided in
<code>standR</code>.</p>
<p>Here we present the output for the summarised statistics for the two
normalization methods in this workshop (i.e. RUV4 and Limma ) as well as
the uncorrected data. Scores for each method will be presented as a
barplot scores of the six summarized statistics (as above) under two
sections (<em>biology</em> and <em>batch</em>). As a general rule, for
the <em>biology</em> of interest defined in the batch correction
process, a higher score is considered a good outcome. On the other hand,
for the batch, a smaller score will be the preferred outcome.</p>
<p>We can see from the results that when it comes to stratifying based
on biological factors (disease status and tissue regions) or quantifying
the amount of clustering due to batch effects for this dataset,
<code>RUV4</code> outperforms <code>limma</code> in most cases.</p>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">spe</span>, <span class="va">spe_ruv</span>, <span class="va">spe_lrb</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/standR/man/plotClusterEvalStats.html" class="external-link">plotClusterEvalStats</a></span><span class="op">(</span>spe_list <span class="op">=</span> <span class="va">spe_list</span>,</span>
<span>                     bio_feature_name <span class="op">=</span> <span class="st">"regions"</span>,</span>
<span>                     batch_feature_name <span class="op">=</span> <span class="st">"SlideName"</span>,</span>
<span>                     data_names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Raw"</span>,<span class="st">"RUV4"</span>,<span class="st">"Limma"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="GeoMXAnalysisWorkflow_files/figure-html/unnamed-chunk-42-1.png" width="700"></p>
</div>
<div class="section level5">
<h5 id="rle-plots">RLE plots<a class="anchor" aria-label="anchor" href="#rle-plots"></a>
</h5>
<p>In addition, we can visualize the outcomes using RLE plots of
normalized count to determine which batch correction performs better for
this dataset.</p>
<p>Plotting both the RLEs for RUV4-corrected and limma-corrected data,
we can clearly see that RUV4-corrected data has in general a
more-closer-to-zero median RLE when compared to that for the
limma-corrected data. Therefore, for the downstream differential
expression analysis, we would suggest using RUV4 as the batch correction
method for this specific dataset.</p>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/standR/man/plotRLExpr.html" class="external-link">plotRLExpr</a></span><span class="op">(</span><span class="va">spe_ruv</span>, assay <span class="op">=</span> <span class="fl">2</span>, color <span class="op">=</span> <span class="va">SlideName</span><span class="op">)</span> <span class="op">+</span> <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"RUV4"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/standR/man/plotRLExpr.html" class="external-link">plotRLExpr</a></span><span class="op">(</span><span class="va">spe_lrb</span>, assay <span class="op">=</span> <span class="fl">2</span>, color <span class="op">=</span> <span class="va">SlideName</span><span class="op">)</span> <span class="op">+</span> <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"Limma removeBatch"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="differential-expression-analysis-with-limma-voom-pipeline">Differential expression analysis with limma-voom pipeline<a class="anchor" aria-label="anchor" href="#differential-expression-analysis-with-limma-voom-pipeline"></a>
</h2>
<p>For the downstream analyses like differential expression analyses,
standR does not provide specific functions. Instead, we recommend
incorporating the workflow with well established pipelines, such as
<code>edgeR</code>, <code>limma-voom</code> or <code>DESeq2</code>.
These pipelines uses linear modelling which borrow information from all
genes, making it more appropriate for complex dataset with various
experimental factors. A simple T-test is definitely not recommended for
performing DE analysis of GeoMX DSP data.</p>
<p>In this workshop, we’ll demonstrate the DE analysis using the
<code>limma-voom</code> pipeline.</p>
<div class="section level3">
<h3 id="linear-modelling">Linear modelling<a class="anchor" aria-label="anchor" href="#linear-modelling"></a>
</h3>
<p>We’ve shown in previous sections that for this dataset, using RUV4
with <code>k = 4</code> is the appropriate batch correction approach and
is able to remove the batch effect and other undesired technical
variations. To incorporate this correction into the linear model, we
need to include the weight matrix generated from the function
<code>geomxBatchCorrection</code> as covariates. The weight matrix can
be found in the <code>colData</code>.</p>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioinf.wehi.edu.au/edgeR" class="external-link">edgeR</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioinf.wehi.edu.au/limma" class="external-link">limma</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">dge</span> <span class="op">&lt;-</span> <span class="fu">edgeR</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/SE2DGEList.html" class="external-link">SE2DGEList</a></span><span class="op">(</span><span class="va">spe_ruv</span><span class="op">)</span></span>
<span></span>
<span><span class="va">design</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="op">~</span><span class="fl">0</span> <span class="op">+</span> <span class="va">biology</span> <span class="op">+</span> <span class="va">ruv_W1</span> <span class="op">+</span> <span class="va">ruv_W2</span> <span class="op">+</span> <span class="va">ruv_W3</span> <span class="op">+</span> <span class="va">ruv_W4</span>, data <span class="op">=</span> <span class="fu">colData</span><span class="op">(</span><span class="va">spe_ruv</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">design</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">design</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">str_replace</span><span class="op">(</span><span class="st">"biology"</span>,<span class="st">""</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">design</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "DKD_glomerulus_abnormal"    "DKD_glomerulus_healthy"    </span></span>
<span><span class="co">##  [3] "DKD_tubule_neg"             "DKD_tubule_PanCK"          </span></span>
<span><span class="co">##  [5] "normal_glomerulus_abnormal" "normal_glomerulus_healthy" </span></span>
<span><span class="co">##  [7] "normal_tubule_neg"          "normal_tubule_PanCK"       </span></span>
<span><span class="co">##  [9] "ruv_W1"                     "ruv_W2"                    </span></span>
<span><span class="co">## [11] "ruv_W3"                     "ruv_W4"</span></span></code></pre>
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">keep</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/filterByExpr.html" class="external-link">filterByExpr</a></span><span class="op">(</span><span class="va">dge</span>, <span class="va">design</span><span class="op">)</span></span>
<span><span class="va">dge_all</span> <span class="op">&lt;-</span> <span class="va">dge</span><span class="op">[</span><span class="va">keep</span>, <span class="op">]</span></span>
<span></span>
<span><span class="va">dge_all</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/estimateDisp.html" class="external-link">estimateDisp</a></span><span class="op">(</span><span class="va">dge_all</span>, <span class="va">design</span>, robust<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="biological-coefficient-of-variation">Biological coefficient of variation<a class="anchor" aria-label="anchor" href="#biological-coefficient-of-variation"></a>
</h3>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/plotBCV.html" class="external-link">plotBCV</a></span><span class="op">(</span><span class="va">dge_all</span><span class="op">)</span></span>
<span><span class="va">bcv_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  <span class="st">'BCV'</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="va">dge_all</span><span class="op">$</span><span class="va">tagwise.dispersion</span><span class="op">)</span>,</span>
<span>  <span class="st">'AveLogCPM'</span> <span class="op">=</span> <span class="va">dge_all</span><span class="op">$</span><span class="va">AveLogCPM</span>,</span>
<span>  <span class="st">'gene_id'</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">dge_all</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">highbcv</span> <span class="op">&lt;-</span> <span class="va">bcv_df</span><span class="op">$</span><span class="va">AveLogCPM</span> <span class="op">&gt;</span> <span class="fl">7</span> <span class="op">&amp;</span> <span class="va">bcv_df</span><span class="op">$</span><span class="va">BCV</span> <span class="op">&gt;</span> <span class="fl">0.8</span></span>
<span><span class="va">highbcv_df</span> <span class="op">&lt;-</span> <span class="va">bcv_df</span><span class="op">[</span><span class="va">highbcv</span>, <span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/points.html" class="external-link">points</a></span><span class="op">(</span><span class="va">highbcv_df</span><span class="op">$</span><span class="va">AveLogCPM</span>, <span class="va">highbcv_df</span><span class="op">$</span><span class="va">BCV</span>, col <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/text.html" class="external-link">text</a></span><span class="op">(</span><span class="va">highbcv_df</span><span class="op">$</span><span class="va">AveLogCPM</span>, <span class="va">highbcv_df</span><span class="op">$</span><span class="va">BCV</span>, labels <span class="op">=</span> <span class="va">highbcv_df</span><span class="op">$</span><span class="va">gene_id</span>, pos <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span></code></pre></div>
<p><img src="GeoMXAnalysisWorkflow_files/figure-html/unnamed-chunk-45-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="differential-expression">Differential expression<a class="anchor" aria-label="anchor" href="#differential-expression"></a>
</h3>
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">v</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/voom.html" class="external-link">voom</a></span><span class="op">(</span><span class="va">dge_all</span>, <span class="va">design</span><span class="op">)</span> </span>
<span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/lmFit.html" class="external-link">lmFit</a></span><span class="op">(</span><span class="va">v</span><span class="op">)</span></span>
<span></span>
<span><span class="va">contr.matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/makeContrasts.html" class="external-link">makeContrasts</a></span><span class="op">(</span></span>
<span>   DvCinGab <span class="op">=</span> <span class="va">DKD_glomerulus_abnormal</span> <span class="op">-</span> <span class="va">normal_glomerulus_abnormal</span>,</span>
<span>   levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">design</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">fit_contrast</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/contrasts.fit.html" class="external-link">contrasts.fit</a></span><span class="op">(</span><span class="va">fit</span>, contrasts <span class="op">=</span> <span class="va">contr.matrix</span><span class="op">)</span></span>
<span></span>
<span><span class="va">efit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/ebayes.html" class="external-link">eBayes</a></span><span class="op">(</span><span class="va">fit_contrast</span>, robust <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">results_efit</span><span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/decideTests.html" class="external-link">decideTests</a></span><span class="op">(</span><span class="va">efit</span>, p.value <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span>
<span><span class="va">summary_efit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">results_efit</span><span class="op">)</span></span></code></pre></div>
<div class="section level4">
<h4 id="visualization">Visualization<a class="anchor" aria-label="anchor" href="#visualization"></a>
</h4>
<p>We can then visualise the DE genes using plots such as volcano plot
and MA plot.</p>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/slowkow/ggrepel" class="external-link">ggrepel</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">de_results_DvCinGab</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/toptable.html" class="external-link">topTable</a></span><span class="op">(</span><span class="va">efit</span>, coef <span class="op">=</span> <span class="fl">1</span>, sort.by <span class="op">=</span> <span class="st">"P"</span>, n <span class="op">=</span> <span class="cn">Inf</span><span class="op">)</span></span>
<span></span>
<span><span class="va">de_genes_toptable_DvCinGab</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/toptable.html" class="external-link">topTable</a></span><span class="op">(</span><span class="va">efit</span>, coef <span class="op">=</span> <span class="fl">1</span>, sort.by <span class="op">=</span> <span class="st">"P"</span>, n <span class="op">=</span> <span class="cn">Inf</span>, p.value <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span>
<span></span>
<span></span>
<span></span>
<span><span class="va">de_results_DvCinGab</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>DE <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">logFC</span> <span class="op">&gt;</span> <span class="fl">0</span> <span class="op">&amp;</span> <span class="va">adj.P.Val</span> <span class="op">&lt;</span><span class="fl">0.05</span>, <span class="st">"UP"</span>, </span>
<span>                       <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">logFC</span> <span class="op">&lt;</span><span class="fl">0</span> <span class="op">&amp;</span> <span class="va">adj.P.Val</span><span class="op">&lt;</span><span class="fl">0.05</span>, <span class="st">"DOWN"</span>, <span class="st">"NOT DE"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">AveExpr</span>, <span class="va">logFC</span>, col <span class="op">=</span> <span class="va">DE</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>shape <span class="op">=</span> <span class="fl">1</span>, size <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/ggrepel/man/geom_text_repel.html" class="external-link">geom_text_repel</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">de_genes_toptable_DvCinGab</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>                    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>DE <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">logFC</span> <span class="op">&gt;</span> <span class="fl">0</span> <span class="op">&amp;</span> <span class="va">adj.P.Val</span> <span class="op">&lt;</span><span class="fl">0.05</span>, <span class="st">"UP"</span>, </span>
<span>                       <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">logFC</span> <span class="op">&lt;</span><span class="fl">0</span> <span class="op">&amp;</span> <span class="va">adj.P.Val</span><span class="op">&lt;</span><span class="fl">0.05</span>, <span class="st">"DOWN"</span>, <span class="st">"NOT DE"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>                    <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html" class="external-link">rownames_to_column</a></span><span class="op">(</span><span class="op">)</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>label <span class="op">=</span> <span class="va">rowname</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"Average log-expression"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"Log-fold-change"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"DKD vs Normal in abnormal glomerulus (limma-voom)"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"blue"</span>,<span class="st">"gray"</span>,<span class="st">"red"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size<span class="op">=</span><span class="fl">15</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="GeoMXAnalysisWorkflow_files/figure-html/unnamed-chunk-47-1.png" width="700"></p>
<div class="sourceCode" id="cb67"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">de_results_DvCinGab</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>DE <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">logFC</span> <span class="op">&gt;</span> <span class="fl">0</span> <span class="op">&amp;</span> <span class="va">adj.P.Val</span> <span class="op">&lt;</span><span class="fl">0.05</span>, <span class="st">"UP"</span>, </span>
<span>                       <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">logFC</span> <span class="op">&lt;</span><span class="fl">0</span> <span class="op">&amp;</span> <span class="va">adj.P.Val</span><span class="op">&lt;</span><span class="fl">0.05</span>, <span class="st">"DOWN"</span>, <span class="st">"NOT DE"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">logFC</span>, <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="va">adj.P.Val</span><span class="op">)</span>, col <span class="op">=</span> <span class="va">DE</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>shape <span class="op">=</span> <span class="fl">1</span>, size <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/ggrepel/man/geom_text_repel.html" class="external-link">geom_text_repel</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">de_genes_toptable_DvCinGab</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>                    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>DE <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">logFC</span> <span class="op">&gt;</span> <span class="fl">0</span> <span class="op">&amp;</span> <span class="va">adj.P.Val</span> <span class="op">&lt;</span><span class="fl">0.05</span>, <span class="st">"UP"</span>, </span>
<span>                       <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">logFC</span> <span class="op">&lt;</span><span class="fl">0</span> <span class="op">&amp;</span> <span class="va">adj.P.Val</span><span class="op">&lt;</span><span class="fl">0.05</span>, <span class="st">"DOWN"</span>, <span class="st">"NOT DE"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>                    <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html" class="external-link">rownames_to_column</a></span><span class="op">(</span><span class="op">)</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>label <span class="op">=</span> <span class="va">rowname</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"Fold change (log2)"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"-log10(p-value)"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"DKD vs Normal in abnormal glomerulus (limma-voom)"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"blue"</span>,<span class="st">"gray"</span>,<span class="st">"red"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size<span class="op">=</span><span class="fl">15</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="GeoMXAnalysisWorkflow_files/figure-html/unnamed-chunk-47-2.png" width="700"></p>
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">updn_cols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu">RColorBrewer</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html" class="external-link">brewer.pal</a></span><span class="op">(</span><span class="fl">6</span>, <span class="st">'Greens'</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>, <span class="fu">RColorBrewer</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html" class="external-link">brewer.pal</a></span><span class="op">(</span><span class="fl">6</span>, <span class="st">'Purples'</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="va">de_genes_toptable_DvCinGab</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"logFC"</span>, <span class="st">"AveExpr"</span>, <span class="st">"P.Value"</span>, <span class="st">"adj.P.Val"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">DT</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/DT/man/datatable.html" class="external-link">datatable</a></span><span class="op">(</span>caption <span class="op">=</span> <span class="st">'Differentially expressed genes. DKD vs Normal in abnormal glomerulus (limma-voom)'</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">DT</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/DT/man/formatCurrency.html" class="external-link">formatStyle</a></span><span class="op">(</span><span class="st">'logFC'</span>,</span>
<span>                valueColumns <span class="op">=</span> <span class="st">'logFC'</span>,</span>
<span>                backgroundColor <span class="op">=</span> <span class="fu">DT</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/DT/man/styleInterval.html" class="external-link">styleInterval</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fu"><a href="https://rdrr.io/r/base/rev.html" class="external-link">rev</a></span><span class="op">(</span><span class="va">updn_cols</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">DT</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/DT/man/formatCurrency.html" class="external-link">formatSignif</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span>, digits <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span></code></pre></div>
<div id="htmlwidget-ae812c65140817f90ea1" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-ae812c65140817f90ea1">{"x":{"filter":"none","vertical":false,"caption":"<caption>Differentially expressed genes. DKD vs Normal in abnormal glomerulus (limma-voom)<\/caption>","data":[["KLK6","RPS2","APLP1","NDNF","DDN","DCN","CDKN1C","CALM2","RPL41","KLK7","PEA15","DSTN","FGF1","AIF1","CHD3","SORT1","PCOLCE2","NDRG2","MSLN","MME","CLIC5","B4GAT1","CD68","SEMA5A","BGN","PMEPA1","MYL9","ARHGEF3","HIPK2","MPP5","TPPP3","SMAD6","DDX3Y","SYNPO","SLC2A3","SPOCK2","RPL18A","PLA2R1","MYO1E","RIPOR1","TJP1","FYN","NPHS1","GBP2","CLIP3","RPLP0","FTH1","JUN","PLS3","RBMY1J","GOLIM4","ZBTB4","ORM2","ARHGEF12","SUPT7L","PALLD","RPS4Y1","NEBL","HMOX1","S100A10","RPS19","RPL29","DAAM2","TUSC3","MYO1D","RPS28","CCNI","SEPTIN11","CRB2","NPR3","TSC22D3","MAL2","DHRS3","TKT","SH3BGRL2","TNS3","TSPY1","DYNLL1","INMT","NPNT","VAMP5","PRKAR2B","NAMPT","SELENOP","SERF2","CSDE1","ARMH4","IFI30","MAP1B","HTRA1","SEMA3G","GPM6B","CLU","PARVA","STARD7","RPL28","RPL18","ITGB1","SLC6A6","WT1","NPHS2","PODXL","RPS20","SGCB","TNS2","MTSS1","FGR","RBM3","BHLHE40","ZDHHC6","BAMBI","FOXC1","TYRO3","RPS12","HPGD","PDE4DIP","SOD2","CNTN1","CERS6","CALD1","RPS9","WFDC1","PRCP","TCN2","WFDC2","TIMP2","APLN","GJA5","TSPAN2","SERINC2","MYLK","RPS8","SOST","S100A9","FLRT3","RPS14","ST3GAL6","VSIG8","RASA4","PRY","RNF145","SLC11A1","KLF4","HEXB","FAM20B","S100A8","ACTN4","ADAMTS1","USP46","CPM","CAVIN2","ITGA8","NET1","RCN2","NES","RPS27","FABP5","ANKRD33B","TCF21","STK17B","KIRREL1","GAGE1","TAF15","ARRDC4","KDM5D","THY1","RPS10","TNS1","EEF1G","NTNG1","MICAL2","ZBTB16","KRT8","ARHGAP19","BEX3","IL1RL1","PARD6G","C1R","TMSB4X","MFSD6","ITGB5","ARRDC2","SLC7A8","PTH1R","DDX17","GATM","MXD1","CD81","TSPAN1","RPL34","DSP","OAZ2","IFIT1","DST","DCBLD2","PKM","SERINC1","SRGAP2B","LPAR6","NKD1","DDIT4","CCN4","OPRPN","CD9","HSFY1","HLA-A","KLF2","ZNF185","SUN2","CTDSPL","HNRNPH3","PTPN13","LMX1B","FAM89A","HEXIM1","LARGE1","GTF2A2","NEDD9","WIPF3","CMYA5","YAP1","LRRC2","STS","BIN2","NCBP2","CLDN4","PDE10A","RPS23","HCK","RPL9","CAPN2","TFRC","ENG","ESRRG","NRIP2","PDPN","POSTN","UBE2J1","PTPN14","ABLIM3","MYH9","SSB","RPL13","LGALS3","NELL1","ATOH8","ALOX5","TMEM252","NINL","SAR1A","AKR1B15","SLC6A8","TBL1X","KAT7","RENBP","PAQR7","TAS1R3","RPS3","RPL27A","RPL12","RPS15A","RPL37A","ALS2CL","SGPP2","ITGB8","RPS5","MORF4L1","CXorf49B","SNAI2","PPFIA4","PHACTR4","ARMCX3","GNAS","EMC10","ANKRD13A","CHTOP","VNN2","PTPRQ","CGNL1","TNFRSF1B","PATL1","SLK","SERPINA1","CNTLN","N4BP2L2","GALNT15","UNC5B","RNF11","AKT3","HKDC1","SLC50A1","AIF1L","LCP1","SCTR","SLC38A2","MIA2","RPS17","MXRA8","HSPA6","SSX2","ANAPC16","CXADR","MGAT5","MIB1","ZNF664","HGD","CHRDL1","PIEZO2","RRAS","TCTN1","SNX25","TRIM22","RPL19","USP9Y","FBXO21","FAM98A","SDCBP","CIB1","GJA4","AFAP1L2","GABARAPL1","CA10","CELF5","RPL17","NRGN","FAM156A","CTSB","PTPRB","ZNF140","RCAN1","TMEM98","KCNG3","RPL7A","TNFAIP3","BANF1","KANK1","TGM2","CRYAB","GPI","LILRB2","ARPP19","OR10G9","PLK2","KRT18","SEPTIN4","ST6GALNAC6","ETS2","HSPA4L","MRFAP1","FAM47E","MXD4","DMGDH","TBCEL","NTN4","CYP1B1","SLC25A6","AKR1C1","DMTF1","PILRA","GBP3","MYC","NFE2L1","TMCC3","OVGP1","MUC20","MPV17L2","CYREN","GAS6","CDC42EP4","THSD7A","SLC16A3","IL6ST","EPHB4","CBLC","RNF220","BEX4","TRMT10A","DENND2D","TMEM176A","PCMTD2","TIA1","RIC3","RPS6","PRX","SLC4A11","C11orf58","LILRB3","SLC40A1","DEDD2","CTSL","KIAA1549","PDGFRB","CALCRL","EPAS1","FAM43A","HLX","CREBL2","PNPLA2","GOLGA8A","FXYD2","SLCO4A1","PPP2CB","SPTAN1","TIPIN","TMEM47","FCER1G","CTNNBIP1","PLD3","CTSH","KIAA0040","TNPO2","F2R","LINGO1","DDX6","UBP1","PAFAH1B1","BACE1","MIDN","FRZB","GARNL3","RPL8","SH3BP2","SLC35F6","PLXNA3","SPRYD3","RACK1","SSX3","CORO1A","FXYD1","LTBP3","AGT","FRY","RUFY3","ALDOB","ZFY","MTRNR2L1","WTIP","SPI1","LRG1","YWHAE","TUFT1","PRDX3","RPS27A","PTPN6","GLUL","CCDC80","AFG3L2","SCOC","PRMT6","ACTR3B","SMAD2","FERMT2","MRC2","LACTB2","ANAPC13","SLC2A13","MYL12B","SLC16A5","CDH11","SEMA6A","SEPTIN2","HNRNPA1","ECM2","MAF","NDN","WDR41","RPL37","PALD1","ITGA6","CTSD","TCIM","FAT1","UACA","ARRB1","SEMA4B","RPS25","RAB32","APBB2","RPS18","GMFB","VCL","SRFBP1","WNT8A","DESI1","IDO1","MYO1B","KLF7","PLXNB2","NFX1","KIDINS220","SPOCK1","LGALS3BP","SLC44A2","S100A11","ARL17B","CCDC198","FCRL4","SRSF1","SORD","SLC35D2","RNF149","CYP4F3","HEATR1","HSPB1","RPL21","HIF1A","PEBP1","HYOU1","POR","PIGR","PCMTD1","TAAR8","CD47","TGFBR3","SKP1","MMP24OS","OAS3","S100A4","PIEZO1","FOSL2","IFI6","CCL14","CT45A1","KHDRBS3","ACVRL1","HSPE1","CALM1","PREX2","PALM2AKAP2","ANKS1B","PAPOLA","NOP53","XAF1","GBP4","ATP5MG","MINDY3","CNPY4","CCND2","RPS21","PKHD1L1","GRN","RAB23","CRTAC1","MILR1","RIPK4","PARD3B","TCEAL2","MOB3A","TTC39A","VEGFA","DUX5","CCDC90B","PGF","STN1","UNC13B","MST1","GIMAP8","RPL30","TUBAL3","SP1","ABTB2","SNX16","ZNHIT6","PRKCA","PTPN21","EFCAB14","PLCE1","PGM1","ADORA1","EXOC3L2","RPL14","CYTIP","DYNLRB1","TALDO1","HSPA12A","TNNT2","THBD","OR6Y1","ALYREF","GSAP","IFITM10","PHF10","PAQR5","REG1A","DEK","DEFA1B","RBM4","FBXO2","SIGLEC14","KIF13A","RARRES2","AK3","AKIRIN1","UPP1","NECTIN2","EPS15","SERINC3","SLC15A1","PRRG4","BBX","ATP10D","LEPR","PABPC1","NEK9","FAM3B","SON","ANKH","RGL1","MAGT1","NOS3","UBE2Q2","FKBP5","NUDT10","ZC3HAV1L","LRP6","COPE","SNURF","AGPAT1","DNAJB1","C1orf198","C12orf42","SPON2","IFNGR2","KLF13","BAHD1","RPS15","PLAUR","KIFC2","SEPHS2","RHEBL1","BMP7","TMEM132B","TMEM106C","GSTP1","NOTCH3","POLR2K","CALCOCO2","R3HDM4","TAF9B","CRYZ","TNFAIP1","ARID1B","CMTR1","SIPA1L1","MGST1","NRP1","AZGP1","EIF1AY","KYNU","WDR72","CNN2","FNBP1L","UGT2B7","TAPBP","YAE1","GNRHR","DDX3X","IDH2","ALDH3B1","MPZL2","RPL15","DYNLL2","PFKP","ADGRF1","NNAT","AGTR1","INPP4B","SFMBT1","MTCH1","AMN","CCDC91","APLNR","TTYH2","NR1H4","CYS1","RPL36","SYNE1","ESM1","ARMCX2","SELENOW","VIM","ARF1","PRRC2C","GPX4","NACA","CASC4","TMEM262","SUN1","PPP1R36","CD63","CXCL14","ZBTB18","UBA7","KIAA1549L","VAMP2","OR5M8","DIDO1","ING5","ARGLU1","CHCHD2","ICE2","XAGE5","BPY2","ZNF106","ADGRB3","HS3ST3A1","ABLIM2","FAM90A1","FOSL1","TMEM107","IQGAP2","GCNT3","RPS7","GOLGA7","EIF4G2","DIAPH1","CFI","DLG5","RAD21","FAXC","ZNF587","ENTPD5","EMP3","RPAP2","ZMYND11","VNN1","CENPI","NR5A2","SPCS3","YIPF6","TRA2A","CLDN5","ZNF559","RBMS1","SYDE2","BID","IGFBP2","MS4A4A","TMX4","FPR1","PLA2G2C","AQP1","CD2AP","TMED7","OSMR","PAM","MAP3K6","TAF6","ADAMTS3","RAET1E","NCKIPSD","ADD3","CAND1","IL1RAP","SYNJ2BP","HYAL3","IP6K3","TLE4","MAD2L2","SERPINA3","PDLIM2","GBE1","SPATA32","PLEKHA2","CECR2","TOP1","PAMR1","FBRS","RPL5","CBX5","RPL36A","BTRC","MFSD2A","DNAJC19","UTRN","SLC44A4","RTL6","QPRT","GPHN","RAPGEF3","TP53I11","GGTLC1","CLIC1","TET2","PRAMEF19","LILRA5","ASAP3","ATP5MC2","KDM3A","REG3G","BCAM","APOL3","IQCJ"],[-1.55511920218114,0.86566267980435,-1.3263316882969,-1.50229686373663,-1.42962633705051,-1.79176484206843,-1.56084540236117,-0.790920920474068,0.682593280708665,-1.29939023850801,-0.904518184867434,-0.818018229876633,-1.10247760255842,-1.17163735977683,-1.12039876901481,-0.937283593351533,-1.27133946735321,-0.97969182036633,-1.17341930674174,-1.22174693579104,-0.920502455756666,-0.956127536625651,1.12538418396543,-1.0051594485488,-1.38620831929371,-0.993666524337533,-0.927802090730701,-0.835440062864626,-0.815219274291932,-0.895388444917446,-1.03433993676786,-0.99810517523801,1.2235267552889,-0.934079075694113,1.4671307426226,-0.80841102077008,0.523608807812535,-0.936318820539494,-0.840640743239258,-0.776490585085691,-0.833384348606364,-0.776357851354473,-0.964836952381184,1.05811840117434,-0.852428892897408,0.664776538009182,1.01096098764827,1.38563508614347,-0.740640751964934,1.01525371708476,-0.765310518357223,-0.70760953063319,0.725983666096694,-0.690896330603045,-1.36913047193674,-0.817152856415762,1.75635242740757,-0.831311548785072,1.50816383369073,0.809215053800911,0.593864876412482,0.533183146891322,-0.84523221158268,-0.75021004432646,-0.812142505856886,0.512367386480881,-0.580113583069275,-0.693051553712851,-0.885906627927236,-1.07650967214025,-1.10336130287503,-0.972703873420676,-0.78629075228362,0.916557308678636,-0.754604258426108,-0.797069117023193,1.07805320240355,-0.571014970873048,-1.43574765839482,-0.84893091695356,-0.704080187624344,-0.94599375728693,1.02489372791378,-0.731304455194239,-0.573435504034095,-0.537060743693964,-0.781822903115045,1.08164186851546,-0.81198539917979,-0.861622323375003,-0.718684565587697,-0.913454384761899,1.19559258463212,-0.612112735558043,-0.642032466730718,0.474817431002364,0.488581495974461,-0.491570216632838,0.789213372124298,-0.927483160178011,-0.875092222280026,-0.762581812069428,0.452420153811088,-0.676050924853762,-0.606829159489387,-0.75144377378313,0.937630662031814,0.970564599840566,1.38434989321104,-0.752795521620583,-0.779524746308024,-0.722595008335642,-0.743262484250948,0.514722349635679,-1.14570532268268,-0.725282194409865,1.09752421308784,-0.81361061610219,-0.720434683831106,-0.73414974530982,0.45432150218459,-0.795700381368307,0.72714628021365,0.956367525288208,1.28940708486018,-0.715667957036963,-0.839232720768919,-0.825922585339688,-0.91625387949127,0.725121553982145,-0.729334751114803,0.460087342150394,-0.995641906297788,1.73971818063943,-0.879473583814976,0.392642325779908,-0.887640297335455,-0.721243601522488,-0.749145810037281,0.891126745174379,-0.591894766174208,0.825755815835888,0.939694306576097,0.691211904438552,-0.723023977957723,1.57177873758914,-0.508290711512119,1.17440623838305,-0.639268826153149,-0.897350342671766,-0.968174802733039,-0.87353232075785,0.781678541232396,-0.624044994806127,-0.63553636166513,0.379806914999618,-0.628623139322719,-0.747114395044848,-0.896891200186775,-0.750748214071972,-0.660694303850478,-0.651475151476682,-0.601171914868237,-0.721665307200212,0.866962203879117,-1.10638672147194,0.436116046643296,-0.640088237377705,0.41779584620588,-0.761649358257701,-0.817343050050232,-0.821178689439977,0.935370896335111,-0.732836483523806,-0.609950187167589,-1.24857592678082,-0.740150371075988,0.980214800807062,-0.564682750128417,-0.669858000638175,-0.652712135060622,-0.737726386135776,-0.872509089306603,-0.821630432256732,-0.403797467849415,-1.08424438057178,0.791545592674177,-0.564428465081557,0.972181168998119,0.43272242206237,0.667949917573956,-0.567048229577033,-0.735292784972796,-0.627860450917664,-0.687202344353698,0.541710638817345,-0.39859529770307,-0.556947676063429,0.70973416266854,-0.72729838527078,1.119002771641,-0.779570826770862,0.768880297981939,0.702779515657496,0.843280424788079,0.702115793605089,0.878503899014232,-0.699261185291694,-0.508490085261158,-0.603158153710183,-0.446658491986017,-0.687021961620948,-0.714782432209287,-0.658979759450313,-0.644322035548172,-0.651333294293204,-0.63399817847564,-0.723605958420659,-0.725229007388319,-0.73833633735216,-0.531187794248702,-0.704054082941311,-0.857226788263334,0.720333533327491,-0.561630391599339,0.983709372240682,0.676697676671014,0.432971282907883,0.655508595422287,0.363721583526356,-0.472611007037183,0.709576591154923,-0.76218359541822,0.676865199139364,-0.639603764092948,-0.704763636892361,0.850254551471144,0.606611885588547,-0.587609482349348,0.87815175128833,-0.4234468402533,-0.587868439411445,0.362206653553054,0.78554712512198,0.873301900925567,-0.70483871795386,0.785341161495634,0.966168594701285,0.60193220725709,-0.547718696101028,1.12905444583095,0.667084538126405,-0.65258259121947,-0.559701999282839,0.686290131439721,-0.691436717098509,0.553974380223696,0.388019482431395,0.367557207319196,0.385213117685816,0.381411125253968,0.34463457251342,-0.691947355576458,0.643282608008834,-0.668659185179026,0.435698058345476,-0.375391945383147,-0.714660262750808,-0.730997538534309,-0.605229946466951,-0.537550611246455,-0.55735635860079,-0.480488309938321,-0.482516206799788,0.606355826317419,-0.504153449358033,0.730819043792021,-0.786621020459187,-0.6361594437782,0.665945558176497,0.553667993017342,-0.565658042611537,1.68307500538597,-0.667022518118455,-0.508113422645631,0.915649999796491,-0.631636611672203,-0.49503671884132,-0.536585491611252,0.848615313725819,0.541717303996825,-0.500647165128135,0.827741286879313,-0.653594415996306,-0.609071107310807,0.727674049949131,0.406333112583902,-0.6922964683686,1.63301079224518,-0.648130334517948,-0.51445839928922,-0.53576026364712,-0.573068828880111,-0.50455058100848,-0.546243363120427,0.613911783664594,-0.620116823196966,-0.64062373476402,-0.500507889520988,-0.618058473800932,-0.649476481876466,-0.577185763949504,0.322727286338274,0.648017880191845,-0.541599564896952,-0.539812080252058,0.44772755027954,0.622088765358519,-0.581976289706761,-0.582337690686263,0.440410495068678,-0.66962010455028,-0.700414135835929,0.350824432374677,-0.696892221450566,-0.589083862019914,0.863891602118617,-0.775722510656518,-0.438309150489695,-0.719045791932559,-0.578741793882235,0.695874560775421,0.293755157448311,0.702134585561661,-0.42925095642686,-0.520596699651738,0.776087009656434,1.0000442754456,0.530026042586511,0.592727229771439,-0.405904009167661,0.595163361379104,0.651253892560788,0.676876093167033,-0.584142938379934,-0.457756301717905,0.679516224532786,0.731247148798252,-0.40617245560276,-0.694457922056461,-0.526150953388122,0.632155867851322,-0.515267942509843,-0.702951274350929,-0.71611355547958,0.47058347832302,1.3642828406231,-0.498045711684727,0.658249992313032,0.612577518879587,0.724275489405285,-0.426801561503722,-0.575644409664847,0.613763939112768,0.624889296023424,0.622900259564903,-0.556237781547607,-0.561646249486021,-0.491791903406093,-0.611247309831565,0.615063900981316,-0.422678437217294,0.559313028119758,0.526521634198776,-0.568199775922403,-0.478532580896135,0.600668602013121,0.626061372595188,0.679295291023639,-0.493225444909801,-0.53553537633354,-0.605373818482102,0.386184841659999,-0.599784843555819,-0.680005787915693,-0.392677124781406,0.528927749078003,-0.545251271331471,0.58753900232885,0.86409854369486,-0.590227155007785,-0.612446885553085,0.649153964337228,-0.769874792188056,-0.602515791124604,0.63076538331299,-0.510739378980043,0.445947507430744,-0.494412718398432,-0.887934083125839,0.794439239145849,-0.4045022798038,-0.3786009483559,0.522297550859598,-0.597196293820172,0.752765430388199,-0.508184170443785,-0.435190765475137,0.945526100987207,-0.539886216024843,0.558463423885497,-0.592466312815001,-0.560909283831421,-0.47135588003868,-0.489390405568688,-0.441099332929933,-0.559640076957804,0.620862394326605,-0.749345955688831,0.538733661169893,0.3442783215767,0.652101368128317,0.722945392805058,-0.600095166484323,-0.514920818604572,0.335461672864508,-0.48320700138628,0.639619408729139,-0.65491009061415,-0.481348439927668,0.970764482221782,-0.506472046450344,0.49973567518484,-1.10455968284409,0.583206959506718,0.620574442305595,-0.604435029358639,0.66073371399295,0.92483287990998,-0.293925575231649,0.552176034547461,-0.449713177468132,0.315933345350497,0.566885002331648,-0.429877744112537,-0.627779678290443,0.489687432086257,-0.453597841080665,-0.55277717362699,0.540740762844676,-0.400234353929346,-0.426704589109226,-0.683936279041197,0.677993891902754,-0.430752825823529,-0.57708187178838,-0.359733123207972,0.638728092356595,-0.573925011389562,0.607261252147679,-0.337623181234836,0.296930416326444,-0.562519611925145,-0.696581787146592,-0.574084672247588,-0.460766403249724,0.41188763264354,-0.588400910273451,0.482335541917068,0.447599682884392,1.07681530472814,-0.549748111476053,-0.509828299855045,-0.449872593259356,0.716197420222541,0.313440582595832,0.548694153006346,-0.512413434406939,0.410916553612171,-0.488843258942732,-0.510407653014166,-0.585225530211199,0.510231207158656,0.552054553194715,-0.601234978412913,-0.568611829559019,-0.502910539483548,-0.397974670452058,0.639457926241233,-0.48112906463331,-0.472220945991827,-0.427760273019834,-0.444080815775958,0.51199830079973,-0.58112403851773,-0.573969024602692,0.568753483171363,-0.344637199544017,0.621487027485456,0.523264093504862,0.487674031374663,0.755899509348634,-0.522341623576697,0.781363089756194,0.251458945213225,0.561790420092771,-0.441415097233966,-0.488609817753243,0.513797267319032,0.861563700334084,-0.441786650347514,0.577910055002583,-0.452983271157304,-0.549962411322632,-0.288880815660626,0.463654338836238,-0.598602483841622,-0.621654287085557,-0.448958042609917,0.505459423272969,-0.488190475562052,0.779948944849144,-0.509027207336405,0.54124602197682,-0.486420887016472,0.639574138506029,-0.389267862626887,-0.500100186410284,0.534342075429892,-0.619366626456511,-0.406750717312506,0.375161867563932,-0.548076782566783,-0.578707374937594,-0.593189279756844,-0.3705634833136,-0.529855520135413,-0.55553319321183,0.303918008862816,-0.585112088214461,0.433389845023341,-0.541420911791223,-0.540699144401124,0.702344193696502,0.537775714869914,-0.421545730812559,-0.606185856699326,-0.497227906046742,0.647494029240365,-0.791845222334478,0.452391090465418,-0.435637397679967,-0.559409174222656,-0.504841353682116,-0.523627497126434,0.591757996893691,-0.521577120571607,0.341245808735505,0.534512838993303,-0.493026627602079,0.621862975404019,-0.476270035326005,-0.45296285294971,-0.455951103281088,-0.482829382183436,-0.472281537382214,-0.440040875110308,0.480486064583477,-0.536953307184433,-0.545290142824905,0.311593024778924,0.536445441196702,-0.406849632580948,0.45541206227294,-0.531975693337568,-0.46389317964305,0.517007884248323,0.5265503695174,0.45238407951331,0.521088099532706,0.584388224650347,-0.498140579196811,0.538928957127298,1.08181583323891,-0.406652158961645,1.06698839618875,-0.439621377865397,-0.570997742622994,0.578888077540725,-0.433172745012614,-0.577568452958745,-0.444850176034904,-0.467245015866646,0.933838153444707,-0.381729564940126,-0.449467901330925,-0.394653747858195,0.725137178212242,0.563324997147849,-0.413361823535894,-0.47376419406174,-0.517105078843137,0.313676053980013,-0.462802741255766,0.538185913754943,-0.36517905109343,-0.512475750770766,-0.494641199366732,-0.428837145655782,0.568514954958552,-0.463187413944975,-0.547117051684999,-0.609943686224113,-0.588813527555853,-0.482094492560945,0.475248782604924,-0.443927839717101,-0.432804269199368,1.14339753998566,-0.468563161403972,0.544260791065138,-0.740545105433956,0.401827560459374,0.473270276622114,0.528111377070394,0.269630967972257,0.548322235797981,0.482614068837377,0.631956191094858,0.51484254240153,-0.547521561092342,0.546939830789391,-0.612750033605073,-0.426870762044109,-0.576801995683241,-0.471358864923329,-0.387658445176704,0.385472404053762,-0.496078677648508,0.522385059709705,0.390613782281433,-0.442776207072328,-0.458262789547245,0.503216235635112,0.669633287831411,-0.517095156951383,0.67026633347642,0.606495025386389,0.65014890445155,0.542953331870334,-0.447996162428112,-0.398503844965429,0.875786507732226,0.422532836630772,-0.481030855112172,0.547086742838871,-0.397351074726977,0.587941953893958,-0.481984449518681,0.583338420915641,0.238229271478479,-0.386755264749434,0.625004812669847,0.503243101229162,0.518495957302649,-0.571583880534176,-0.510783283271651,0.516118128339202,0.357648225789542,-0.586401347889066,-0.476776122023955,-0.557737173150644,0.535074992365963,0.614734510248814,-0.466601304452355,0.30153773538493,-0.436167299610085,-0.755667781990751,-0.482601708321423,-0.3417046947539,-0.5969671527042,-0.262200304317159,-0.412431328607654,0.361602284953059,0.299830480401046,-0.356722170533308,0.562694047870419,0.459083565714007,0.579887370698815,0.288049858304307,-0.91546063994277,0.478528405968314,-0.471529441938321,0.540800110035006,-0.42114275383978,0.515859992241998,-0.454786265060646,-0.49202482795376,-0.355971011941648,-0.367452285741652,-0.47853973780851,-0.519177244160993,0.517004315460914,-0.460213789292226,0.536003146219215,-0.507371220768386,-0.478802378829863,0.505923952427774,0.482766535293038,-0.47650693965255,-0.412947945299623,0.625756493370682,0.294612040645229,-0.38229331417527,-0.302814159715613,0.5014814708408,0.508625880787736,-0.454786746285771,-0.364136488008688,0.528888305680981,0.295256529955417,0.496253360015415,0.482063677641303,0.520962967376111,-0.409972886303053,0.67759326204501,-0.513602021433838,0.515664952235841,-0.367902610810613,-0.438918430247767,-0.371044662916922,-0.456975964220678,-0.472586013668934,-0.387390589742055,-0.531183828071629,0.475784762311314,-0.563307417416948,0.612404000575625,-0.441851306991526,0.545617088545026,0.591072706637956,-0.785966802712802,-0.424114894367298,-0.426552546111612,0.461344498739787,-0.408677215109407,-0.453341075592868,-0.459356357008623,0.489797879629759,-0.499950825661044,0.472726318169722,-0.377485947409905,-0.40776764884568,-0.518754518342512,-0.425571197836403,0.484825879300789,-0.507622552981625,-0.418383124547756,0.477416846111245,1.2635824286643,-0.478745437005848,0.46780985349805,0.494515533283009,-0.410066248732813,0.531821550088411,0.463974518573707,0.543196707390447,-0.434180070872453,0.265044332586005,-0.39026037589719,0.317383239659272,-0.429954880386497,0.529189393763696,-0.519923628298722,-0.411225622503446,0.518561300980393,-0.465147829456223,-0.513007082083364,-0.397203541702823,-0.489179887966584,0.478128723656331,0.563850241044767,0.383720717329172,-0.448623451171113,-0.554901061612799,0.505090247049526,0.505105714209158,-1.01308326688361,0.43579647390972,0.459935810280632,-0.46263841955797,-0.478182783406123,-0.475736687617049],[6.02509916812993,8.78066217165782,5.89496400099668,7.47795618266897,6.14177637610806,7.8177980382032,8.027719331052,8.17895274012871,9.38081029879967,6.21996637269689,6.92400841806551,8.01015196362424,7.69899926557679,6.55999107321515,6.17743332326351,6.90545266164239,6.14621799921949,6.7959343878457,5.45348244294603,8.29361083578622,7.95347662697357,6.1859094883583,6.25103084913948,6.72518698343375,8.17599031894043,5.80539683293974,8.12107161846274,6.72926506507896,7.05156398667299,6.67761250722403,6.58873832230075,5.8161920697852,5.42331783638675,7.12779555313787,6.91357216950285,7.57420012642524,8.50813071668189,7.58818051664162,6.56434148705724,7.19954729937757,6.76936756970463,6.51735388216539,6.93029681028646,6.18383885192914,6.31064673166519,7.89396009409575,8.2361220097081,6.51114892995711,6.71650414929618,5.14260242830257,6.87642297256281,7.00537206089558,6.2299781042972,7.21285107172766,9.51541988533804,6.56478431637801,5.96374383157882,6.74602561093858,6.15281635737147,7.59008943547654,8.44194990103289,8.51926515283345,6.045140729644,6.51554946061301,6.17653520424072,9.46856117844827,7.69154209176057,6.82039797482369,6.09131428989692,6.19454788734864,6.41612953592764,7.21847467670128,6.9623017291345,6.29606148229852,6.27184039888324,6.68671390975919,5.11554843283977,7.50630278936941,6.62386458357346,8.35116310132761,6.85662247250085,5.88725164203179,7.23111548601811,7.39970104045516,7.59523453220025,7.93591275447657,7.2119894932857,6.58555148158287,5.93562576902985,7.99791906600753,6.41169693525354,5.45532861948763,7.38906899986457,7.18019229930002,6.95402546532776,8.8337220497479,8.39945011600592,8.30217882324037,6.24545447002372,6.30929970486261,7.70257067100564,8.61035866758892,9.26753283007974,6.82983985341247,7.46060740031179,6.12482943624905,5.26304266157089,6.74699385949477,6.04023503724003,7.69488168004658,6.1219764976631,6.34365568500232,7.10933742250713,9.59091860009747,6.480738021295,6.46278370696777,9.02116326193567,5.53056722972677,5.82121812158122,6.85391390084832,8.59690370629342,5.67606967550313,6.3744858195274,6.10617354464919,7.37607597107614,7.40486076100927,5.32611388024977,6.05146022831107,5.7389107950058,5.84160771836592,6.7302612748809,9.0632498102584,6.39096656070859,7.1363689542852,6.36781107507398,9.17362109707208,5.9935165030689,5.80001025650769,6.53643426258754,5.0457248798625,6.4798620363404,5.48555129731092,5.94295065176478,6.36626474751579,5.78437567946213,6.4063836627593,8.62126884207435,6.86495685832841,6.3775312528364,6.16998214267641,5.81951670649344,5.6580710958247,5.73284209013165,6.37242681613977,6.58515400106792,9.53015845314462,6.1331966976404,5.81619298473171,6.55758973032696,6.05439955486967,6.41035821189457,6.92358639834786,6.46307324435819,6.29611078537365,5.16339426190489,5.38040517057533,7.95314560677202,6.84729952453009,8.27600388915369,6.42470646263097,5.10543837614532,6.5755022307394,7.35220503112681,5.74109489584886,6.85676950513746,7.51638244671186,5.80222907173263,7.22561309306055,10.377405121493,5.77632204805892,6.3384939820897,5.50895941567008,6.33892736987249,7.2339262946497,8.81557012091221,6.39887819591883,5.9013888296859,8.95958842162864,6.14938545820759,9.5589591186987,5.72885201996014,6.33421104540522,5.64659222222873,6.50600300524936,5.98100959119535,7.31223572890298,7.91988309833876,8.09839792711609,5.58868367268417,5.47977154523932,6.52522290768874,5.28452710908754,4.86662899566515,8.16215217889799,4.58320191793997,8.73157594946462,7.51658172407231,5.74132008436593,7.12921997360923,6.4715109382554,7.21361649745288,6.37250552286212,5.66786329024571,5.7789659475892,5.98030123832617,5.67763537814576,5.90701519250336,6.17163508726539,6.30318142575962,5.10854580170209,6.72286728805529,5.75403665457837,5.01945517228505,5.12385703835396,6.18345017614918,5.93434568756421,5.48726595887546,8.9729816606124,5.80264315047115,9.04688371444604,7.89365328431327,5.8820832024961,7.25180323244767,5.33897009958588,5.77447045684046,5.33650528898801,7.05366188311562,6.40460759219621,6.18505672006364,5.36833606455295,9.15750452469622,6.08646968404691,8.86532358228998,6.12502544155458,5.03766029698539,5.47000251286928,5.62101625961573,5.61050196461237,5.43037884522087,6.33303064736973,5.66222840983325,6.06107919616854,5.48648317166594,6.71285510958526,5.49144682727326,5.95106819795108,6.08488905920164,8.54830327648139,8.6492099976544,8.72852584951688,7.95094183620593,9.56001366929827,7.61699443519753,5.62316221321143,7.4272813491793,7.5767091923665,7.95541720795723,4.85501062537257,5.44750612958642,5.80144899254305,6.33713182225071,6.50026999980736,8.00897505840679,7.06879663827822,5.99587995267469,6.21572759260493,5.04991949104754,6.29571600220048,6.73406988067365,5.55584058411463,5.92769209828684,6.27475428398057,7.6210850026476,5.35787583329756,6.58051057183766,5.24248640207966,5.66418732574133,6.88884785055446,6.27541883683874,5.26356476413824,6.01613735087432,7.2932026933188,5.57917559393699,7.05375150380948,6.1038619506399,5.40071744159431,9.38150411968185,6.28053186926519,5.78080719256634,5.18605167226589,6.45102168297056,6.68370439641431,6.47142420608001,6.62675785266019,6.08378309951975,5.904470911251,5.83592036056704,5.29884806430258,6.06912038774349,5.49072752346806,5.60075085780751,6.0198389507827,9.33162387963846,4.95428921119533,6.51973547063715,6.03602662322877,7.0521717613645,6.19699484178295,5.88338179255923,5.99091825426943,7.30024734610384,5.44627598464015,4.76350965229227,8.94774725561903,5.7977239167335,5.61585617276029,7.86611792223911,6.51297245891881,6.99521205349554,7.03023972309507,5.97777462937537,4.82433004936913,9.54680405482011,6.10308720904755,6.82823955411343,6.4042953716642,6.61660233739805,8.98470394552401,6.99547866089455,5.71239858555746,7.32166355501412,5.19083253684048,5.44844975455787,7.66132052630164,5.56391736561304,6.7153018017411,6.86835184108339,5.17912814650188,7.34353804657674,4.85341575558255,6.34877484329601,5.25132750441321,5.89024657400879,7.07084132988591,6.4013452980648,7.31347665644443,4.65399276257313,5.96244079849423,5.26274901968184,5.61098167279109,5.82547261446615,7.41863361770319,5.62043314130913,5.12114276967372,6.32523222491011,5.27755289980177,5.66749288819665,6.96575783016206,6.68836824795171,6.43120952357862,6.21287372451093,7.97185076882614,5.86222279681798,5.71834441746139,5.77606246353485,6.7351000563592,5.19668161186104,5.27651245902219,5.87968278744274,6.31922413568031,5.96744628281521,5.28968224961263,9.56833188516981,5.5751830248396,5.08500678112042,7.31554523456505,6.02526308239474,6.14180928458453,5.62589985704882,6.79312673882761,5.39408111302411,6.76842515381958,5.60333379959781,8.06273153281908,5.53844171146145,5.38928856415468,6.01271361201633,7.20516210207242,6.16815360727067,8.6484354624354,5.59787122220448,7.01959778003401,7.45575349862782,5.73099055334252,5.66682109290275,5.50808482052609,6.0699497150153,7.12231938164579,6.13678589268888,5.76057652366555,5.32762093175838,5.48731914793817,5.43221920540562,6.44797886416809,5.73764806815621,6.69357718403566,5.95784003709269,6.36750932212194,5.29946276605185,5.38206561919113,7.86455704545295,6.03060775110913,5.6670696695945,5.08690321800922,5.48802597962435,8.34923718682439,6.17544930544559,5.31650159341641,5.14181525655409,6.35805224785709,5.13540316119474,6.25750172317684,6.16122458120087,7.95365332029976,4.85371211065472,5.8529808217585,5.30074479269354,5.60193269178448,5.65993210289595,8.34724605328916,5.22524147859765,6.46349713944058,8.68802111499562,5.55285413526889,7.59049074834177,5.22464152985597,6.17938173057272,6.38918255334108,5.59356985050206,5.89809087922182,6.76940856186348,7.09284365699171,5.27879517596342,5.50467703520578,6.33272513378756,5.55354860468177,8.2972036097158,5.67051989409953,5.4758447464297,6.21689710921298,7.55312952492764,8.88706834245832,5.46003105266099,6.79024084341957,5.76905263923573,6.09408187521911,9.43742752191227,5.17194899683661,6.12002117442291,9.63396292620468,7.55422079162575,6.56702380180434,6.17150084993949,6.29100246359846,6.43963824974074,8.71916436149204,5.4006977979804,6.35976165743094,9.15174382086034,5.79298387378419,5.92424793574108,5.34485622641064,5.94101664181747,5.60861787767282,5.27469546840651,6.89905586146861,6.23392980040355,7.05829397481381,4.85611324623617,5.76887937863874,7.11400704508018,7.00698524035091,7.40193155123461,8.03865686125911,5.4507830888888,6.08554698483875,5.1011024732324,7.46757856901658,5.69300447425449,5.5767506465333,6.17642661091697,5.39837254283355,5.23933717321783,7.32237752518186,9.48592686794001,6.83878283436133,7.46034611890995,6.05934224607855,6.24006922367761,6.66272695544929,6.60245956097892,4.97728155793277,6.3457136523508,7.11559284226129,8.84745790676538,6.4627479943798,5.11869708278116,6.79466366082186,6.23860933686306,7.62649224576495,6.19972469576876,5.73777525246569,5.66862437041142,5.75199824422288,6.0443841326457,7.48065857232728,6.71370865310928,6.06696462706742,5.79538526509204,5.21521491445964,6.47591546431728,7.38181540914276,5.87836299854575,5.4031679177059,7.69230922731896,6.67618727486135,5.29231570088966,5.57781632305097,8.45146121740997,5.53032835128038,6.95049905050447,4.98375718888097,5.28195173888658,4.30725879863269,5.32703993176565,6.51147430055769,5.40046681926426,5.59188013225093,5.22703837370299,9.01341890048603,6.12169009665472,6.11118360013229,5.4133176515414,5.9205873770253,5.40377546691239,5.99175769080586,5.91927461023148,8.99436609803608,5.11071843006558,6.04173092825425,5.35054745785226,5.46473163475308,5.75264523017116,6.05739272154779,5.86947509067182,5.81959495102831,6.95394316241506,5.66960804070575,5.52983241997035,5.40288171625569,8.33948980280815,5.40054478051468,6.71000753226102,6.33704271680655,5.70240411793381,6.62121110461788,5.65880364099265,5.01530405936492,5.83079153191535,5.44064944155103,5.58453684898884,5.75934904993074,5.86816735904351,5.51187197917081,6.52514569880444,5.95346742555991,6.22072765467299,4.97660459419615,4.94418676912743,6.29747212814461,6.44443483844029,6.30351411183541,6.17825218118797,6.25728838922946,7.16419350863643,6.1801553286483,6.64310668927649,5.32015342791391,5.18321347151825,6.03735672389023,5.89851474055773,5.48633644168603,7.99838566640385,5.71163761659342,5.08346963354288,6.66960306576564,5.49926448395003,6.10448957191224,6.34384707230944,5.39275041356654,5.73241593384144,6.99950901509011,4.71160189400846,5.18170488822983,6.08972810253759,5.98536299152677,6.4263098461384,6.10133416279358,6.44831200742233,5.70124293205993,4.5858977855163,5.70190705129731,6.42315215817007,5.65831193624553,5.42532842632377,8.37031616901947,5.00369781840572,5.57411837209368,5.9916334867318,5.12005040231511,5.50084264222356,4.74570344977811,10.025806433777,7.47941554769273,6.32134266758503,5.94114626099736,6.47286899554719,6.86561617497688,5.42574640518853,5.76159907967158,6.4863813468593,6.16256272897468,5.65506734674694,5.51223339691666,5.48812438871489,6.85576939941298,5.37220187490879,5.03446597863054,4.73936131105525,5.49780683133054,6.56141006647518,6.87571096934769,5.46063390781101,6.93567187612394,5.52752101648049,4.95618086937544,6.86129871724051,6.21363883372566,5.75151507680751,5.65457667647123,9.29038202076124,6.7496842495321,6.15495851665373,5.32001291061944,5.20448713959331,6.10484619309323,5.40256694580468,5.21729008506733,7.1518460589149,5.02752906279739,5.81004044879751,4.99777523146321,5.17254078543029,5.18515173151809,6.21515787679833,7.85069248467914,6.2125968586341,5.98037995786395,5.38440049600555,7.21511279984115,7.80760874009535,8.33217094260052,6.34839787607369,7.33730175013737,7.99584865815345,6.74935624444877,4.88068976437291,5.95396445878225,4.73499463757694,8.76250590425375,6.29221982965966,5.51746383230909,5.94656793116782,4.8250916307104,6.26227058043854,5.17428033580913,5.82711881675042,5.41279200753956,7.31871240402882,7.09880135352302,5.64280752565931,4.99352125698865,5.06390170079012,5.92429717307511,5.07191761856573,5.33021717213309,5.71454246754954,6.43202718440233,5.72970150314859,5.51308302582467,6.84151692243661,5.0170426226696,8.30084757123065,6.54616489078954,8.06936256324245,5.22742257120332,5.98295822780111,5.8023886280221,6.47488273197831,4.82242840992481,7.8967508992099,5.72743891585336,6.44435187461587,4.99101412190373,6.29403056015139,5.40001233100273,4.93784557188202,4.9116434852781,6.74453047764292,5.88455257544475,6.69725569692826,6.48085020960316,5.45719979980414,6.36178614549708,5.17568944981138,5.3416046734057,7.34599357882002,4.87012529928581,5.96636119809412,5.46306082263069,4.52800502286447,6.87220353608308,5.96526181856169,5.73419426390788,6.11235099700989,6.56246292396951,5.82113712702062,5.59034448019991,4.95117563600418,5.46928681683344,5.59124431262541,6.55691925013883,5.98672659459466,5.20401564143986,5.66687437252265,5.6507338662895,5.20376670511549,5.93708290691277,5.35906862551213,6.1527139410318,6.18202780001214,6.20992510779122,5.08588233244326,6.33999956047354,5.37550091097054,5.32001379225526,5.34441892851878,6.28503986595568,8.63982766363762,6.61865192315813,8.05456215453963,5.46065661603231,5.2970781748515,5.44482789519478,6.30663420044793,5.35470209347648,5.35779151909168,5.53641435915946,6.38887484179778,5.82168439216643,5.94360836915135,6.04973973031993,7.50281037329256,5.84694979152839,4.47667483288172,5.4689983573988,5.16006420383876,9.64293080769905,5.80817475342063,5.22251840745784,9.02060660100884,5.75910810387467,5.41218910930814],[3.17123685397488e-22,6.40480653314719e-21,4.90428051008563e-18,4.9658762266613e-17,1.07778755314229e-16,1.29270930861792e-16,4.72187550793548e-16,3.00465037985799e-15,4.71687188002337e-15,6.45715566123585e-15,6.81793707828626e-15,2.60282384819987e-14,3.10288082393197e-14,6.0742369129893e-14,1.13611822838919e-13,2.97243448756591e-13,5.19896949027911e-13,5.78017110842748e-13,2.12469882429752e-12,2.1965148528035e-12,5.29397436956393e-12,5.8595699845219e-12,1.24288977382902e-11,1.60695326110462e-11,1.65008208954339e-11,1.70005961444839e-11,2.31291925101832e-11,3.0017098606782e-11,4.15002405637642e-11,4.75055388318923e-11,5.50393383149668e-11,5.80753861084274e-11,9.52504135249956e-11,1.33822369213912e-10,2.4124035171357e-10,2.78481620722556e-10,3.15497798467089e-10,3.3936745968765e-10,3.53669345084456e-10,4.08371606917091e-10,4.13954075400111e-10,4.73013940554468e-10,7.82913986580493e-10,8.61494149619849e-10,8.95025518096749e-10,9.02173329638553e-10,1.08739749923549e-09,1.26507411941659e-09,1.43928978394267e-09,1.86192370424349e-09,1.95389837415267e-09,2.09133635364347e-09,2.24386839588746e-09,2.25301574614264e-09,2.73312732605843e-09,3.06111165549707e-09,3.48873069745372e-09,3.89399784054121e-09,4.07233495141805e-09,4.87642540293488e-09,6.26223503587105e-09,6.48194515907748e-09,7.28953670702812e-09,7.79019956702949e-09,7.89758648735153e-09,8.10985859014961e-09,8.53958212127622e-09,8.70042224590585e-09,9.01285890768895e-09,9.11026876116409e-09,9.61409038319826e-09,1.15644830463645e-08,1.18396186226124e-08,1.23424668771648e-08,1.42459152226137e-08,1.42512304036968e-08,1.47217287663414e-08,1.56111352146796e-08,1.85583206309642e-08,1.88432344381061e-08,2.1266291132471e-08,2.41205714479728e-08,2.57518881503046e-08,2.60673135272167e-08,3.22248769389748e-08,3.25635226697486e-08,3.29195525765419e-08,3.42105989792903e-08,3.80257515469728e-08,4.77015277633178e-08,4.97431246435837e-08,5.19760937446026e-08,5.50016399627298e-08,5.64327870564852e-08,6.00812247256992e-08,6.44106822118465e-08,6.50651785368365e-08,6.82673170566243e-08,7.31567220598684e-08,8.55113013862021e-08,9.07236138218047e-08,1.01690838152009e-07,1.02326757564645e-07,1.15859865934059e-07,1.18744561045168e-07,1.35852779707011e-07,1.37280022073971e-07,1.61948480803619e-07,1.70187173872017e-07,1.74059159243383e-07,1.78070329698059e-07,1.81789018299707e-07,2.15307685430599e-07,2.22272108228247e-07,2.44222638748157e-07,2.47967258760115e-07,2.60258462070148e-07,2.69118537624585e-07,2.8868070234704e-07,3.15282778507188e-07,3.26139003184527e-07,3.3736517946881e-07,3.5179422209611e-07,3.57206826012258e-07,3.62298679683394e-07,3.83336898708114e-07,4.09848265939873e-07,4.12393991886232e-07,4.16436570300598e-07,4.16701498133682e-07,4.25506909955642e-07,4.52587910404876e-07,4.66688152306986e-07,4.93294906819627e-07,5.05093428987648e-07,5.12188352911578e-07,5.18275946253591e-07,5.2397415395993e-07,5.37862730681178e-07,5.44899216527301e-07,5.52513731321711e-07,5.89745767347629e-07,6.29623693398707e-07,6.37933747966022e-07,7.2283981770452e-07,7.30148119944374e-07,7.36974637409392e-07,7.98240955264588e-07,1.03851686851206e-06,1.11371558887536e-06,1.15445709146657e-06,1.1588057007559e-06,1.26211357872952e-06,1.29082793879789e-06,1.33744495513208e-06,1.47218533422187e-06,1.60906510476335e-06,1.67625100218291e-06,1.73976318190826e-06,1.82124993894789e-06,1.98555452372401e-06,1.98795962543871e-06,2.08596301407788e-06,2.1838035237426e-06,2.19610875679179e-06,2.25930463753799e-06,2.32493522913291e-06,2.3358111372693e-06,2.42157834771206e-06,2.73582522879374e-06,2.77745287200559e-06,2.80021029192625e-06,2.85886600689495e-06,3.20678097432723e-06,3.48555668387688e-06,3.61263399922059e-06,3.65446739729725e-06,3.92818428245301e-06,4.10046651019557e-06,4.23037386174099e-06,4.2328380461186e-06,4.40686777053016e-06,4.42481041619419e-06,4.87027127171622e-06,5.03890257301359e-06,5.18115373602796e-06,5.23979499302931e-06,5.3038660025433e-06,5.38904190195641e-06,5.447101612101e-06,5.548841453557e-06,5.65231427570106e-06,5.68458428707458e-06,5.72128260997754e-06,6.0566058459317e-06,6.29455679107212e-06,6.34342855399001e-06,6.62505940033025e-06,6.98238134679542e-06,7.04607587737781e-06,7.07299503213034e-06,7.74248489423065e-06,7.76336125159619e-06,7.94838387655982e-06,8.31045215701429e-06,8.4942033172466e-06,8.8831947813806e-06,8.92252930491379e-06,8.96408866565264e-06,9.02662565770515e-06,9.82180526005411e-06,1.02717125266638e-05,1.03787186794152e-05,1.09437630441233e-05,1.16414046909376e-05,1.20005804949091e-05,1.28146491163629e-05,1.29918288056761e-05,1.30303250123079e-05,1.32692205440077e-05,1.32824510474637e-05,1.33434647365932e-05,1.34098738058056e-05,1.34404201956249e-05,1.34573450745838e-05,1.36120621739853e-05,1.43500618829632e-05,1.44058154709399e-05,1.47335168842582e-05,1.51849950989653e-05,1.59950444619345e-05,1.62517937018565e-05,1.65108725428989e-05,1.66439912496249e-05,1.6854545329786e-05,1.75715724307782e-05,1.7959959024738e-05,1.80587172570158e-05,1.82137800621862e-05,1.84140593236139e-05,1.86047450744404e-05,1.88281629275467e-05,1.89118237079529e-05,1.9697792131576e-05,2.01119149232581e-05,2.06679593816395e-05,2.07032348342937e-05,2.12339896910015e-05,2.15236522976962e-05,2.15881432598887e-05,2.23123320928352e-05,2.28639812738187e-05,2.33329593662217e-05,2.33715603032657e-05,2.36001201966235e-05,2.38742985464536e-05,2.40917977955115e-05,2.4240103253488e-05,2.49456679415096e-05,2.50174355365162e-05,2.51118551123481e-05,2.51198453159111e-05,2.5632162258907e-05,2.7051197936012e-05,2.70524479131681e-05,2.73224803097896e-05,2.80734768647689e-05,2.87805961865083e-05,2.93450030594677e-05,2.96680658999386e-05,3.04372535882259e-05,3.10228172862477e-05,3.13517368110509e-05,3.19568312033656e-05,3.32054805387285e-05,3.35109428116793e-05,3.35382762731509e-05,3.39484407281618e-05,3.4256860881357e-05,3.44687060578377e-05,3.45707143526291e-05,3.48155955156801e-05,3.49682707477043e-05,3.5424684327855e-05,3.57804122748337e-05,3.65065994527347e-05,3.77991160777686e-05,3.81628073831624e-05,3.86283578258535e-05,3.88199131975736e-05,3.93615356332351e-05,3.98592584654501e-05,4.00778011475446e-05,4.18866271441529e-05,4.21476955816406e-05,4.31552288020486e-05,4.40786449253218e-05,4.69774318495533e-05,4.75036734080972e-05,4.81000539962669e-05,4.91185679464206e-05,5.04125212451418e-05,5.07871210516361e-05,5.0994825961635e-05,5.31993186096224e-05,5.69927776794974e-05,5.77244368458039e-05,5.79345707378957e-05,5.94621488006291e-05,6.04072569519089e-05,6.29767125945643e-05,6.35898599305048e-05,6.44913369696997e-05,6.46455823603803e-05,6.61379904462487e-05,6.75573205125058e-05,6.91437008924444e-05,7.0769668494931e-05,7.33728856984787e-05,7.47178820028937e-05,7.50229220724866e-05,7.62304848946384e-05,7.67659342665269e-05,7.78379929322643e-05,7.89632382121031e-05,8.10950114354684e-05,8.28248162935586e-05,8.2865779129629e-05,8.30024726778695e-05,8.31410241911168e-05,8.37781857397666e-05,8.65643732775154e-05,8.6721834377291e-05,8.69014087431749e-05,8.78561816234231e-05,9.06031617884838e-05,9.0694606374485e-05,9.17360938544815e-05,9.1821056805125e-05,9.24023010556715e-05,9.30393979633141e-05,9.36752372963439e-05,9.58938314894523e-05,9.69616779466917e-05,9.77043559810292e-05,9.78025265237098e-05,9.84323129966538e-05,9.86023825379561e-05,9.88143605853571e-05,0.00010063105932209,0.000101904546226038,0.00010211985856005,0.000103429480410375,0.000103807207363336,0.000104256431440063,0.000106023010704382,0.000107631731963783,0.000108729986449183,0.000113841177385203,0.000115263233469864,0.000116686126139333,0.000118530052794774,0.000118841337003604,0.000120730737837701,0.000122150085563421,0.000123022559308785,0.000125272744740728,0.00012679888502821,0.000130458045617701,0.000131021283677071,0.000131967815203595,0.000132196750185326,0.000133852617956777,0.000137139735136937,0.000144674517079247,0.000150540165804098,0.000151821080120826,0.000152831381873233,0.000154141811360087,0.000154179948791735,0.000155645930421384,0.00015694000775392,0.000158952202539406,0.000165704501311817,0.000166340641173939,0.00016913994822989,0.000170204433494299,0.000172602840205371,0.00017355843761645,0.000175502933531467,0.000176744093648993,0.000178273429140276,0.000178423581243167,0.000180503508132169,0.000180711690298959,0.000181953904641431,0.000182415111816331,0.000184412357525395,0.000191668977865852,0.000192013654450802,0.000195457584725526,0.000199000501311306,0.000199421474987228,0.000200378447642497,0.000201268084899137,0.000203878890617012,0.000207317505040809,0.000208652778089097,0.000214390301515384,0.000219191695104752,0.000221575599309741,0.000223473637652927,0.0002285114534486,0.000230142670056527,0.000231218689182554,0.000231758248549773,0.00023190054441929,0.000232246056421578,0.000233285520417551,0.000235289635892629,0.000235410332858775,0.000236071462231192,0.000238193239226526,0.000241169652900249,0.000247437180252655,0.000248959012350975,0.000250210107678131,0.00025466274406226,0.000257670979060819,0.000262004089999653,0.000268841065639772,0.000270794098342026,0.000283724036048127,0.000286795233105775,0.00028732252818254,0.000287727144505209,0.000288504302317016,0.000290068439294417,0.000290168585914446,0.000293958281274406,0.000295155070159994,0.000297047977682943,0.000297503429591334,0.000298788174993118,0.000302642861388608,0.000310747213946613,0.000313654695398593,0.000316839765549745,0.000318085500866264,0.000327147908408352,0.000328612164150681,0.000329742866450628,0.000330085705595729,0.000337391045236033,0.000338476308880635,0.000339397858458335,0.000339626189484932,0.000340590665804305,0.000340974406828189,0.000342435125085539,0.000342802964265044,0.000343409830652041,0.000346356646663554,0.000349225365776516,0.00035714216404285,0.000357892430407832,0.000359636050037062,0.000361173404345156,0.000361897792729276,0.000369432978117946,0.000374514576270259,0.000374970244680571,0.000376407897238832,0.000382986274665387,0.000385027677682215,0.000388133080198456,0.000390103288908546,0.00039273321301634,0.000397223833882014,0.000398544900207469,0.000415351956645619,0.00041662265393549,0.000417710610022892,0.000420878365522096,0.000422469966379647,0.000427811857127423,0.000431770830681107,0.00043272866748737,0.000436209001835634,0.000439700107992205,0.000447763316737349,0.000455381179814151,0.000456748597522271,0.000464723285767364,0.000464749945228983,0.000468887440411735,0.000471719567755345,0.000473969943673566,0.00047635102142,0.000477492086728732,0.000483618633407126,0.000503281801442249,0.000503434498403834,0.000503577345517457,0.000508375489288066,0.00051194159399223,0.000523036059130949,0.000523235623779512,0.000538697945495046,0.000540592627513106,0.000542872298036576,0.000551027836555642,0.00055334134809593,0.000557779186072245,0.000564487430698742,0.000584940441024243,0.000592042255546495,0.000594133713057888,0.000596543387945796,0.00059970145727995,0.000600606636536964,0.000608111143415432,0.000609031210448388,0.000610690029800124,0.000617855214222879,0.000620666622043723,0.000621019400405368,0.000624188160261163,0.000627687764172371,0.00063399404615958,0.000637112609528384,0.00063776006624261,0.000640298497645778,0.000640983619517798,0.000641111193302549,0.000645058171803913,0.000646617734256154,0.000649671276725319,0.000657853679169235,0.000658484791864109,0.000659185474489814,0.000661136529240573,0.000664732708230457,0.000664873267437291,0.000668329423708325,0.000670913401699535,0.000674661078824778,0.000682124221663461,0.000685317622643121,0.000686524034277301,0.000689859880011963,0.000708522440955453,0.000715955112101685,0.000717502914007209,0.000720227477500268,0.000723765486932433,0.000724261922851897,0.000727582590444768,0.0007306166877764,0.00073497883611763,0.00074371402896283,0.000747967264861981,0.000752192942097257,0.000753670081714087,0.000753926639018617,0.000754563581483443,0.000754621838608722,0.000754975073714931,0.000756251510778725,0.000763371155236576,0.000765387318907635,0.000766117973209167,0.000781705012061412,0.000781926880463836,0.00078664902049159,0.000794993528222783,0.000798299247931588,0.000801251739407861,0.000802563638292188,0.000811419315280485,0.000811850026049797,0.000812043631665962,0.000812170503403786,0.000817650210357053,0.00081964414322001,0.000819854019335269,0.0008227260089536,0.00082694065462214,0.000828028034214135,0.000834463897638125,0.00083807332229847,0.00084695418294557,0.000849366386208069,0.00085040079172092,0.000853847877861747,0.000854459579831007,0.000857262226768088,0.000858439531059289,0.000860635005950512,0.000881827034537577,0.000886962291127711,0.00090253543512844,0.000938396731819659,0.00095008531247821,0.000961600235427576,0.000963962940721393,0.000966541949650518,0.000970644364892035,0.000986241972191063,0.00098668839749547,0.000987999888333547,0.000993976818645102,0.000994430094276045,0.00100162498856219,0.00100259415587579,0.00100961896529212,0.00101034603966261,0.00101132877060614,0.00102138367790974,0.00102642298345716,0.00103776346699702,0.00105382043095949,0.00105495378777824,0.00105538455091823,0.00106308505218118,0.00106667202557513,0.00106924326017114,0.00107022177057114,0.0010717416397667,0.0010720880493121,0.00107518655682504,0.00109614931270724,0.00110166102499321,0.00110347641364375,0.00111415871930577,0.00111618794659651,0.00112412188569107,0.00112715507081534,0.00113082788242869,0.00113819676751599,0.00114052202091853,0.00116128599110532,0.00116520214363166,0.00116761510701537,0.00117498762639924,0.0011789408320537,0.00118050356734477,0.00118968431841549,0.00119372072723772,0.00120691164679855,0.00121151937574525,0.00121333683067309,0.00121739742847268,0.00122409083006182,0.00122762289577503,0.00124398217330479,0.00124636067390636,0.00124674455836749,0.00124819460956582,0.00125134179033909,0.00125452413094122,0.00125493442158097,0.00125907447664013,0.00126143531033553,0.00126219003103416,0.00126751143287819,0.00126935177569962,0.00126954403066348,0.00127063201083605,0.00127547004134097,0.00128799246567003,0.00129153713652277,0.00130614654278024,0.00130670502436668,0.00130780301490623,0.00131304365704682,0.00131993658376717,0.00132695049001293,0.00133779652107787,0.00134019173511181,0.00134433839099225,0.00134636248568398,0.0013566308721943,0.00136637582955736,0.00137002909343216,0.00137238710674761,0.0013761894311881,0.00137637119027526,0.0013836394105054,0.00138920747828254,0.00138929237149831,0.00139473421460195,0.00140606355363512,0.00141560298873836,0.00141785354397414,0.00142796227461359,0.00143675328290176,0.00143916867848798,0.00144283172891371,0.00145238405489281,0.00146622169308674,0.00146999424965615,0.00147150956266217,0.00147330366472681,0.00147443955255984,0.00147473997413963,0.00148156191440385,0.00148371531215461,0.00149509773100032,0.00150726246385612,0.00151295103483471,0.00152030593082594,0.00152118214615129,0.00153277138963979,0.00154080068343165,0.00154109704621148,0.00154678721720615,0.00155434752141642,0.00156769488580959,0.00159305732448409,0.00160187531573498,0.00160510337930772,0.00162612497801736,0.00162834589666411,0.0016354553382236,0.00165697471111562,0.00167439435200846,0.00169823014898589,0.0017056886088621,0.00170726541801325,0.00172908642739997,0.00175606088476211,0.00176340945440446,0.00176371486636069,0.00176811416834932,0.00177341044306191,0.00179249092451894,0.0018060875358481,0.00180756609420693,0.00181671688444897,0.00182707282413805,0.00183146019907759,0.00183155873780598,0.00183677759489148,0.00183731960176217,0.00183779402351649,0.00184223361116175,0.00184646047821157,0.00184762505940459,0.00185011205072391,0.00186360156148469,0.00187488406811551,0.00187759360935535,0.00188877523037678,0.00190998309647328,0.00191932772183232,0.00192006774111252,0.00192436400647202,0.00194584437082043,0.00194762376591459,0.00195142552499411,0.00195296403653594,0.00195532760345461,0.0019718813341661,0.00199193935378552,0.00200994857625684,0.00201861976353406,0.00202086874111289,0.00202201524007922,0.00202705500956665,0.00202987624991755,0.00203001619530336,0.00203918241183677,0.00203931666335985,0.00205581008244932,0.00205787394412221,0.00206060328430443,0.00207618387759346,0.00207652701595169,0.0020860738835071,0.00208934654066567,0.00211771868360212,0.00212210223102757,0.0021229009862013,0.00212553735307418,0.00212568669024315,0.0021277115160454,0.00212987829664327,0.00213257945180188,0.00214324506846687,0.00215099249969865,0.002153237771346,0.00215326499913149],[5.82397648232487e-18,5.88121359906241e-17,3.00223705225742e-14,2.27995792256587e-13,3.95676774212801e-13,3.95676774212801e-13,1.23881776718907e-12,6.89755052826151e-12,9.62503911962547e-12,1.13828558584297e-11,1.13828558584297e-11,3.98340499768256e-11,4.38341587165467e-11,7.96809720764633e-11,1.39098741762449e-10,3.41179746025925e-10,5.61641615817504e-10,5.89738013368171e-10,2.01694976358682e-09,2.01694976358682e-09,4.62970663319246e-09,4.89140921662476e-09,9.92420465059565e-09,1.2008305699748e-08,1.2008305699748e-08,1.2008305699748e-08,1.57321340907227e-08,1.96880005683411e-08,2.62811006190872e-08,2.90813073549234e-08,3.26063692953021e-08,3.33298270587897e-08,5.30082983147438e-08,7.22837591356911e-08,1.26582258834849e-07,1.42064304571382e-07,1.56597758617516e-07,1.64012720977992e-07,1.66541987755796e-07,1.85421136456659e-07,1.85421136456659e-07,2.06830976625781e-07,3.34377101477924e-07,3.5957591040383e-07,3.60182895626348e-07,3.60182895626348e-07,4.24894788797015e-07,4.84022629230953e-07,5.39439936369533e-07,6.83884576568635e-07,7.03594973359093e-07,7.38603694897353e-07,7.66233966257586e-07,7.66233966257586e-07,9.12616060782966e-07,1.00388063487864e-06,1.12404454839891e-06,1.23298741968171e-06,1.26760053191174e-06,1.49259254208165e-06,1.88534338416019e-06,1.92001488462029e-06,2.12495780356463e-06,2.23137193600324e-06,2.23137193600324e-06,2.25662959103178e-06,2.34073769637668e-06,2.34975374332443e-06,2.39014408283969e-06,2.39014408283969e-06,2.48679957587938e-06,2.94974626592339e-06,2.97855610964763e-06,3.06310005674503e-06,3.44373482057753e-06,3.44373482057753e-06,3.51122790641375e-06,3.6756217720204e-06,4.31422225807162e-06,4.32570000569774e-06,4.82167205738062e-06,5.40212554441489e-06,5.69799308289572e-06,5.69912158246827e-06,6.94905267894474e-06,6.94905267894474e-06,6.94905267894474e-06,7.13951875289394e-06,7.84654974337253e-06,9.73376174859257e-06,1.00388185063672e-05,1.03754452349959e-05,1.08613453539305e-05,1.10254056839612e-05,1.16146493903944e-05,1.23187835446289e-05,1.23187835446289e-05,1.27931558953562e-05,1.35709414204998e-05,1.5704150499576e-05,1.64964274043311e-05,1.82449602201427e-05,1.82449602201427e-05,2.04592926719133e-05,2.07689891770906e-05,2.35371348992382e-05,2.35621271531634e-05,2.75387393514673e-05,2.86741967721063e-05,2.90599678136794e-05,2.94618162604041e-05,2.98085296524474e-05,3.49922623268403e-05,3.58072567334365e-05,3.9001293570521e-05,3.92579198890476e-05,4.08516808198143e-05,4.18844232497924e-05,4.45514378033898e-05,4.82514018940375e-05,4.95003536651556e-05,5.07845206634812e-05,5.25260234861386e-05,5.29040593525412e-05,5.32289220190843e-05,5.58728741648771e-05,5.88671001017314e-05,5.88671001017314e-05,5.88671001017314e-05,5.88671001017314e-05,5.965217100256e-05,6.29680073832238e-05,6.44415632865999e-05,6.76071713712123e-05,6.87114135063568e-05,6.91642580972142e-05,6.94754580507094e-05,6.97303285324211e-05,7.10636622227326e-05,7.14791007965992e-05,7.19639338703775e-05,7.62724015305577e-05,8.08604134913794e-05,8.13587033430277e-05,9.15514017389208e-05,9.18436316628659e-05,9.20716953470986e-05,9.90519942123929e-05,0.000128002431478013,0.000136355911931306,0.000140009649305145,0.000140009649305145,0.00015149487498933,0.000153935422701449,0.000158465655490327,0.000173312074762722,0.000188219621968019,0.000194837656044868,0.000200948118463807,0.000209045344554863,0.000225363447661616,0.000225363447661616,0.000235022765359143,0.000244433559505947,0.000244433559505947,0.000249952588363766,0.000255340306761612,0.000255340306761612,0.000263149623406698,0.00029554959015763,0.000298291941487618,0.000298987569832707,0.000303485978130785,0.000338462830997239,0.000365784277139422,0.000376966042020943,0.000379176800855163,0.000405287103074435,0.000420698700892412,0.000429481053684907,0.000429481053684907,0.000444052695592384,0.000444052695592384,0.000486100716875372,0.000500213220288619,0.000511569292269642,0.000514592700786007,0.000518114357110147,0.000523649494864706,0.000526505374243341,0.000533531273793583,0.000540649748298177,0.000540919121409973,0.000541604923362049,0.000570408032618132,0.000589793548306324,0.000591355661898612,0.000614490989328611,0.000644379062481899,0.000646246536144645,0.000646246536144645,0.000702335612736769,0.000702335612736769,0.000715549362220691,0.000744494896895451,0.000757262349132203,0.000787681762414884,0.000787681762414884,0.000787681762414884,0.000789399905732167,0.000854869448345468,0.000889811323359341,0.000894859946232207,0.000939169197688428,0.000994392544879391,0.00102032713328243,0.00108452088028574,0.00109270282580381,0.00109270282580381,0.0010984184101988,0.0010984184101988,0.0010984184101988,0.0010984184101988,0.0010984184101988,0.0010984184101988,0.00110613062754531,0.00116036316282373,0.00116036316282373,0.00118157658331617,0.00121248884779347,0.00127164065603215,0.00128648358333877,0.00130138272210445,0.00130626880042462,0.0013171647871554,0.00136738104953916,0.00139170737337263,0.00139348043035754,0.00139956514996673,0.00140905916449237,0.00141774333316223,0.00142883971968758,0.00142928247899816,0.0014825817725262,0.00150757272475769,0.00153933161025022,0.00153933161025022,0.00157242830917436,0.00158586500387142,0.00158586500387142,0.001632533780418,0.00166625800037175,0.00168983742113967,0.00168983742113967,0.00169967140161173,0.00171270114377195,0.0017215792471384,0.00172546316376088,0.00176078610391873,0.00176078610391873,0.00176078610391873,0.00176078610391873,0.0017898656269385,0.0018747856827371,0.0018747856827371,0.00188638101838077,0.0019309715453988,0.00197222257076576,0.00200342372188522,0.0020179778898236,0.00206265742489952,0.00209461043919831,0.00210906463932216,0.0021419241060212,0.00221752236397727,0.00222357560922894,0.00222357560922894,0.00224267307184422,0.00225493638023699,0.00225939917824212,0.00225939917824212,0.00226733479306903,0.00226923071477594,0.0022907546749333,0.00230563954886779,0.00234420873758557,0.00241874831626558,0.00243354151941589,0.00245470516080208,0.00245837139956358,0.00248410516118338,0.00250690164971915,0.00251204374769507,0.00261648948130057,0.00262387264188756,0.00267751951672169,0.002725603751022,0.00289510246952029,0.00291774234829333,0.00294452497213814,0.00299688538317613,0.00306564884989083,0.00307823590136402,0.00308065782495206,0.00320329667628104,0.00342049791530709,0.0034531246992612,0.00345444282987485,0.00353405295379791,0.00357864281910261,0.00371886600256969,0.00374303774879398,0.00378094305747893,0.00378094305747893,0.00385594982395351,0.00392623478231699,0.00400575415422631,0.00408705962864593,0.00422411613119925,0.00428810594682232,0.00429219926436516,0.00434774178599389,0.00436472564335841,0.00441202080308961,0.00446203036850853,0.00456843523009932,0.00462692396748442,0.00462692396748442,0.00462692396748442,0.00462692396748442,0.00464829722389974,0.00477827656158206,0.00477827656158206,0.00477827656158206,0.00481635455377363,0.00494245236221786,0.00494245236221786,0.00497431772338089,0.00497431772338089,0.00499108311437473,0.00501075819236441,0.00503025068113262,0.00513437380555041,0.00517645702177614,0.00519116589482061,0.00519116589482061,0.00519978719813778,0.00519978719813778,0.00519978719813778,0.00528025544128622,0.00532792955242987,0.00532792955242987,0.0053809699935879,0.00538536543284652,0.00539343482646975,0.00546941739209546,0.00553685366250665,0.00557772681882473,0.00582365800189206,0.0058800257852057,0.00593612384085556,0.0060124549698931,0.0060124549698931,0.00609126373733344,0.00614598992156773,0.00617297623416893,0.00626875737646722,0.00632788457484531,0.00649285096956388,0.00650325912089027,0.00652632612138041,0.00652632612138041,0.00659035744980218,0.00673414768927768,0.00708519334976097,0.00735284613029856,0.00739574041490441,0.00742526012725377,0.00745135463042161,0.00745135463042161,0.00750246066191265,0.00754503466597052,0.00762182036458534,0.00792490407966542,0.00793466461080363,0.00804729313275111,0.00807701400806924,0.0081697194854939,0.00819383215122392,0.00826438813924461,0.00830154803034209,0.00833778389193577,0.00833778389193577,0.0084019498540263,0.0084019498540263,0.00843834206752497,0.0084384219861635,0.00850937926119064,0.00881582690997247,0.00881582690997247,0.00895156744011042,0.00908778011945519,0.00908778011945519,0.00910878760137243,0.0091266379732658,0.00922225572951088,0.00935475670779965,0.00939193203334869,0.00962659630154042,0.0098181840990214,0.0099008172294973,0.00996139163955342,0.0101612901757471,0.0102038249430198,0.0102038249430198,0.0102038249430198,0.0102038249430198,0.0102038249430198,0.0102250324163922,0.0102691467053478,0.0102691467053478,0.0102735838954404,0.0103414156936055,0.0104459449894176,0.0106921972125647,0.0107327048399663,0.0107613785187561,0.0109272927446341,0.0110306003040838,0.011190011890334,0.0114553739454163,0.0115118833704891,0.0120336995889696,0.0121195160753169,0.0121195160753169,0.0121195160753169,0.0121244428193409,0.0121388293401339,0.0121388293401339,0.012269417808192,0.0122914350645993,0.0123332968046159,0.0123332968046159,0.0123586595354699,0.0124899688750602,0.0127956784397523,0.012886506668893,0.012988308692681,0.0130103345732938,0.0133512696398208,0.0133812913406369,0.0133819513979372,0.0133819513979372,0.0136426905912847,0.0136426905912847,0.0136426905912847,0.0136426905912847,0.0136426905912847,0.0136426905912847,0.0136509124240795,0.0136509124240795,0.0136509124240795,0.0137383149373135,0.0138222496605295,0.0141044946018022,0.0141044946018022,0.0141428609398943,0.0141711150607103,0.0141711150607103,0.0144353971130555,0.0145896791177091,0.0145896791177091,0.0146146533462815,0.0148386981734807,0.0148863858960713,0.0149749244072366,0.0150193855362798,0.0150890072323119,0.0152296778898605,0.0152484939423128,0.0158585003821139,0.0158740146048242,0.015882516258945,0.015969899138044,0.0159972390362108,0.016166182625813,0.0162822819413933,0.0162849630705032,0.0163823687499211,0.016479780578116,0.016747807152508,0.0169981206652172,0.0170145801085122,0.0172426924123844,0.0172426924123844,0.0173611246837934,0.0174308447924083,0.0174788313565563,0.0175314358885337,0.0175382843455463,0.0177278566916604,0.0183495991079923,0.0183495991079923,0.0183495991079923,0.0184877541797531,0.0185806469835322,0.0189157917927377,0.0189157917927377,0.0194365182102486,0.0194666345181925,0.0195104691848174,0.0197648949577038,0.0198091888065921,0.0199292115801883,0.020129731387927,0.0208186651151361,0.0210306692903508,0.0210642193828342,0.0211089004231687,0.0211710957389661,0.0211710957389661,0.0213859621030299,0.0213859621030299,0.0214032870177086,0.0216131638270537,0.0216414066194394,0.0216414066194394,0.0217106355363566,0.0217910884480635,0.0219684918070202,0.0220074898411239,0.0220074898411239,0.0220074898411239,0.0220074898411239,0.0220074898411239,0.0221016666514531,0.0221138448596169,0.0221769758309674,0.0223769708669232,0.0223769708669232,0.0223769708669232,0.0224017940212235,0.0224455837435402,0.0224455837435402,0.0225208621401897,0.0225665286121098,0.0226510981949123,0.0228598746913311,0.0229236616172775,0.0229236616172775,0.0229932426432299,0.023572490268382,0.0237767009651853,0.0237850920861776,0.0238323921158422,0.0238798388028278,0.0238798388028278,0.0239463338235093,0.0240031761556594,0.0241033684380362,0.0243463603242467,0.0244087152819882,0.0244087152819882,0.0244087152819882,0.0244087152819882,0.0244087152819882,0.0244087152819882,0.0244087152819882,0.0244087152819882,0.0245952829226662,0.0245974765349412,0.0245974765349412,0.0250175734489867,0.0250175734489867,0.0251248856718749,0.0253473196975893,0.0254086060455175,0.0254560988207876,0.0254560988207876,0.0255840674013903,0.0255840674013903,0.0255840674013903,0.0255840674013903,0.0256938891895772,0.0256938891895772,0.0256938891895772,0.0257399713022706,0.0258178859903949,0.0258178859903949,0.0259744567459732,0.0260426676210007,0.0262741783273571,0.0262922736362874,0.0262922736362874,0.0263291110463028,0.0263291110463028,0.0263632809162272,0.0263632809162272,0.0263865807750937,0.026991255815471,0.027103265352014,0.0275333276846077,0.0285798606631311,0.0288879416616926,0.0291897327663263,0.0292131673372086,0.0292430690367904,0.029318887765201,0.0296965923882906,0.0296965923882906,0.0296965923882906,0.0297923469516796,0.0297923469516796,0.029939254752291,0.029939254752291,0.0300534836119445,0.0300534836119445,0.0300534836119445,0.030303249183865,0.0304036420825657,0.0306900580859908,0.0310611174320727,0.0310611174320727,0.0310611174320727,0.0312376911732919,0.0312522175009788,0.0312522175009788,0.0312522175009788,0.0312522175009788,0.0312522175009788,0.0312928702315243,0.0318525033668806,0.0319620927709327,0.0319642655150906,0.0322228738268511,0.0322308044642216,0.0324089457311092,0.032445459052545,0.0325002410967182,0.0326609119303611,0.0326765786492494,0.0332196530010113,0.033279840385374,0.0332969742862381,0.0334552678431349,0.0335084204239361,0.0335084204239361,0.0337169020180563,0.0337791697314649,0.0340998959899313,0.0341762743793117,0.0341762743793117,0.0342381374791743,0.0343737432631275,0.0344202969174174,0.0347846646504951,0.0347846646504951,0.0347846646504951,0.0347846646504951,0.0348140040065477,0.0348140040065477,0.0348140040065477,0.0348573231878833,0.0348573231878833,0.0348573231878833,0.0348806530328908,0.0348806530328908,0.0348806530328908,0.0348806530328908,0.0349612049391447,0.0352518355171833,0.0352962492741678,0.0355819294351896,0.0355819294351896,0.0355819294351896,0.0356716668071966,0.0358059606512321,0.0359431353231379,0.0361835539169293,0.0361950311990124,0.0362537071227204,0.0362550543249067,0.0364780760876257,0.0366863919734224,0.0367307800012869,0.036739908298554,0.036739908298554,0.036739908298554,0.0368803160724698,0.0369238124494449,0.0369238124494449,0.0370148755074637,0.0372616986471991,0.0374604450838328,0.037466014870626,0.0376789183524118,0.0378564907324115,0.0378658062756901,0.0379078751094423,0.0381043330972948,0.0383620391290003,0.0383620391290003,0.0383620391290003,0.0383620391290003,0.0383620391290003,0.0383620391290003,0.0384849852305893,0.0384864854628805,0.0387270378417783,0.0389871480967854,0.0390792486001963,0.0391816411137006,0.0391816411137006,0.0394248551410851,0.0395282782872539,0.0395282782872539,0.0396188943430837,0.0397570922434716,0.0400427212488082,0.040634024672431,0.0408022748591856,0.0408278719681249,0.0413046580003264,0.0413046580003264,0.0414277755675536,0.0419150696551492,0.0422974584245328,0.0428406547886343,0.0429505882216622,0.0429505882216622,0.0434400441028735,0.0440574564872351,0.044128914878357,0.044128914878357,0.0441787982336534,0.0442509276995,0.044666344408128,0.0449200965089448,0.0449200965089448,0.0450864940309531,0.0452427442920647,0.0452427442920647,0.0452427442920647,0.0452427442920647,0.0452427442920647,0.0452427442920647,0.0452913256612925,0.0453025823978175,0.0453025823978175,0.0453030770820595,0.0455726267332441,0.0457875610517838,0.045792837497757,0.0460044523950526,0.0464593901546115,0.0465812999544669,0.0465812999544669,0.046623937966832,0.0470633032381862,0.0470633032381862,0.0470636847148674,0.0470636847148674,0.0470636847148674,0.0474000009187964,0.0478195637023151,0.0481889107088209,0.0482890895761442,0.0482890895761442,0.0482890895761442,0.0482917712781687,0.0482917712781687,0.0482917712781687,0.0483876621739065,0.0483876621739065,0.0487021327110881,0.0487021327110881,0.0487039630839779,0.0489543243234311,0.0489543243234311,0.049116342141805,0.0491304087315302,0.0496385572019537,0.0496385572019537,0.0496385572019537,0.0496385572019537,0.0496385572019537,0.0496385572019537,0.0496385572019537,0.0496385572019537,0.0498236654207521,0.0498672278802646,0.0498672278802646,0.0498672278802646]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>logFC<\/th>\n      <th>AveExpr<\/th>\n      <th>P.Value<\/th>\n      <th>adj.P.Val<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"targets":1,"render":"function(data, type, row, meta) {\n    return type !== 'display' ? data : DTWidget.formatSignif(data, 4, 3, \",\", \".\", null);\n  }"},{"targets":2,"render":"function(data, type, row, meta) {\n    return type !== 'display' ? data : DTWidget.formatSignif(data, 4, 3, \",\", \".\", null);\n  }"},{"targets":3,"render":"function(data, type, row, meta) {\n    return type !== 'display' ? data : DTWidget.formatSignif(data, 4, 3, \",\", \".\", null);\n  }"},{"targets":4,"render":"function(data, type, row, meta) {\n    return type !== 'display' ? data : DTWidget.formatSignif(data, 4, 3, \",\", \".\", null);\n  }"},{"className":"dt-right","targets":[1,2,3,4]},{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false,"rowCallback":"function(row, data, displayNum, displayIndex, dataIndex) {\nvar value=data[1]; $(this.api().cell(row, 1).node()).css({'background-color':isNaN(parseFloat(value)) ? '' : value <= 0 ? \"#DADAEB\" : \"#C7E9C0\"});\n}"}},"evals":["options.columnDefs.0.render","options.columnDefs.1.render","options.columnDefs.2.render","options.columnDefs.3.render","options.rowCallback"],"jsHooks":[]}</script>
</div>
</div>
</div>
<div class="section level2">
<h2 id="gsea-and-visualisation-with-visse">GSEA and visualisation with vissE<a class="anchor" aria-label="anchor" href="#gsea-and-visualisation-with-visse"></a>
</h2>
<p>Here we try to do GSEA with the DE genes using limma fry. We select
the following gene sets to conduct gene set enrichment analysis:</p>
<ul>
<li>MSigDB Hallmarks - genesets from the hallmarks collection of
MSigDB</li>
<li>MSigDB C2 - genesets from the C2 collection of MSigDB which contains
curated genesets such as those obtained from databases such as BioCarta,
KEGG, PID, and Reactome, and from chemical or genetic perturbation
experiments</li>
<li>GO BP - biological processes from the gene ontology database</li>
<li>GO MF - molecular functions from the gene ontology database</li>
<li>GO CC - cellular component from the gene ontolgoy database</li>
</ul>
<p>FDR &lt; 0.05 indicates significantly enriched gene set.</p>
<div class="section level3">
<h3 id="load-gene-sets">Load Gene sets<a class="anchor" aria-label="anchor" href="#load-gene-sets"></a>
</h3>
<div class="sourceCode" id="cb69"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://davislaboratory.github.io/msigdb" class="external-link">msigdb</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">GSEABase</span><span class="op">)</span></span>
<span></span>
<span><span class="va">msigdb_hs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/msigdb/man/getMsigdb.html" class="external-link">getMsigdb</a></span><span class="op">(</span>version <span class="op">=</span> <span class="st">'7.2'</span><span class="op">)</span></span>
<span><span class="va">msigdb_hs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/msigdb/man/appendKEGG.html" class="external-link">appendKEGG</a></span><span class="op">(</span><span class="va">msigdb_hs</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/msigdb/man/listSubCollections.html" class="external-link">listSubCollections</a></span><span class="op">(</span><span class="va">msigdb_hs</span><span class="op">)</span></span>
<span></span>
<span><span class="va">gsc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/msigdb/man/subsetCollection.html" class="external-link">subsetCollection</a></span><span class="op">(</span><span class="va">msigdb_hs</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'h'</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/msigdb/man/subsetCollection.html" class="external-link">subsetCollection</a></span><span class="op">(</span><span class="va">msigdb_hs</span>, <span class="st">'c2'</span>, <span class="va">sc</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"^CP:"</span>,<span class="va">sc</span><span class="op">)</span><span class="op">]</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/msigdb/man/subsetCollection.html" class="external-link">subsetCollection</a></span><span class="op">(</span><span class="va">msigdb_hs</span>, <span class="st">'c5'</span>, <span class="va">sc</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"^GO:"</span>,<span class="va">sc</span><span class="op">)</span><span class="op">]</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/GSEABase/man/GeneSetCollection-methods.html" class="external-link">GeneSetCollection</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="enrichment-analysis">Enrichment analysis<a class="anchor" aria-label="anchor" href="#enrichment-analysis"></a>
</h3>
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fry_indices</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/ids2indices.html" class="external-link">ids2indices</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">gsc</span>, <span class="va">geneIds</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">v</span><span class="op">)</span>, remove.empty <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">fry_indices</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">gsc</span>, <span class="va">setName</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">gsc_category</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">gsc</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/pkg/GSEABase/man/CollectionType-class.html" class="external-link">bcCategory</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/GSEABase/man/GeneSet-class.html" class="external-link">collectionType</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">gsc_category</span> <span class="op">&lt;-</span> <span class="va">gsc_category</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">fry_indices</span>, <span class="va">length</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">5</span><span class="op">]</span></span>
<span></span>
<span></span>
<span><span class="va">gsc_subcategory</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">gsc</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/pkg/GSEABase/man/CollectionType-class.html" class="external-link">bcSubCategory</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/GSEABase/man/GeneSet-class.html" class="external-link">collectionType</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">gsc_subcategory</span> <span class="op">&lt;-</span> <span class="va">gsc_subcategory</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">fry_indices</span>, <span class="va">length</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">5</span><span class="op">]</span></span>
<span></span>
<span><span class="va">fry_indices</span> <span class="op">&lt;-</span> <span class="va">fry_indices</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">fry_indices</span>, <span class="va">length</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">5</span><span class="op">]</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">gsc_category</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">gsc_subcategory</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">fry_indices</span><span class="op">)</span></span>
<span></span>
<span><span class="va">fry_indices_cat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/split.html" class="external-link">split</a></span><span class="op">(</span><span class="va">fry_indices</span>, <span class="va">gsc_category</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">fry_indices</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">fry_res_out</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">fry_indices_cat</span>, <span class="kw">function</span> <span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu">limma</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/limma/man/roast.html" class="external-link">fry</a></span><span class="op">(</span><span class="va">v</span>, index <span class="op">=</span> <span class="va">x</span>, design <span class="op">=</span> <span class="va">design</span>, contrast <span class="op">=</span> <span class="va">contr.matrix</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>, robust <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span></span>
<span><span class="va">post_fry_format</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">fry_output</span>, <span class="va">gsc_category</span>, <span class="va">gsc_subcategory</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">fry_output</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span>
<span>  <span class="va">fry_output</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">rbind</span>, <span class="va">fry_output</span><span class="op">)</span></span>
<span>  <span class="va">fry_output</span><span class="op">$</span><span class="va">GenesetName</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">fry_output</span><span class="op">)</span></span>
<span>  <span class="va">fry_output</span><span class="op">$</span><span class="va">GenesetCat</span> <span class="op">&lt;-</span> <span class="va">gsc_category</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">fry_output</span><span class="op">)</span><span class="op">]</span></span>
<span>  <span class="va">fry_output</span><span class="op">$</span><span class="va">GenesetSubCat</span> <span class="op">&lt;-</span> <span class="va">gsc_subcategory</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">fry_output</span><span class="op">)</span><span class="op">]</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">fry_output</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">fry_res_sig</span> <span class="op">&lt;-</span> <span class="fu">post_fry_format</span><span class="op">(</span><span class="va">fry_res_out</span>, <span class="va">gsc_category</span>, <span class="va">gsc_subcategory</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">FDR</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span> </span></code></pre></div>
</div>
<div class="section level3">
<h3 id="visualization-1">Visualization<a class="anchor" aria-label="anchor" href="#visualization-1"></a>
</h3>
<p>To interpret the GSEA output, we can use the <code>vissE</code>
package to visualise the important enriched gene sets.</p>
<div class="sourceCode" id="cb71"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://davislaboratory.github.io/vissE" class="external-link">vissE</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://igraph.org" class="external-link">igraph</a></span><span class="op">)</span></span>
<span><span class="va">dovissE</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">fry_out</span>, <span class="va">de_table</span>, <span class="va">topN</span> <span class="op">=</span> <span class="fl">6</span>, <span class="va">title</span> <span class="op">=</span> <span class="st">""</span>, <span class="va">specific_clusters</span> <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span><span class="op">{</span></span>
<span>  </span>
<span>  <span class="va">n_row</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="fl">1000</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">fry_out</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">gs_sig_name</span> <span class="op">&lt;-</span> <span class="va">fry_out</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">FDR</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="va">FDR</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="va">.</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="va">n_row</span>,<span class="op">]</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="op">)</span></span>
<span>  <span class="va">gsc_sig</span> <span class="op">&lt;-</span> <span class="va">gsc</span><span class="op">[</span><span class="va">gs_sig_name</span>,<span class="op">]</span></span>
<span>  </span>
<span>  <span class="va">gs_ovlap</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://DavisLaboratory.github.io/vissE/reference/computeMsigOverlap.html" class="external-link">computeMsigOverlap</a></span><span class="op">(</span><span class="va">gsc_sig</span>, thresh <span class="op">=</span> <span class="fl">0.15</span><span class="op">)</span></span>
<span>  <span class="va">gs_ovnet</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://DavisLaboratory.github.io/vissE/reference/computeMsigNetwork.html" class="external-link">computeMsigNetwork</a></span><span class="op">(</span><span class="va">gs_ovlap</span>, <span class="va">gsc</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="va">gs_stats</span> <span class="op">&lt;-</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="va">fry_out</span><span class="op">[</span><span class="va">gs_sig_name</span>,<span class="op">]</span><span class="op">$</span><span class="va">FDR</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">gs_stats</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">gs_sig_name</span></span>
<span>  </span>
<span>  <span class="co">#identify clusters</span></span>
<span>  <span class="va">grps</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/igraph/man/cluster_walktrap.html" class="external-link">cluster_walktrap</a></span><span class="op">(</span><span class="va">gs_ovnet</span><span class="op">)</span></span>
<span>  <span class="co">#extract clustering results</span></span>
<span>  <span class="va">grps</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/igraph/man/groups.html" class="external-link">groups</a></span><span class="op">(</span><span class="va">grps</span><span class="op">)</span></span>
<span>  <span class="co">#sort by cluster size</span></span>
<span>  <span class="va">grps</span> <span class="op">=</span> <span class="va">grps</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/order.html" class="external-link">order</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">grps</span>, <span class="va">length</span><span class="op">)</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">]</span></span>
<span>  </span>
<span>  <span class="co"># write output</span></span>
<span>  <span class="va">output_clusters</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span>  <span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">grps</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="va">output_clusters</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>geneset <span class="op">=</span> <span class="va">grps</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>, cluster <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"cluster"</span>,<span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">grps</span><span class="op">)</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span>  <span class="va">output_clusters</span> <span class="op">&lt;&lt;-</span> <span class="va">output_clusters</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind.html" class="external-link">bind_rows</a></span><span class="op">(</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="kw">if</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">specific_clusters</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="va">grps</span> <span class="op">&lt;-</span> <span class="va">grps</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="va">topN</span><span class="op">]</span></span>
<span>  <span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>    <span class="va">grps</span> <span class="op">&lt;-</span> <span class="va">grps</span><span class="op">[</span><span class="va">specific_clusters</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="op">)</span><span class="op">]</span></span>
<span>  <span class="op">}</span></span>
<span>  </span>
<span>  <span class="co">#plot the top 12 clusters</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">36</span><span class="op">)</span> <span class="co">#set seed for reproducible layout</span></span>
<span>  <span class="va">p1</span> <span class="op">&lt;&lt;-</span> <span class="fu"><a href="https://DavisLaboratory.github.io/vissE/reference/plotMsigNetwork.html" class="external-link">plotMsigNetwork</a></span><span class="op">(</span><span class="va">gs_ovnet</span>, markGroups <span class="op">=</span> <span class="va">grps</span>, </span>
<span>                        genesetStat <span class="op">=</span> <span class="va">gs_stats</span>, rmUnmarkedGroups <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">scico</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scico/man/ggplot2-scales.html" class="external-link">scale_fill_scico</a></span><span class="op">(</span>name <span class="op">=</span> <span class="st">"-log10(FDR)"</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="va">p2</span> <span class="op">&lt;&lt;-</span> <span class="fu"><a href="https://DavisLaboratory.github.io/vissE/reference/plotMsigWordcloud.html" class="external-link">plotMsigWordcloud</a></span><span class="op">(</span><span class="va">gsc</span>, <span class="va">grps</span>, type <span class="op">=</span> <span class="st">'Name'</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/GSEABase/man/GeneSet-class.html" class="external-link">geneIds</a></span><span class="op">(</span><span class="va">gsc_sig</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="va">genes_logfc</span> <span class="op">&lt;-</span> <span class="va">de_table</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html" class="external-link">rownames_to_column</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">rowname</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="va">genes</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="va">.</span><span class="op">$</span><span class="va">logFC</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">genes_logfc</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">de_table</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html" class="external-link">rownames_to_column</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">rowname</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="va">genes</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="va">.</span><span class="op">$</span><span class="va">rowname</span></span>
<span>  </span>
<span>  <span class="va">p3</span> <span class="op">&lt;&lt;-</span> <span class="fu"><a href="https://DavisLaboratory.github.io/vissE/reference/plotGeneStats.html" class="external-link">plotGeneStats</a></span><span class="op">(</span><span class="va">genes_logfc</span>, <span class="va">gsc</span>, <span class="va">grps</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="fl">0</span>, colour <span class="op">=</span> <span class="fl">2</span>, lty <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"logFC"</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="co">#p4 &lt;- plotMsigPPI(ppi, gsc, grps[1:topN], geneStat = genes_logfc) +</span></span>
<span>  <span class="co">#  guides(col=guide_legend(title="logFC"))</span></span>
<span>  </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">p2</span> <span class="op">+</span> <span class="va">p1</span> <span class="op">+</span> <span class="va">p3</span> <span class="op">+</span> <span class="fu">patchwork</span><span class="fu">::</span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html" class="external-link">plot_layout</a></span><span class="op">(</span>ncol <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="op">+</span></span>
<span>          <span class="fu">patchwork</span><span class="fu">::</span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_annotation.html" class="external-link">plot_annotation</a></span><span class="op">(</span>title <span class="op">=</span> <span class="va">title</span><span class="op">)</span><span class="op">)</span>  </span>
<span>    </span>
<span><span class="op">}</span></span></code></pre></div>
<div class="sourceCode" id="cb72"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">dovissE</span><span class="op">(</span><span class="va">fry_res_sig</span>, <span class="va">de_genes_toptable_DvCinGab</span>, title <span class="op">=</span> <span class="st">"DKD vs Normal in abnormal glomerulus"</span> <span class="op">)</span></span></code></pre></div>
<p><img src="GeoMXAnalysisWorkflow_files/figure-html/unnamed-chunk-51-1.png" width="1440"></p>
</div>
</div>
<div class="section level2">
<h2 id="summary">Summary<a class="anchor" aria-label="anchor" href="#summary"></a>
</h2>
<p>To write.</p>
</div>
<div class="section level2">
<h2 class="unnumbered" id="packages-used">Packages used<a class="anchor" aria-label="anchor" href="#packages-used"></a>
</h2>
<p>This workflow depends on various packages from version 3.16 of the
Bioconductor project, running on R version 4.2.0 (2022-04-22) or higher.
The complete list of the packages used for this workflow are shown
below:</p>
<div class="sourceCode" id="cb73"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.2.0 (2022-04-22)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">## Running under: Ubuntu 20.04.4 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3</span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/liblapack.so.3</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] igraph_1.3.3                ggrepel_0.9.1              </span></span>
<span><span class="co">##  [3] ExperimentHub_2.5.0         AnnotationHub_3.5.0        </span></span>
<span><span class="co">##  [5] BiocFileCache_2.5.0         dbplyr_2.2.1               </span></span>
<span><span class="co">##  [7] ggalluvial_0.12.3           msigdb_1.5.0               </span></span>
<span><span class="co">##  [9] GSEABase_1.59.0             graph_1.75.0               </span></span>
<span><span class="co">## [11] annotate_1.75.0             XML_3.99-0.10              </span></span>
<span><span class="co">## [13] AnnotationDbi_1.59.1        vissE_1.5.0                </span></span>
<span><span class="co">## [15] forcats_0.5.1               stringr_1.4.0              </span></span>
<span><span class="co">## [17] dplyr_1.0.9                 purrr_0.3.4                </span></span>
<span><span class="co">## [19] readr_2.1.2                 tidyr_1.2.0                </span></span>
<span><span class="co">## [21] tibble_3.1.7                ggplot2_3.3.6              </span></span>
<span><span class="co">## [23] tidyverse_1.3.1             edgeR_3.39.3               </span></span>
<span><span class="co">## [25] limma_3.53.4                SpatialExperiment_1.7.0    </span></span>
<span><span class="co">## [27] SingleCellExperiment_1.19.0 SummarizedExperiment_1.27.1</span></span>
<span><span class="co">## [29] Biobase_2.57.1              GenomicRanges_1.49.0       </span></span>
<span><span class="co">## [31] GenomeInfoDb_1.33.3         IRanges_2.31.0             </span></span>
<span><span class="co">## [33] S4Vectors_0.35.1            BiocGenerics_0.43.0        </span></span>
<span><span class="co">## [35] MatrixGenerics_1.9.1        matrixStats_0.62.0         </span></span>
<span><span class="co">## [37] standR_1.1.0               </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] rappdirs_0.3.3                R.methodsS3_1.8.2            </span></span>
<span><span class="co">##   [3] ragg_1.2.2                    bit64_4.0.5                  </span></span>
<span><span class="co">##   [5] knitr_1.39                    irlba_2.3.5                  </span></span>
<span><span class="co">##   [7] DelayedArray_0.23.0           R.utils_2.12.0               </span></span>
<span><span class="co">##   [9] data.table_1.14.2             KEGGREST_1.37.3              </span></span>
<span><span class="co">##  [11] RCurl_1.98-1.7                generics_0.1.3               </span></span>
<span><span class="co">##  [13] ScaledMatrix_1.5.0            qdapRegex_0.7.5              </span></span>
<span><span class="co">##  [15] cowplot_1.1.1                 RSQLite_2.2.15               </span></span>
<span><span class="co">##  [17] bit_4.0.4                     tzdb_0.3.0                   </span></span>
<span><span class="co">##  [19] textstem_0.1.4                xml2_1.3.3                   </span></span>
<span><span class="co">##  [21] lubridate_1.8.0               httpuv_1.6.5                 </span></span>
<span><span class="co">##  [23] assertthat_0.2.1              viridis_0.6.2                </span></span>
<span><span class="co">##  [25] xfun_0.31                     hms_1.1.1                    </span></span>
<span><span class="co">##  [27] jquerylib_0.1.4               evaluate_0.15                </span></span>
<span><span class="co">##  [29] promises_1.2.0.1              fansi_1.0.3                  </span></span>
<span><span class="co">##  [31] readxl_1.4.0                  DBI_1.1.3                    </span></span>
<span><span class="co">##  [33] htmlwidgets_1.5.4             ellipsis_0.3.2               </span></span>
<span><span class="co">##  [35] RSpectra_0.16-1               crosstalk_1.2.0              </span></span>
<span><span class="co">##  [37] ggpubr_0.4.0                  backports_1.4.1              </span></span>
<span><span class="co">##  [39] sparseMatrixStats_1.9.0       vctrs_0.4.1                  </span></span>
<span><span class="co">##  [41] ggwordcloud_0.5.0             prettydoc_0.4.1              </span></span>
<span><span class="co">##  [43] abind_1.4-5                   cachem_1.0.6                 </span></span>
<span><span class="co">##  [45] withr_2.5.0                   ggforce_0.3.3                </span></span>
<span><span class="co">##  [47] vroom_1.5.7                   cluster_2.1.3                </span></span>
<span><span class="co">##  [49] crayon_1.5.1                  koRpus_0.13-8                </span></span>
<span><span class="co">##  [51] slam_0.1-50                   pkgconfig_2.0.3              </span></span>
<span><span class="co">##  [53] labeling_0.4.2                sylly_0.1-6                  </span></span>
<span><span class="co">##  [55] tweenr_1.0.2                  nlme_3.1-158                 </span></span>
<span><span class="co">##  [57] vipor_0.4.5                   rlang_1.0.4                  </span></span>
<span><span class="co">##  [59] lifecycle_1.0.1               textshape_1.7.3              </span></span>
<span><span class="co">##  [61] filelock_1.0.2                lexicon_1.2.1                </span></span>
<span><span class="co">##  [63] modelr_0.1.8                  rsvd_1.0.5                   </span></span>
<span><span class="co">##  [65] cellranger_1.1.0              rprojroot_2.0.3              </span></span>
<span><span class="co">##  [67] polyclip_1.10-0               Matrix_1.4-1                 </span></span>
<span><span class="co">##  [69] carData_3.0-5                 Rhdf5lib_1.19.2              </span></span>
<span><span class="co">##  [71] reprex_2.0.1                  beeswarm_0.4.0               </span></span>
<span><span class="co">##  [73] png_0.1-7                     viridisLite_0.4.0            </span></span>
<span><span class="co">##  [75] rjson_0.2.21                  bitops_1.0-7                 </span></span>
<span><span class="co">##  [77] R.oo_1.25.0                   rhdf5filters_1.9.0           </span></span>
<span><span class="co">##  [79] Biostrings_2.65.1             blob_1.2.3                   </span></span>
<span><span class="co">##  [81] DelayedMatrixStats_1.19.0     mclustcomp_0.3.3             </span></span>
<span><span class="co">##  [83] rstatix_0.7.0                 ggsignif_0.6.3               </span></span>
<span><span class="co">##  [85] beachmat_2.13.4               scales_1.2.0                 </span></span>
<span><span class="co">##  [87] memoise_2.0.1                 magrittr_2.0.3               </span></span>
<span><span class="co">##  [89] plyr_1.8.7                    zlibbioc_1.43.0              </span></span>
<span><span class="co">##  [91] compiler_4.2.0                dqrng_0.3.0                  </span></span>
<span><span class="co">##  [93] RColorBrewer_1.1-3            cli_3.3.0                    </span></span>
<span><span class="co">##  [95] XVector_0.37.0                sylly.en_0.1-3               </span></span>
<span><span class="co">##  [97] patchwork_1.1.1               MASS_7.3-57                  </span></span>
<span><span class="co">##  [99] mgcv_1.8-40                   tidyselect_1.1.2             </span></span>
<span><span class="co">## [101] stringi_1.7.8                 textshaping_0.3.6            </span></span>
<span><span class="co">## [103] highr_0.9                     koRpus.lang.en_0.1-4         </span></span>
<span><span class="co">## [105] yaml_2.3.5                    BiocSingular_1.13.0          </span></span>
<span><span class="co">## [107] locfit_1.5-9.6                grid_4.2.0                   </span></span>
<span><span class="co">## [109] sass_0.4.2                    tools_4.2.0                  </span></span>
<span><span class="co">## [111] ruv_0.9.7.1                   parallel_4.2.0               </span></span>
<span><span class="co">## [113] rstudioapi_0.13               gridExtra_2.3                </span></span>
<span><span class="co">## [115] farver_2.1.1                  ggraph_2.0.5                 </span></span>
<span><span class="co">## [117] DropletUtils_1.17.0           digest_0.6.29                </span></span>
<span><span class="co">## [119] BiocManager_1.30.18           FNN_1.1.3.1                  </span></span>
<span><span class="co">## [121] shiny_1.7.1                   Rcpp_1.0.9                   </span></span>
<span><span class="co">## [123] car_3.1-0                     broom_1.0.0                  </span></span>
<span><span class="co">## [125] scuttle_1.7.2                 BiocVersion_3.16.0           </span></span>
<span><span class="co">## [127] later_1.3.0                   org.Hs.eg.db_3.15.0          </span></span>
<span><span class="co">## [129] httr_1.4.3                    Rdpack_2.3.1                 </span></span>
<span><span class="co">## [131] colorspace_2.0-3              rvest_1.0.2                  </span></span>
<span><span class="co">## [133] fs_1.5.2                      splines_4.2.0                </span></span>
<span><span class="co">## [135] uwot_0.1.11                   statmod_1.4.36               </span></span>
<span><span class="co">## [137] pkgdown_2.0.6                 scater_1.25.1                </span></span>
<span><span class="co">## [139] graphlayouts_0.8.0            systemfonts_1.0.4            </span></span>
<span><span class="co">## [141] xtable_1.8-4                  jsonlite_1.8.0               </span></span>
<span><span class="co">## [143] tidygraph_1.2.1               NLP_0.2-1                    </span></span>
<span><span class="co">## [145] scico_1.3.0                   R6_2.5.1                     </span></span>
<span><span class="co">## [147] tm_0.7-8                      pillar_1.7.0                 </span></span>
<span><span class="co">## [149] htmltools_0.5.2               mime_0.12                    </span></span>
<span><span class="co">## [151] glue_1.6.2                    fastmap_1.1.0                </span></span>
<span><span class="co">## [153] DT_0.23                       BiocParallel_1.31.10         </span></span>
<span><span class="co">## [155] BiocNeighbors_1.15.1          interactiveDisplayBase_1.35.0</span></span>
<span><span class="co">## [157] codetools_0.2-18              utf8_1.2.2                   </span></span>
<span><span class="co">## [159] lattice_0.20-45               bslib_0.4.0                  </span></span>
<span><span class="co">## [161] curl_4.3.2                    ggbeeswarm_0.6.0             </span></span>
<span><span class="co">## [163] magick_2.7.3                  textclean_0.9.3              </span></span>
<span><span class="co">## [165] syuzhet_1.0.6                 rmarkdown_2.14               </span></span>
<span><span class="co">## [167] desc_1.4.1                    munsell_0.5.0                </span></span>
<span><span class="co">## [169] rhdf5_2.41.1                  GenomeInfoDbData_1.2.8       </span></span>
<span><span class="co">## [171] HDF5Array_1.25.1              haven_2.5.0                  </span></span>
<span><span class="co">## [173] reshape2_1.4.4                gtable_0.3.0                 </span></span>
<span><span class="co">## [175] rbibutils_2.2.8</span></span></code></pre>
</div>
<div class="section level2">
<h2 class="unnumbered" id="acknowledgments">Acknowledgments<a class="anchor" aria-label="anchor" href="#acknowledgments"></a>
</h2>
<p>To write.</p>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-R-rmarkdown" class="csl-entry">
Allaire, JJ, Yihui Xie, Jonathan McPherson, Javier Luraschi, Kevin
Ushey, Aron Atkins, Hadley Wickham, Joe Cheng, Winston Chang, and
Richard Iannone. 2022. <em>Rmarkdown: Dynamic Documents for r</em>. <a href="https://CRAN.R-project.org/package=rmarkdown" class="external-link">https://CRAN.R-project.org/package=rmarkdown</a>.
</div>
<div id="ref-R-prettydoc" class="csl-entry">
Qiu, Yixuan. 2021. <em>Prettydoc: Creating Pretty Documents from r
Markdown</em>. <a href="https://github.com/yixuan/prettydoc" class="external-link">https://github.com/yixuan/prettydoc</a>.
</div>
<div id="ref-R-ggplot2" class="csl-entry">
Wickham, Hadley, Winston Chang, Lionel Henry, Thomas Lin Pedersen,
Kohske Takahashi, Claus Wilke, Kara Woo, Hiroaki Yutani, and Dewey
Dunnington. 2021. <em>Ggplot2: Create Elegant Data Visualisations Using
the Grammar of Graphics</em>. <a href="https://CRAN.R-project.org/package=ggplot2" class="external-link">https://CRAN.R-project.org/package=ggplot2</a>.
</div>
<div id="ref-R-knitr" class="csl-entry">
Xie, Yihui. 2022. <em>Knitr: A General-Purpose Package for Dynamic
Report Generation in r</em>. <a href="https://yihui.org/knitr/" class="external-link">https://yihui.org/knitr/</a>.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Ning Liu, Chin Wee Tan, Melissa J. Davis.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
