<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Analysing Nanostring's GeoMX transcriptomics data using standR, limma and vissE • GeoMXAnalysisWorkflow</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Analysing Nanostring's GeoMX transcriptomics data using standR, limma and vissE">
<meta property="og:description" content="GeoMXAnalysisWorkflow">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">GeoMXAnalysisWorkflow</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/GeoMXAnalysisWorkflow.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/DavisLaboratory/GeoMXAnalysisWorkflow" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="GeoMXAnalysisWorkflow_files/htmlwidgets-1.5.4/htmlwidgets.js"></script><link href="GeoMXAnalysisWorkflow_files/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="GeoMXAnalysisWorkflow_files/datatables-binding-0.23/datatables.js"></script><link href="GeoMXAnalysisWorkflow_files/dt-core-1.11.3/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="GeoMXAnalysisWorkflow_files/dt-core-1.11.3/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="GeoMXAnalysisWorkflow_files/dt-core-1.11.3/js/jquery.dataTables.min.js"></script><link href="GeoMXAnalysisWorkflow_files/crosstalk-1.2.0/css/crosstalk.min.css" rel="stylesheet">
<script src="GeoMXAnalysisWorkflow_files/crosstalk-1.2.0/js/crosstalk.min.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Analysing Nanostring’s GeoMX transcriptomics
data using standR, limma and vissE</h1>
                        <h4 data-toc-skip class="author">Ning Liu</h4>
            <address class="author_afil">
      Bioinformatics Division, Walter and Eliza Hall Institute of
Medical Research, Parkville, VIC 3052, AustraliaDepartment of Medical
Biology, University of Melbourne, Parkville, VIC 3010,
Australia<br><a class="author_email" href="mailto:#"></a><a href="mailto:liu.n@wehi.edu.au" class="email">liu.n@wehi.edu.au</a>
      </address>
                              <h4 data-toc-skip class="author">Chin Wee
Tan</h4>
            <address class="author_afil">
      Bioinformatics Division, Walter and Eliza Hall Institute of
Medical Research, Parkville, VIC 3052, AustraliaDepartment of Medical
Biology, University of Melbourne, Parkville, VIC 3010,
Australia<br><a class="author_email" href="mailto:#"></a><a href="mailto:cwtan@wehi.edu.au" class="email">cwtan@wehi.edu.au</a>
      </address>
                              <h4 data-toc-skip class="author">Melissa J
Davis</h4>
            <address class="author_afil">
      Bioinformatics Division, Walter and Eliza Hall Institute of
Medical Research, Parkville, VIC 3052, AustraliaDepartment of Medical
Biology, University of Melbourne, Parkville, VIC 3010,
AustraliaDepartment of Biochemistry and Molecular Biology, Faculty of
Medicine, Dentistry and Health Sciences, University of Melbourne,
Parkville, VIC, 3010,
Australia<br><a class="author_email" href="mailto:#"></a><a href="mailto:davis.m@wehi.edu.au" class="email">davis.m@wehi.edu.au</a>
      </address>
                  
            <h4 data-toc-skip class="date">Jun 2022</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/DavisLaboratory/GeoMXAnalysisWorkflow/blob/HEAD/vignettes/GeoMXAnalysisWorkflow.Rmd" class="external-link"><code>vignettes/GeoMXAnalysisWorkflow.Rmd</code></a></small>
      <div class="hidden name"><code>GeoMXAnalysisWorkflow.Rmd</code></div>

    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>Abstract</p>
    </div>
    
<p>
</p>
<p><strong>R version</strong>: R version 4.2.0 (2022-04-22) <br><strong>Bioconductor version</strong>: 3.16</p>

<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>A bioinformatics analysis of a GeoMX DSP dataset often starts with a
count table (genes by regions of interest (ROIs)), and ends by either
identifying differential expressed genes or performing gene
signature/gene-set scoring analysis in various conditions or
experimental designs. Nevertheless, before performing the differential
expression (DE) analysis or other downstream analysis that based on the
gene counts, proper quality control and normalisation of the data is
essential and can largely impact the validity of the DE and downstream
analysis’ results. We therefore developed a bioconductor package called
<code>standR</code> (<strong>S</strong>patial
<strong>t</strong>ranscriptomics <strong>a</strong>nalyzes
a<strong>n</strong>d <strong>d</strong>ecoding in <strong>R</strong>) to
assist the QC, normalisation and batch correction of the GeoMX
transcriptomics data.</p>
<p>There are three major advantages of using the <code>standR</code>
package to analyse the GeoMX DSP datasets:</p>
<ol style="list-style-type: decimal">
<li><p>The package uses the <code>SpatialExperiment</code>
infrastructure to analyse the data, this infrastructure is a lineage of
the <code>SummarisedExperiment</code> family, which is highly
recommended in the bioconductor community and compatible and
transferable with many other highly developed packages in the RNA-seq
analysis world, such as <code>scater</code>, <code>scran</code>,
<code>edgeR</code> and <code>limma</code>.</p></li>
<li><p>The package features a comprehensive route of quality control,
and provides various visualisation functions to help assessing different
quality control metrics.</p></li>
<li><p>Batch effect is commonly seen in transcriptome dataset,
especially in GeoMX DSP data due to the experimental design of slide
usage, the package currently provides three batch correction methods
that can remove the unwanted batch effect and provides statistics to
assess the correction process.</p></li>
</ol>
<p>In this workshop, we first use the <code>standR</code> to process a
published GeoMX WTA dataset in a workflow fashion. This will demonstrate
our recommended flow to process and analyse the GeoMX transcriptomics
data. Then, we will perform the DE analysis of the processed data using
the <code>limma-voom</code> pipeline, followed by a gene-set enrichment
analysis using <code>fry</code> and visualisation of the results using
<code>vissE</code>.</p>
</div>
<div class="section level2">
<h2 id="processing-and-analysing-geomx-transcriptomics-data-with-standr">Processing and analysing GeoMX transcriptomics data with standR<a class="anchor" aria-label="anchor" href="#processing-and-analysing-geomx-transcriptomics-data-with-standr"></a>
</h2>
<div class="section level3">
<h3 id="load-data">Load data<a class="anchor" aria-label="anchor" href="#load-data"></a>
</h3>
<p>Data we are using in this workshop is a published GeoMX whole
transcriptome atlas (WTA) dataset of diabetic kidney disease (DKD).</p>
<p>In this dataset, <strong>Seven slides</strong> were analyzed,
including <strong>4 DKD and 3 healthy</strong> slides. ROIs were focused
on two different parts of a kidney’s structure: <strong>tubules or
glomeruli</strong>. Individual glomeruli were identified by a
pathologist as either <strong>relatively healthy or diseased</strong>
regardless if the tissue was DKD or healthy. Tubule ROIs were segmented
into <strong>distal (PanCK) and proximal (neg) tubules</strong>. While
both distal and proximal tubules are called tubules, they perform very
different functions in the kidney.</p>
<p>This GeoMX DSP data can be retireved by querying the
<code>ExperimentHub</code>. The query below searches for all objects in
the hub associated with the search term “standR”. Here, Three files that
are associated with the DKD GeoMX data in the <code>standR</code>
package are retrieved.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">ExperimentHub</span><span class="op">)</span></span>
<span></span>
<span><span class="va">eh</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ExperimentHub/man/ExperimentHub-class.html" class="external-link">ExperimentHub</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/AnnotationHub/man/AnnotationHub-class.html" class="external-link">query</a></span><span class="op">(</span><span class="va">eh</span>, <span class="st">"standR"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## ExperimentHub with 3 records</span></span>
<span><span class="co">## # snapshotDate(): 2022-06-15</span></span>
<span><span class="co">## # $dataprovider: Nanostring</span></span>
<span><span class="co">## # $species: NA</span></span>
<span><span class="co">## # $rdataclass: data.frame</span></span>
<span><span class="co">## # additional mcols(): taxonomyid, genome, description,</span></span>
<span><span class="co">## #   coordinate_1_based, maintainer, rdatadateadded, preparerclass, tags,</span></span>
<span><span class="co">## #   rdatapath, sourceurl, sourcetype </span></span>
<span><span class="co">## # retrieve records with, e.g., 'object[["EH7364"]]' </span></span>
<span><span class="co">## </span></span>
<span><span class="co">##            title                   </span></span>
<span><span class="co">##   EH7364 | GeomxDKDdata_count      </span></span>
<span><span class="co">##   EH7365 | GeomxDKDdata_sampleAnno </span></span>
<span><span class="co">##   EH7366 | GeomxDKDdata_featureAnno</span></span></code></pre>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">countFile</span> <span class="op">&lt;-</span> <span class="va">eh</span><span class="op">[[</span><span class="st">"EH7364"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">sampleAnnoFile</span> <span class="op">&lt;-</span> <span class="va">eh</span><span class="op">[[</span><span class="st">"EH7365"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">featureAnnoFile</span> <span class="op">&lt;-</span> <span class="va">eh</span><span class="op">[[</span><span class="st">"EH7366"</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<p>We can have a first glance at the format of the three files, they are
the usual files we can obtain from the Nanostring.</p>
<p>The <code>countFile</code> is a tab-delimited file, it contains the
count table (features by samples) we generally see in transcriptomics
analysis.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.delim</a></span><span class="op">(</span><span class="va">countFile</span>, nrows <span class="op">=</span> <span class="fl">5</span><span class="op">)</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##   TargetName disease3_scan...001...PanCK disease3_scan...001...neg</span></span>
<span><span class="co">## 1      PADI2                          15                        35</span></span>
<span><span class="co">## 2    CYP24A1                          23                        32</span></span>
<span><span class="co">## 3    SUPT16H                          35                        66</span></span>
<span><span class="co">## 4      ZMIZ2                          51                        81</span></span>
<span><span class="co">## 5      SPAG9                          48                        65</span></span></code></pre>
<p>The <code>sampleAnnoFile</code> is a tab-delimited file, contains all
the metadata of the samples.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.delim</a></span><span class="op">(</span><span class="va">sampleAnnoFile</span>, nrows <span class="op">=</span> <span class="fl">5</span><span class="op">)</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##   SlideName      ScanName ROILabel      SegmentLabel</span></span>
<span><span class="co">## 1  disease3 disease3_scan        7 Geometric Segment</span></span>
<span><span class="co">## 2  disease3 disease3_scan        8 Geometric Segment</span></span>
<span><span class="co">## 3  disease3 disease3_scan        9 Geometric Segment</span></span>
<span><span class="co">## 4  disease3 disease3_scan       10 Geometric Segment</span></span>
<span><span class="co">## 5  disease3 disease3_scan       11 Geometric Segment</span></span>
<span><span class="co">##                        SegmentDisplayName</span></span>
<span><span class="co">## 1 disease3_scan | 007 | Geometric Segment</span></span>
<span><span class="co">## 2 disease3_scan | 008 | Geometric Segment</span></span>
<span><span class="co">## 3 disease3_scan | 009 | Geometric Segment</span></span>
<span><span class="co">## 4 disease3_scan | 010 | Geometric Segment</span></span>
<span><span class="co">## 5 disease3_scan | 011 | Geometric Segment</span></span></code></pre>
<p>The <code>featureAnnoFile</code> is a tab-delimited file, contains
all the metadata of the genes.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.delim</a></span><span class="op">(</span><span class="va">featureAnnoFile</span>, nrows <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##       RTS_ID TargetName ProbeID Negative</span></span>
<span><span class="co">## 1 RTS0020877        A2M   35614    FALSE</span></span>
<span><span class="co">## 2 RTS0020878       NAT2   35615    FALSE</span></span>
<span><span class="co">## 3 RTS0020879      ACADM   35616    FALSE</span></span>
<span><span class="co">## 4 RTS0020880      ACADS   35617    FALSE</span></span>
<span><span class="co">## 5 RTS0020881      ACAT1   35618    FALSE</span></span></code></pre>
<p>As we described in the introduction, there are many advantages of
using a mature infrastructure throughout out the analysis, such as being
compatible with many other tools.</p>
<p>Hence, the first step in the <code>standR</code> package is to
construct a <code>SpatialExperiment</code> object that includes all the
information about the data. Here we can use the function
<code>readGeoMX</code> to do so. For more information about the
<code>SpatialExperiment</code> infrastructure, see <a href="http://www.bioconductor.org/packages/release/bioc/html/SpatialExperiment.html" class="external-link">here</a>.</p>
<p><strong>Notes: if you plan to use <code>readGeoMx</code> to construct
your <code>SpatialExperiment</code> object with your own data, make sure
that the files you use as inputs are tab-delimited files.</strong></p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/DavisLaboratory/standR" class="external-link">standR</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">spe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/standR/man/readGeoMx.html" class="external-link">readGeoMx</a></span><span class="op">(</span><span class="va">countFile</span>, <span class="va">sampleAnnoFile</span>, featureAnnoFile <span class="op">=</span> <span class="va">featureAnnoFile</span><span class="op">)</span></span></code></pre></div>
<p>We can check out some basic informations about the dataset by typing
the object directly. We can see that the dataset have 18503 genes and
231 ROIs.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe</span></span></code></pre></div>
<pre><code><span><span class="co">## class: SpatialExperiment </span></span>
<span><span class="co">## dim: 18503 231 </span></span>
<span><span class="co">## metadata(1): NegProbes</span></span>
<span><span class="co">## assays(2): counts logcounts</span></span>
<span><span class="co">## rownames(18503): PADI2 CYP24A1 ... FAM166A AGTPBP1</span></span>
<span><span class="co">## rowData names(3): RTS_ID ProbeID Negative</span></span>
<span><span class="co">## colnames(231): disease3_scan | 001 | PanCK disease3_scan | 001 | neg</span></span>
<span><span class="co">##   ... disease1B_scan | 023 | Geometric Segment disease1B_scan | 024 |</span></span>
<span><span class="co">##   Geometric Segment</span></span>
<span><span class="co">## colData names(25): SlideName ScanName ... RoiReportY sample_id</span></span>
<span><span class="co">## reducedDimNames(0):</span></span>
<span><span class="co">## mainExpName: NULL</span></span>
<span><span class="co">## altExpNames(0):</span></span>
<span><span class="co">## spatialCoords names(2) : ROICoordinateX ROICoordinateY</span></span>
<span><span class="co">## imgData names(0):</span></span></code></pre>
<p>In the <code>spatialExperiment</code> object, the raw count data is
stored in the <code>counts</code> assay slot, while the log-CPM (count
per million) of the data is calculated by default with the
<code>readGeoMX</code> function and stored in the <code>logcounts</code>
assay of the object.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">assayNames</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "counts"    "logcounts"</span></span></code></pre>
<p>We can have a look at the count table by using the <code>assay</code>
function and specify the table name.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">assay</span><span class="op">(</span><span class="va">spe</span>, <span class="st">"counts"</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##         disease3_scan | 001 | PanCK disease3_scan | 001 | neg</span></span>
<span><span class="co">## PADI2                            15                        35</span></span>
<span><span class="co">## CYP24A1                          23                        32</span></span>
<span><span class="co">## SUPT16H                          35                        66</span></span>
<span><span class="co">## ZMIZ2                            51                        81</span></span>
<span><span class="co">## SPAG9                            48                        65</span></span>
<span><span class="co">##         disease3_scan | 002 | PanCK disease3_scan | 002 | neg</span></span>
<span><span class="co">## PADI2                            36                        41</span></span>
<span><span class="co">## CYP24A1                          23                        41</span></span>
<span><span class="co">## SUPT16H                          46                        61</span></span>
<span><span class="co">## ZMIZ2                            87                       113</span></span>
<span><span class="co">## SPAG9                            46                        91</span></span>
<span><span class="co">##         disease3_scan | 003 | PanCK</span></span>
<span><span class="co">## PADI2                             9</span></span>
<span><span class="co">## CYP24A1                           3</span></span>
<span><span class="co">## SUPT16H                          20</span></span>
<span><span class="co">## ZMIZ2                            68</span></span>
<span><span class="co">## SPAG9                            25</span></span></code></pre>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">assay</span><span class="op">(</span><span class="va">spe</span>, <span class="st">"logcounts"</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##         disease3_scan | 001 | PanCK disease3_scan | 001 | neg</span></span>
<span><span class="co">## PADI2                      4.694722                  5.097130</span></span>
<span><span class="co">## CYP24A1                    5.205348                  4.988548</span></span>
<span><span class="co">## SUPT16H                    5.738537                  5.903808</span></span>
<span><span class="co">## ZMIZ2                      6.236242                  6.175525</span></span>
<span><span class="co">## SPAG9                      6.155076                  5.883737</span></span>
<span><span class="co">##         disease3_scan | 002 | PanCK disease3_scan | 002 | neg</span></span>
<span><span class="co">## PADI2                      5.507768                  5.054985</span></span>
<span><span class="co">## CYP24A1                    4.952999                  5.054985</span></span>
<span><span class="co">## SUPT16H                    5.824568                  5.551068</span></span>
<span><span class="co">## ZMIZ2                      6.679218                  6.363792</span></span>
<span><span class="co">## SPAG9                      5.824568                  6.073582</span></span>
<span><span class="co">##         disease3_scan | 003 | PanCK</span></span>
<span><span class="co">## PADI2                      4.573874</span></span>
<span><span class="co">## CYP24A1                    3.518078</span></span>
<span><span class="co">## SUPT16H                    5.538414</span></span>
<span><span class="co">## ZMIZ2                      7.183436</span></span>
<span><span class="co">## SPAG9                      5.827206</span></span></code></pre>
<p>Sample metadata is stored in the <code>colData</code> of the
object.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">colData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 231 rows and 25 columns</span></span>
<span><span class="co">##                                            SlideName       ScanName  ROILabel</span></span>
<span><span class="co">##                                          &lt;character&gt;    &lt;character&gt; &lt;numeric&gt;</span></span>
<span><span class="co">## disease3_scan | 001 | PanCK                 disease3  disease3_scan         1</span></span>
<span><span class="co">## disease3_scan | 001 | neg                   disease3  disease3_scan         1</span></span>
<span><span class="co">## disease3_scan | 002 | PanCK                 disease3  disease3_scan         2</span></span>
<span><span class="co">## disease3_scan | 002 | neg                   disease3  disease3_scan         2</span></span>
<span><span class="co">## disease3_scan | 003 | PanCK                 disease3  disease3_scan         3</span></span>
<span><span class="co">## ...                                              ...            ...       ...</span></span>
<span><span class="co">## disease1B_scan | 020 | Geometric Segment   disease1B disease1B_scan        20</span></span>
<span><span class="co">## disease1B_scan | 021 | Geometric Segment   disease1B disease1B_scan        21</span></span>
<span><span class="co">## disease1B_scan | 022 | Geometric Segment   disease1B disease1B_scan        22</span></span>
<span><span class="co">## disease1B_scan | 023 | Geometric Segment   disease1B disease1B_scan        23</span></span>
<span><span class="co">## disease1B_scan | 024 | Geometric Segment   disease1B disease1B_scan        24</span></span>
<span><span class="co">##                                               SegmentLabel</span></span>
<span><span class="co">##                                                &lt;character&gt;</span></span>
<span><span class="co">## disease3_scan | 001 | PanCK                          PanCK</span></span>
<span><span class="co">## disease3_scan | 001 | neg                              neg</span></span>
<span><span class="co">## disease3_scan | 002 | PanCK                          PanCK</span></span>
<span><span class="co">## disease3_scan | 002 | neg                              neg</span></span>
<span><span class="co">## disease3_scan | 003 | PanCK                          PanCK</span></span>
<span><span class="co">## ...                                                    ...</span></span>
<span><span class="co">## disease1B_scan | 020 | Geometric Segment Geometric Segment</span></span>
<span><span class="co">## disease1B_scan | 021 | Geometric Segment Geometric Segment</span></span>
<span><span class="co">## disease1B_scan | 022 | Geometric Segment Geometric Segment</span></span>
<span><span class="co">## disease1B_scan | 023 | Geometric Segment Geometric Segment</span></span>
<span><span class="co">## disease1B_scan | 024 | Geometric Segment Geometric Segment</span></span>
<span><span class="co">##                                                       Sample_ID AOISurfaceArea</span></span>
<span><span class="co">##                                                     &lt;character&gt;      &lt;numeric&gt;</span></span>
<span><span class="co">## disease3_scan | 001 | PanCK              DSP-1001250007864-D-..        22152.1</span></span>
<span><span class="co">## disease3_scan | 001 | neg                DSP-1001250007864-D-..        71817.9</span></span>
<span><span class="co">## disease3_scan | 002 | PanCK              DSP-1001250007864-D-..        37316.7</span></span>
<span><span class="co">## disease3_scan | 002 | neg                DSP-1001250007864-D-..       102112.7</span></span>
<span><span class="co">## disease3_scan | 003 | PanCK              DSP-1001250007864-D-..        18398.7</span></span>
<span><span class="co">## ...                                                         ...            ...</span></span>
<span><span class="co">## disease1B_scan | 020 | Geometric Segment DSP-1001250007868-B-..        34871.4</span></span>
<span><span class="co">## disease1B_scan | 021 | Geometric Segment DSP-1001250007868-B-..        29638.5</span></span>
<span><span class="co">## disease1B_scan | 022 | Geometric Segment DSP-1001250007868-B-..        24203.5</span></span>
<span><span class="co">## disease1B_scan | 023 | Geometric Segment DSP-1001250007868-B-..        24850.1</span></span>
<span><span class="co">## disease1B_scan | 024 | Geometric Segment DSP-1001250007868-B-..        22985.6</span></span>
<span><span class="co">##                                          AOINucleiCount ROICoordinateX</span></span>
<span><span class="co">##                                               &lt;numeric&gt;      &lt;numeric&gt;</span></span>
<span><span class="co">## disease3_scan | 001 | PanCK                         183          23572</span></span>
<span><span class="co">## disease3_scan | 001 | neg                           421          23572</span></span>
<span><span class="co">## disease3_scan | 002 | PanCK                         323          22404</span></span>
<span><span class="co">## disease3_scan | 002 | neg                           565          22404</span></span>
<span><span class="co">## disease3_scan | 003 | PanCK                         162          26308</span></span>
<span><span class="co">## ...                                                 ...            ...</span></span>
<span><span class="co">## disease1B_scan | 020 | Geometric Segment            185          28871</span></span>
<span><span class="co">## disease1B_scan | 021 | Geometric Segment            167          28494</span></span>
<span><span class="co">## disease1B_scan | 022 | Geometric Segment            145          32218</span></span>
<span><span class="co">## disease1B_scan | 023 | Geometric Segment            162          32556</span></span>
<span><span class="co">## disease1B_scan | 024 | Geometric Segment            126          15685</span></span>
<span><span class="co">##                                          ROICoordinateY  RawReads TrimmedReads</span></span>
<span><span class="co">##                                               &lt;numeric&gt; &lt;numeric&gt;    &lt;numeric&gt;</span></span>
<span><span class="co">## disease3_scan | 001 | PanCK                      -39228   7485137      7423555</span></span>
<span><span class="co">## disease3_scan | 001 | neg                        -39228  11365209     11264938</span></span>
<span><span class="co">## disease3_scan | 002 | PanCK                      -47078   5764010      5716732</span></span>
<span><span class="co">## disease3_scan | 002 | neg                        -47078  12870214     12767450</span></span>
<span><span class="co">## disease3_scan | 003 | PanCK                      -51612   3411689      3382824</span></span>
<span><span class="co">## ...                                                 ...       ...          ...</span></span>
<span><span class="co">## disease1B_scan | 020 | Geometric Segment         -20331   3596971      3532718</span></span>
<span><span class="co">## disease1B_scan | 021 | Geometric Segment         -20693   6624280      6501597</span></span>
<span><span class="co">## disease1B_scan | 022 | Geometric Segment         -31068   1274498      1246856</span></span>
<span><span class="co">## disease1B_scan | 023 | Geometric Segment         -30116   1762715      1733661</span></span>
<span><span class="co">## disease1B_scan | 024 | Geometric Segment         -13025    811601       796263</span></span>
<span><span class="co">##                                          StitchedReads AlignedReads</span></span>
<span><span class="co">##                                              &lt;numeric&gt;    &lt;numeric&gt;</span></span>
<span><span class="co">## disease3_scan | 001 | PanCK                    7364647      6981003</span></span>
<span><span class="co">## disease3_scan | 001 | neg                     11163492     10486650</span></span>
<span><span class="co">## disease3_scan | 002 | PanCK                    5675008      5412997</span></span>
<span><span class="co">## disease3_scan | 002 | neg                     12672391     11929146</span></span>
<span><span class="co">## disease3_scan | 003 | PanCK                    3357777      3198844</span></span>
<span><span class="co">## ...                                                ...          ...</span></span>
<span><span class="co">## disease1B_scan | 020 | Geometric Segment       3480276      1969705</span></span>
<span><span class="co">## disease1B_scan | 021 | Geometric Segment       6392274      2316137</span></span>
<span><span class="co">## disease1B_scan | 022 | Geometric Segment       1225128      1160940</span></span>
<span><span class="co">## disease1B_scan | 023 | Geometric Segment       1708493      1624815</span></span>
<span><span class="co">## disease1B_scan | 024 | Geometric Segment        784340       741467</span></span>
<span><span class="co">##                                          DeduplicatedReads SequencingSaturation</span></span>
<span><span class="co">##                                                  &lt;numeric&gt;            &lt;numeric&gt;</span></span>
<span><span class="co">## disease3_scan | 001 | PanCK                         729971              90.2477</span></span>
<span><span class="co">## disease3_scan | 001 | neg                          1213754              89.3204</span></span>
<span><span class="co">## disease3_scan | 002 | PanCK                         898058              84.4196</span></span>
<span><span class="co">## disease3_scan | 002 | neg                          1471691              88.5651</span></span>
<span><span class="co">## disease3_scan | 003 | PanCK                         486670              85.7352</span></span>
<span><span class="co">## ...                                                    ...                  ...</span></span>
<span><span class="co">## disease1B_scan | 020 | Geometric Segment             39143              98.9118</span></span>
<span><span class="co">## disease1B_scan | 021 | Geometric Segment             49662              99.2503</span></span>
<span><span class="co">## disease1B_scan | 022 | Geometric Segment             24887              98.0473</span></span>
<span><span class="co">## disease1B_scan | 023 | Geometric Segment             30385              98.2762</span></span>
<span><span class="co">## disease1B_scan | 024 | Geometric Segment             19827              97.5571</span></span>
<span><span class="co">##                                             UMIQ30    RTSQ30 disease_status</span></span>
<span><span class="co">##                                          &lt;numeric&gt; &lt;numeric&gt;    &lt;character&gt;</span></span>
<span><span class="co">## disease3_scan | 001 | PanCK                 0.9944    0.9937            DKD</span></span>
<span><span class="co">## disease3_scan | 001 | neg                   0.9943    0.9935            DKD</span></span>
<span><span class="co">## disease3_scan | 002 | PanCK                 0.9943    0.9936            DKD</span></span>
<span><span class="co">## disease3_scan | 002 | neg                   0.9945    0.9938            DKD</span></span>
<span><span class="co">## disease3_scan | 003 | PanCK                 0.9943    0.9937            DKD</span></span>
<span><span class="co">## ...                                            ...       ...            ...</span></span>
<span><span class="co">## disease1B_scan | 020 | Geometric Segment    0.9870    0.9850            DKD</span></span>
<span><span class="co">## disease1B_scan | 021 | Geometric Segment    0.9871    0.9852            DKD</span></span>
<span><span class="co">## disease1B_scan | 022 | Geometric Segment    0.9856    0.9830            DKD</span></span>
<span><span class="co">## disease1B_scan | 023 | Geometric Segment    0.9879    0.9861            DKD</span></span>
<span><span class="co">## disease1B_scan | 024 | Geometric Segment    0.9872    0.9852            DKD</span></span>
<span><span class="co">##                                            pathology      region       LOQ</span></span>
<span><span class="co">##                                          &lt;character&gt; &lt;character&gt; &lt;numeric&gt;</span></span>
<span><span class="co">## disease3_scan | 001 | PanCK                       NA      tubule   36.2389</span></span>
<span><span class="co">## disease3_scan | 001 | neg                         NA      tubule   62.9229</span></span>
<span><span class="co">## disease3_scan | 002 | PanCK                       NA      tubule   42.6423</span></span>
<span><span class="co">## disease3_scan | 002 | neg                         NA      tubule   71.2849</span></span>
<span><span class="co">## disease3_scan | 003 | PanCK                       NA      tubule   19.6686</span></span>
<span><span class="co">## ...                                              ...         ...       ...</span></span>
<span><span class="co">## disease1B_scan | 020 | Geometric Segment    abnormal  glomerulus   4.93200</span></span>
<span><span class="co">## disease1B_scan | 021 | Geometric Segment    abnormal  glomerulus   5.70526</span></span>
<span><span class="co">## disease1B_scan | 022 | Geometric Segment    abnormal  glomerulus   2.36403</span></span>
<span><span class="co">## disease1B_scan | 023 | Geometric Segment    abnormal  glomerulus   2.82057</span></span>
<span><span class="co">## disease1B_scan | 024 | Geometric Segment    abnormal  glomerulus   2.28887</span></span>
<span><span class="co">##                                          NormalizationFactor RoiReportX</span></span>
<span><span class="co">##                                                    &lt;numeric&gt;  &lt;numeric&gt;</span></span>
<span><span class="co">## disease3_scan | 001 | PanCK                         0.400861       4236</span></span>
<span><span class="co">## disease3_scan | 001 | neg                           0.246243       4236</span></span>
<span><span class="co">## disease3_scan | 002 | PanCK                         0.325227       3945</span></span>
<span><span class="co">## disease3_scan | 002 | neg                           0.202788       3945</span></span>
<span><span class="co">## disease3_scan | 003 | PanCK                         0.574567       4910</span></span>
<span><span class="co">## ...                                                      ...        ...</span></span>
<span><span class="co">## disease1B_scan | 020 | Geometric Segment             5.74567       5496</span></span>
<span><span class="co">## disease1B_scan | 021 | Geometric Segment             4.30925       5403</span></span>
<span><span class="co">## disease1B_scan | 022 | Geometric Segment             8.61850       6336</span></span>
<span><span class="co">## disease1B_scan | 023 | Geometric Segment             8.61850       6418</span></span>
<span><span class="co">## disease1B_scan | 024 | Geometric Segment             8.61850       2202</span></span>
<span><span class="co">##                                          RoiReportY   sample_id</span></span>
<span><span class="co">##                                           &lt;numeric&gt; &lt;character&gt;</span></span>
<span><span class="co">## disease3_scan | 001 | PanCK                   -8628    sample01</span></span>
<span><span class="co">## disease3_scan | 001 | neg                     -8628    sample01</span></span>
<span><span class="co">## disease3_scan | 002 | PanCK                  -10603    sample01</span></span>
<span><span class="co">## disease3_scan | 002 | neg                    -10603    sample01</span></span>
<span><span class="co">## disease3_scan | 003 | PanCK                  -11731    sample01</span></span>
<span><span class="co">## ...                                             ...         ...</span></span>
<span><span class="co">## disease1B_scan | 020 | Geometric Segment      -3866    sample01</span></span>
<span><span class="co">## disease1B_scan | 021 | Geometric Segment      -3958    sample01</span></span>
<span><span class="co">## disease1B_scan | 022 | Geometric Segment      -6550    sample01</span></span>
<span><span class="co">## disease1B_scan | 023 | Geometric Segment      -6304    sample01</span></span>
<span><span class="co">## disease1B_scan | 024 | Geometric Segment      -2039    sample01</span></span></code></pre>
<p>Gene metadata are stored in the <code>rowData</code> of the
object.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">rowData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 18503 rows and 3 columns</span></span>
<span><span class="co">##              RTS_ID   ProbeID  Negative</span></span>
<span><span class="co">##         &lt;character&gt; &lt;numeric&gt; &lt;logical&gt;</span></span>
<span><span class="co">## PADI2    RTS0051249     34691     FALSE</span></span>
<span><span class="co">## CYP24A1  RTS0021490     36149     FALSE</span></span>
<span><span class="co">## SUPT16H  RTS0030451     44219     FALSE</span></span>
<span><span class="co">## ZMIZ2    RTS0034986     48277     FALSE</span></span>
<span><span class="co">## SPAG9    RTS0024185     38596     FALSE</span></span>
<span><span class="co">## ...             ...       ...       ...</span></span>
<span><span class="co">## TMPRSS4  RTS0051782     35235     FALSE</span></span>
<span><span class="co">## ATP6V0B  RTS0028259     42276     FALSE</span></span>
<span><span class="co">## DCHS1    RTS0028059     42092     FALSE</span></span>
<span><span class="co">## FAM166A  RTS0021731     36360     FALSE</span></span>
<span><span class="co">## AGTPBP1  RTS0050106     35406     FALSE</span></span></code></pre>
<p>By default, the <code>readGeoMx</code> function will remove the
negative probe, the one with name “NegProbe-WTX”, in the count table and
put it in the metadata of the object. User can turn this off by
specifying <code>hasNegProbe = FALSE</code> in the function.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">metadata</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">NegProbes</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##              disease3_scan | 001 | PanCK disease3_scan | 001 | neg</span></span>
<span><span class="co">## NegProbe-WTX                    18.08935                  34.26331</span></span>
<span><span class="co">## NA                                    NA                        NA</span></span>
<span><span class="co">## NA.1                                  NA                        NA</span></span>
<span><span class="co">## NA.2                                  NA                        NA</span></span>
<span><span class="co">## NA.3                                  NA                        NA</span></span>
<span><span class="co">##              disease3_scan | 002 | PanCK disease3_scan | 002 | neg</span></span>
<span><span class="co">## NegProbe-WTX                    21.55552                  40.01078</span></span>
<span><span class="co">## NA                                    NA                        NA</span></span>
<span><span class="co">## NA.1                                  NA                        NA</span></span>
<span><span class="co">## NA.2                                  NA                        NA</span></span>
<span><span class="co">## NA.3                                  NA                        NA</span></span>
<span><span class="co">##              disease3_scan | 003 | PanCK</span></span>
<span><span class="co">## NegProbe-WTX                    8.992909</span></span>
<span><span class="co">## NA                                    NA</span></span>
<span><span class="co">## NA.1                                  NA</span></span>
<span><span class="co">## NA.2                                  NA</span></span>
<span><span class="co">## NA.3                                  NA</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="import-from-dgelist-object">Import from DGEList object<a class="anchor" aria-label="anchor" href="#import-from-dgelist-object"></a>
</h3>
<p>Alternatively, <code>standR</code> provided a function to generate a
spatial experiment object from a DGEList object, which would be useful
for users who are familiar with the <code>edgeR</code> package.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dge</span> <span class="op">&lt;-</span> <span class="fu">edgeR</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/SE2DGEList.html" class="external-link">SE2DGEList</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span></span>
<span></span>
<span><span class="va">spe2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/standR/man/readGeoMxFromDGE.html" class="external-link">readGeoMxFromDGE</a></span><span class="op">(</span><span class="va">dge</span><span class="op">)</span></span>
<span></span>
<span><span class="va">spe2</span></span></code></pre></div>
<pre><code><span><span class="co">## class: SpatialExperiment </span></span>
<span><span class="co">## dim: 18503 231 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(2): counts logcounts</span></span>
<span><span class="co">## rownames(18503): PADI2 CYP24A1 ... FAM166A AGTPBP1</span></span>
<span><span class="co">## rowData names(3): RTS_ID ProbeID Negative</span></span>
<span><span class="co">## colnames(231): disease3_scan | 001 | PanCK disease3_scan | 001 | neg</span></span>
<span><span class="co">##   ... disease1B_scan | 023 | Geometric Segment disease1B_scan | 024 |</span></span>
<span><span class="co">##   Geometric Segment</span></span>
<span><span class="co">## colData names(28): group lib.size ... RoiReportY sample_id</span></span>
<span><span class="co">## reducedDimNames(0):</span></span>
<span><span class="co">## mainExpName: NULL</span></span>
<span><span class="co">## altExpNames(0):</span></span>
<span><span class="co">## spatialCoords names(0) :</span></span>
<span><span class="co">## imgData names(0):</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="quality-control">Quality control<a class="anchor" aria-label="anchor" href="#quality-control"></a>
</h3>
</div>
</div>
<div class="section level2">
<h2 id="quality-control-1">Quality control<a class="anchor" aria-label="anchor" href="#quality-control-1"></a>
</h2>
<p>The quality control (QC) for the GeoMX transcriptome data are
consisting of three major steps:</p>
<ol style="list-style-type: decimal">
<li><p><strong>Inspection of the sample metadata</strong>: sample
metadata can be seen as tabular-like format using the
<code>colData</code> function, however here we aim to visualise the
relations among important sample metadata, such as what slides did the
ROI come from, which are the control groups and treatment groups, what
are the pre-defined tissue types etc. By doing this step, we can have a
general idea of some important question such as how the experiment was
designed, should we be looking out for batch effect, and what interested
comparison can be established.</p></li>
<li><p><strong>Gene level QC</strong>: at gene level, by default we aim
at removing genes that are not expressed in more than 90% of the ROIs,
and identifying any ROIs with only few genes that are expressed. This is
the similar process as <code><a href="https://rdrr.io/pkg/edgeR/man/filterByExpr.html" class="external-link">edgeR::filterByExpr</a></code>, as genes with
consistently low counts are unlikely be identified as DE genes, and only
keeping genes with sufficiently large counts in the DE analysis can
increase statistical power while reduce multiple testing
burden.</p></li>
<li><p><strong>ROI level QC</strong>: at ROI level, we aim at identify
low-quality ROIs with relatively small library size (total feature
count) and low cell count. Problematic ROIs that are not removed could
show up as separate clusters in PCA/UMAPs and further affect the
comparisons in DE analysis.</p></li>
</ol>
<div class="section level4">
<h4 id="qc-on-sample-information">QC on sample information<a class="anchor" aria-label="anchor" href="#qc-on-sample-information"></a>
</h4>
<p>To visualise sample metadata, we can use the
<code>plotSampleInfo</code> function. In this dataset, slides, diseases
status, tissue regions, and different subtypes of the tissues are the
important features that we could input into the function.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/standR/man/plotSampleInfo.html" class="external-link">plotSampleInfo</a></span><span class="op">(</span><span class="va">spe</span>, column2plot <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SlideName"</span>,<span class="st">"disease_status"</span>,<span class="st">"region"</span>,<span class="st">"SegmentLabel"</span>,<span class="st">"pathology"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="GeoMXAnalysisWorkflow_files/figure-html/unnamed-chunk-14-1.png" width="700"></p>
<p>Based on the description of the data, we know that all glomerulus are
classified as abnormal and healthy, and tubule are classified as neg and
PanCK.</p>
<p>We therefore merge the region-related annotations to avoid
collinearity, which can affect the process of batch correction.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">colData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">regions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">region</span>,<span class="st">"_"</span>,<span class="fu">colData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">SegmentLabel</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"_Geometric Segment"</span>,<span class="st">""</span>,<span class="va">.</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"_"</span>,<span class="fu">colData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">pathology</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"_NA"</span>,<span class="st">""</span>,<span class="va">.</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="fu">colData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">regions</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## .</span></span>
<span><span class="co">## glomerulus_abnormal  glomerulus_healthy          tubule_neg        tubule_PanCK </span></span>
<span><span class="co">##                  77                  72                  41                  41</span></span></code></pre>
<p>Re-check the sample annotation, we can then have a better idea of the
data and what comparisons we can establish.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/standR/man/plotSampleInfo.html" class="external-link">plotSampleInfo</a></span><span class="op">(</span><span class="va">spe</span>, column2plot <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SlideName"</span>,<span class="st">"disease_status"</span>,<span class="st">"regions"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="GeoMXAnalysisWorkflow_files/figure-html/unnamed-chunk-16-1.png" width="700"></p>
</div>
<div class="section level4">
<h4 id="qc-on-genes">QC on genes<a class="anchor" aria-label="anchor" href="#qc-on-genes"></a>
</h4>
<p>Using the <code>addPerROIQC</code> function, we can add key
statistics to the <code>colData</code> of the object. In this case,
argument <code>rm_genes</code> is set to TRUE, with the default settings
of <code>min_count = 5</code> and <code>sample_fraction = 0.9</code>,
meaning here we first calculated a expression threshold on logCPM scale
(to account for library size differences), then genes that have
expression value below the threshold in more than 90% of the ROIs will
be removed. The count matrix with the removed gene are stored in the
<code>metadata</code> of the object, along with the calculated
expression threshold.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/standR/man/addPerROIQC.html" class="external-link">addPerROIQC</a></span><span class="op">(</span><span class="va">spe</span>, rm_genes <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Here we can see that 121 genes are removed.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 18382   231</span></span></code></pre>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">metadata</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "NegProbes"         "lcpm_threshold"    "genes_rm_rawCount"</span></span>
<span><span class="co">## [4] "genes_rm_logCPM"</span></span></code></pre>
<p>Using the <code>plotGeneQC</code> function, we can have a look at
which were the genes removed and the overall distribution of percentage
of non-expressed genes in all ROIs. By default, top 9 genes are plotted
here (arranging by mean expression), user can increase the number of
plotted genes by changing the <code>top_n</code> parameter.</p>
<p>Moreover, users can order the samples (using parameter
<code>ordannots</code>) and color/shape the dots with annotation to find
out if these genes are specifically expressed in some samples (e.g. in
some tissue types or in some treatment group) so that we may need to
retain them.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/standR/man/plotGeneQC.html" class="external-link">plotGeneQC</a></span><span class="op">(</span><span class="va">spe</span>, ordannots <span class="op">=</span> <span class="st">"regions"</span>, col <span class="op">=</span> <span class="va">regions</span>, point_size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="GeoMXAnalysisWorkflow_files/figure-html/unnamed-chunk-19-1.png" width="700"></p>
</div>
<div class="section level4">
<h4 id="qc-on-rois">QC on ROIs<a class="anchor" aria-label="anchor" href="#qc-on-rois"></a>
</h4>
</div>
<div class="section level3">
<h3 id="roi-level-qc">ROI level QC<a class="anchor" aria-label="anchor" href="#roi-level-qc"></a>
</h3>
<p>Using the <code>plotROIQC</code> function, we are able to visualise
some QC statistics in the ROI level, such as library size and cell count
(AOINucleiCount) (the default settinngs for this function).</p>
<p>In the ROI level QC, we first aim to identify (if any) ROI(s) that
have relatively low library size and low cell count because they are
considered as low quality samples due to insufficient sequencing depth
or lack of RNA in the chosen region.</p>
<p>In this case, looking at the distribution plots of library size and
nuclei count, we don’t see any particular spike in the low ends, rather
the distributions are relatively smooth. Looking at the dot plot,
library sizes are mostly positively correlate with the nuclei count,
with some data have relatively low library size while the nuclei count
is reasonable. We therefore can try to draw an filtering threshold at
the low end of the library size, in this case 50000. By coloring the dot
with their slide names, we find that the ROIs below the threshold are
all from slide disease1B, suggesting the reason for this might be some
technical issues of slide disease1B.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/standR/man/plotROIQC.html" class="external-link">plotROIQC</a></span><span class="op">(</span><span class="va">spe</span>, y_threshold <span class="op">=</span> <span class="fl">50000</span>, col <span class="op">=</span> <span class="va">SlideName</span><span class="op">)</span></span></code></pre></div>
<p><img src="GeoMXAnalysisWorkflow_files/figure-html/unnamed-chunk-20-1.png" width="700"></p>
<p>Since library size of 50000 seems to be a reasonable threshold, here
we subset the spatial experiment object based on the library size in
<code>colData</code>.</p>
<div class="section level4">
<h4 id="relative-log-expression-distribution">Relative log expression distribution<a class="anchor" aria-label="anchor" href="#relative-log-expression-distribution"></a>
</h4>
<p>We can use function <code>plotRLExpr</code> to visualise the relative
log expression (RLE) of the data to inspect the technical variation of
the data by looking at the distance from the median of the RLE (the
boxplot dot) to zero.</p>
<p>By default, we are plotting RLE of the raw count, where most of the
variation are from library size differences.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/standR/man/plotRLExpr.html" class="external-link">plotRLExpr</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span></span></code></pre></div>
<p><img src="GeoMXAnalysisWorkflow_files/figure-html/unnamed-chunk-21-1.png" width="700"></p>
<p>Using <code>assay = 2</code> to run RLE on the logCPM data, and we
can see that most technical variation from library sizes are removed.
And we can also sort the data using the annotation in the object by
specifying <code>ordannots</code>, with color or shape mapping parameter
as we usually write in ggplot, so that we can have a look at what factor
is heavily contributing to the technical variation.</p>
<p>Here we can see obvious variation from slides to slides, and small
variations are also observed within each slide.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/standR/man/plotRLExpr.html" class="external-link">plotRLExpr</a></span><span class="op">(</span><span class="va">spe</span>, ordannots <span class="op">=</span> <span class="st">"SlideName"</span>, assay <span class="op">=</span> <span class="fl">2</span>, col <span class="op">=</span> <span class="va">SlideName</span><span class="op">)</span></span></code></pre></div>
<p><img src="GeoMXAnalysisWorkflow_files/figure-html/unnamed-chunk-22-1.png" width="700"></p>
</div>
<div class="section level4">
<h4 id="dimension-reduction">Dimension reduction<a class="anchor" aria-label="anchor" href="#dimension-reduction"></a>
</h4>
<div class="section level5">
<h5 id="pca">PCA<a class="anchor" aria-label="anchor" href="#pca"></a>
</h5>
<p>Using the <code>drawPCA</code> function, we can perform principal
component analysis (PCA) on the data. The PCA plot can help visualising
the variation (both biological and technical) in the data and finding
out which are the main factors contributing to the variations.</p>
<p>Here we color the PCA with slide information, and shape by regions
(tissue). We can see that PC1 is mainly spread out by regions,
especially glomerulus and tubule. And grouping based on slide within
each tissue are observed. The subtypes in tubule are clearly separated,
but different subtypes of glomerulus is still grouping together.
Moreover, diseased tissues and control tissues are mixed as well
(disease slides and normal slides).</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/standR/man/drawPCA.html" class="external-link">drawPCA</a></span><span class="op">(</span><span class="va">spe</span>, assay <span class="op">=</span> <span class="fl">2</span>, col <span class="op">=</span> <span class="va">SlideName</span>, shape <span class="op">=</span> <span class="va">regions</span><span class="op">)</span></span></code></pre></div>
<p><img src="GeoMXAnalysisWorkflow_files/figure-html/unnamed-chunk-23-1.png" width="700"></p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/standR/man/drawPCA.html" class="external-link">drawPCA</a></span><span class="op">(</span><span class="va">spe</span>, assay <span class="op">=</span> <span class="fl">2</span>, col <span class="op">=</span> <span class="va">disease_status</span>, shape <span class="op">=</span> <span class="va">regions</span><span class="op">)</span></span></code></pre></div>
<p><img src="GeoMXAnalysisWorkflow_files/figure-html/unnamed-chunk-24-1.png" width="700"></p>
<p>The <code>standR</code> package also provide other functions to
visualise the PCA, including the PCA scree plot, pair-dimension PCA plot
and PCA bi-plot.</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/standR/man/plotScreePCA.html" class="external-link">plotScreePCA</a></span><span class="op">(</span><span class="va">spe</span>, assay <span class="op">=</span> <span class="fl">2</span>, dims <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<p><img src="GeoMXAnalysisWorkflow_files/figure-html/unnamed-chunk-25-1.png" width="700"></p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/standR/man/plotPairPCA.html" class="external-link">plotPairPCA</a></span><span class="op">(</span><span class="va">spe</span>, col <span class="op">=</span> <span class="va">disease_status</span>, </span>
<span>            shape <span class="op">=</span> <span class="va">regions</span>, assay <span class="op">=</span> <span class="fl">2</span>, n_dimension <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span></code></pre></div>
<p><img src="GeoMXAnalysisWorkflow_files/figure-html/unnamed-chunk-26-1.png" width="700"></p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/standR/man/plotPCAbiplot.html" class="external-link">plotPCAbiplot</a></span><span class="op">(</span><span class="va">spe</span>, n_loadings <span class="op">=</span> <span class="fl">10</span>, assay <span class="op">=</span> <span class="fl">2</span>, col <span class="op">=</span> <span class="va">regions</span><span class="op">)</span></span></code></pre></div>
<p><img src="GeoMXAnalysisWorkflow_files/figure-html/unnamed-chunk-27-1.png" width="700"></p>
</div>
<div class="section level5">
<h5 id="mds">MDS<a class="anchor" aria-label="anchor" href="#mds"></a>
</h5>
<p><code>plotMDS</code> can be used to visualise the multidimension
scaling of the data.</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">standR</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/standR/man/plotMDS.html" class="external-link">plotMDS</a></span><span class="op">(</span><span class="va">spe</span>, assay <span class="op">=</span> <span class="fl">2</span>, color <span class="op">=</span> <span class="va">disease_status</span>, shape <span class="op">=</span> <span class="va">regions</span><span class="op">)</span></span></code></pre></div>
<p><img src="GeoMXAnalysisWorkflow_files/figure-html/unnamed-chunk-28-1.png" width="700"></p>
</div>
<div class="section level5">
<h5 id="umap">UMAP<a class="anchor" aria-label="anchor" href="#umap"></a>
</h5>
<p>Further more, since we’re using <code>SpatialExperiment</code> as our
infrastructure, it is easy to use other popular packages such as the
<code>scater</code> package, which is commonly used in single cell data
and visium data analysis, we provide the function <code>plotDR</code> to
visualise any kinds of dimension reduction results generated from
<code>scater::run*</code>, such as UMAP, TSNE and NMF, by specifying
<code>dimred</code>.</p>
<p>Here we plot the UMAP of our data. Similar variation can be observed
like PCA and MDS above.</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html" class="external-link">runUMAP</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/standR/man/plotDR.html" class="external-link">plotDR</a></span><span class="op">(</span><span class="va">spe</span>, dimred <span class="op">=</span> <span class="st">"UMAP"</span>, col <span class="op">=</span> <span class="va">regions</span><span class="op">)</span></span></code></pre></div>
<p><img src="GeoMXAnalysisWorkflow_files/figure-html/unnamed-chunk-29-1.png" width="700"></p>
</div>
</div>
</div>
<div class="section level3">
<h3 id="normalization">Normalization<a class="anchor" aria-label="anchor" href="#normalization"></a>
</h3>
<p>As we observed the technical variations in the data in both RLE and
PCA plots. It is necessary to perform normalization in the data.</p>
<p>In the <code>standR</code> package, we offer normalization options
including TMM, RPKM, TPM, CPM, upperquartile and sizefactor. Among them,
RPKM and TPM required gene length information (add
<code>genelength</code> column to the <code>rowData</code> of the
object). For TMM, upperquartile and sizefactor, their normalized factor
will be stored their <code>metadata</code>.</p>
<p>Here we used TMM to normalize the data.</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">colData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">biology</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">disease_status</span>, <span class="st">"_"</span>, <span class="fu">colData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">regions</span><span class="op">)</span></span>
<span></span>
<span><span class="va">spe_tmm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/standR/man/geomxNorm.html" class="external-link">geomxNorm</a></span><span class="op">(</span><span class="va">spe</span>, method <span class="op">=</span> <span class="st">"TMM"</span><span class="op">)</span></span></code></pre></div>
<p>Then we use RLE and PCA plot to assess the normalization count.</p>
<p>The RLE plot show most of the median of RLE are close to zero,
indicating that lots of technical variation are removed.</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/standR/man/plotRLExpr.html" class="external-link">plotRLExpr</a></span><span class="op">(</span><span class="va">spe_tmm</span>, assay <span class="op">=</span> <span class="fl">2</span>, color <span class="op">=</span> <span class="va">SlideName</span><span class="op">)</span> <span class="op">+</span> <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"TMM"</span><span class="op">)</span></span></code></pre></div>
<p><img src="GeoMXAnalysisWorkflow_files/figure-html/unnamed-chunk-31-1.png" width="700"></p>
<p>However, batch effect from the different slides are still observed,
confounding the separation between disease and normal.</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/standR/man/plotPairPCA.html" class="external-link">plotPairPCA</a></span><span class="op">(</span><span class="va">spe_tmm</span>, assay <span class="op">=</span> <span class="fl">2</span>, color <span class="op">=</span> <span class="va">disease_status</span>, shape <span class="op">=</span> <span class="va">regions</span><span class="op">)</span></span></code></pre></div>
<p><img src="GeoMXAnalysisWorkflow_files/figure-html/unnamed-chunk-32-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="batch-correction">Batch correction<a class="anchor" aria-label="anchor" href="#batch-correction"></a>
</h3>
<p>In the Nanostring’s GeoMX DSP protocol, due to the fact that one
slide is only big enough for a handful of tissue segments (ROIs), it is
common that we see the DSP data being confounded by the batch effect
introduced by different slides. In order to establish fair comparison
between ROIs later on, it is necessary to remove this batch effect from
the data.</p>
<p>In the <code>standR</code> package, we provide two approaches of
removing batch effects, including RUV4 and Limma.</p>
<div class="section level4">
<h4 id="correction-method-ruv4">Correction method : RUV4<a class="anchor" aria-label="anchor" href="#correction-method-ruv4"></a>
</h4>
<p>To run RUV4 batch correction, we need to provide a list of “negative
control genes (NCGs)”.</p>
<p>The function <code>findNCGs</code> allows identifying the NCGs from
the data. In this case, since the batch effect is mostly introduced by
slide, we therefore want to identify NCGs across all slides, so here we
set the <code>batch_name</code> to “SlideName”, and select the top 500
least variable genes across different slides as NCGs.</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/standR/man/findNCGs.html" class="external-link">findNCGs</a></span><span class="op">(</span><span class="va">spe</span>, batch_name <span class="op">=</span> <span class="st">"SlideName"</span>, top_n <span class="op">=</span> <span class="fl">300</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">metadata</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "NegProbes"         "lcpm_threshold"    "genes_rm_rawCount"</span></span>
<span><span class="co">## [4] "genes_rm_logCPM"   "NCGs"</span></span></code></pre>
<p>Now we can run RUV4 on the data using function
<code>geomxBatchCorrection</code>. By default this function will use
<code>RUV4</code> to normalize the data. For RUV4 correction, the
function is requiring 3 parameters other than the input object,
including <code>factors</code>: the factor of interest, i.e. the
biological variation we plan to keep; <code>NCGs</code>: the list of
negative control genes detected using the function
<code>findNCGs</code>; and <code>k</code>: is the number of unwanted
factors to use, in the RUV documentation, it is suggest that we should
use the smallest k once we don’t observe technical variation in the
data.</p>
<p>Choosing the best k is one of the most important task when performing
batch correction with RUV. And the best way to do so is to test out each
k have a look with the corresponded diagnostic plot, such as PCA plot.
The best k would be the smallest number that satisfy the separation of
main biology on a PCA plot.</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">5</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">spe_ruv</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/standR/man/geomxBatchCorrection.html" class="external-link">geomxBatchCorrection</a></span><span class="op">(</span><span class="va">spe</span>, factors <span class="op">=</span> <span class="st">"biology"</span>, </span>
<span>                   NCGs <span class="op">=</span> <span class="fu">metadata</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">NCGs</span>, k <span class="op">=</span> <span class="va">i</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/standR/man/plotPairPCA.html" class="external-link">plotPairPCA</a></span><span class="op">(</span><span class="va">spe_ruv</span>, assay <span class="op">=</span> <span class="fl">2</span>, n_dimension <span class="op">=</span> <span class="fl">4</span>, color <span class="op">=</span> <span class="va">disease_status</span>, shape <span class="op">=</span> <span class="va">regions</span>, title <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"k = "</span>, <span class="va">i</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>  </span>
<span><span class="op">}</span></span></code></pre></div>
<p><img src="GeoMXAnalysisWorkflow_files/figure-html/unnamed-chunk-34-1.png" width="700"><img src="GeoMXAnalysisWorkflow_files/figure-html/unnamed-chunk-34-2.png" width="700"><img src="GeoMXAnalysisWorkflow_files/figure-html/unnamed-chunk-34-3.png" width="700"><img src="GeoMXAnalysisWorkflow_files/figure-html/unnamed-chunk-34-4.png" width="700"><img src="GeoMXAnalysisWorkflow_files/figure-html/unnamed-chunk-34-5.png" width="700"></p>
<p>Here we choose k = 4 to be our best k.</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe_ruv</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/standR/man/geomxBatchCorrection.html" class="external-link">geomxBatchCorrection</a></span><span class="op">(</span><span class="va">spe</span>, factors <span class="op">=</span> <span class="st">"biology"</span>, </span>
<span>                   NCGs <span class="op">=</span> <span class="fu">metadata</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">NCGs</span>, k <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/standR/man/plotPairPCA.html" class="external-link">plotPairPCA</a></span><span class="op">(</span><span class="va">spe_ruv</span>, assay <span class="op">=</span> <span class="fl">2</span>, color <span class="op">=</span> <span class="va">disease_status</span>, shape <span class="op">=</span> <span class="va">regions</span>, title <span class="op">=</span> <span class="st">"RUV4, k = 4"</span>, n_dimension <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span></code></pre></div>
<p><img src="GeoMXAnalysisWorkflow_files/figure-html/unnamed-chunk-35-1.png" width="700"></p>
</div>
<div class="section level4">
<h4 id="correction-method-limma">Correction method: limma<a class="anchor" aria-label="anchor" href="#correction-method-limma"></a>
</h4>
<p>Another option is set the parameter <code>method</code> to “Limma”,
which uses the remove batch correction method from <code>limma</code>.
In this mode, the function is requiring 2 parameters, including
<code>batch</code>: a vector that indicating batches for all samples;
and <code>design</code>: a design matrix which is generated by
<code>model.matrix</code>, in the design matrix, all
biologically-relevant factors should be included.</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe_lrb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/standR/man/geomxBatchCorrection.html" class="external-link">geomxBatchCorrection</a></span><span class="op">(</span><span class="va">spe</span>,</span>
<span>                       batch <span class="op">=</span> <span class="fu">colData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">$</span><span class="va">SlideName</span>, method <span class="op">=</span> <span class="st">"Limma"</span>,</span>
<span>                       design <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="op">~</span> <span class="fl">0</span> <span class="op">+</span> <span class="va">disease_status</span> <span class="op">+</span> <span class="va">regions</span>, </span>
<span>                                             data <span class="op">=</span> <span class="fu">colData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Coefficients not estimable: batch6</span></span></code></pre>
<p>Again, using PCA to inspect the batch correction process.</p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/standR/man/plotPairPCA.html" class="external-link">plotPairPCA</a></span><span class="op">(</span><span class="va">spe_lrb</span>, assay <span class="op">=</span> <span class="fl">2</span>, color <span class="op">=</span> <span class="va">disease_status</span>, shape <span class="op">=</span> <span class="va">regions</span>, title <span class="op">=</span> <span class="st">"Limma removeBatch"</span>, n_dimension <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span></code></pre></div>
<p><img src="GeoMXAnalysisWorkflow_files/figure-html/unnamed-chunk-37-1.png" width="700"></p>
</div>
<div class="section level4">
<h4 id="evaluation">Evaluation<a class="anchor" aria-label="anchor" href="#evaluation"></a>
</h4>
<div class="section level5">
<h5 id="summary-statistics">Summary statistics<a class="anchor" aria-label="anchor" href="#summary-statistics"></a>
</h5>
<p>Besides looking at the PCA plots of normalized count, another way to
evaluate the batch correction is by summarising statistics such as
adjusted rand index, jaccard similarity coefficient, silhouette
coefficient and etc. We can do this by using the
<code>plotClusterEvalStats</code> function. In the biology section,
higher the score is better while in the batch section, lower is
better.</p>
<p>We can see that when it comes to stratifying for biology factors
(disease status and tissue regions) or measuring batch clustering for
this dataset, RUV4 outperform Limma in most statistics.</p>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">spe</span>, <span class="va">spe_ruv</span>, <span class="va">spe_lrb</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/standR/man/plotClusterEvalStats.html" class="external-link">plotClusterEvalStats</a></span><span class="op">(</span>spe_list <span class="op">=</span> <span class="va">spe_list</span>,</span>
<span>                     bio_feature_name <span class="op">=</span> <span class="st">"regions"</span>,</span>
<span>                     batch_feature_name <span class="op">=</span> <span class="st">"SlideName"</span>,</span>
<span>                     data_names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Raw"</span>,<span class="st">"RUV4"</span>,<span class="st">"Limma"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="GeoMXAnalysisWorkflow_files/figure-html/unnamed-chunk-38-1.png" width="700"></p>
</div>
<div class="section level5">
<h5 id="rle-plots">RLE plots<a class="anchor" aria-label="anchor" href="#rle-plots"></a>
</h5>
<p>Moreover, we can also have a look at the RLE plots of the normalized
count to determine which batch correction performs better for this
dataset.</p>
<p>We can clearly see that RUV4-corrected count have a overall
more-closer-to-zero median RLE compared to the limma-corrected data.
Therefore, in the downstream differential expression analysis, we would
suggest using the RUV4 as the batch correction method for this specific
dataset.</p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/standR/man/plotRLExpr.html" class="external-link">plotRLExpr</a></span><span class="op">(</span><span class="va">spe_ruv</span>, assay <span class="op">=</span> <span class="fl">2</span>, color <span class="op">=</span> <span class="va">SlideName</span><span class="op">)</span> <span class="op">+</span> <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"RUV4"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/standR/man/plotRLExpr.html" class="external-link">plotRLExpr</a></span><span class="op">(</span><span class="va">spe_lrb</span>, assay <span class="op">=</span> <span class="fl">2</span>, color <span class="op">=</span> <span class="va">SlideName</span><span class="op">)</span> <span class="op">+</span> <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"Limma removeBatch"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="differential-expression-analysis-with-limma-voom-pipeline">Differential expression analysis with limma-voom pipeline<a class="anchor" aria-label="anchor" href="#differential-expression-analysis-with-limma-voom-pipeline"></a>
</h2>
<p>We recommend using comprehensive pipelines, such as
<code>edgeR</code>, <code>limma-voom</code> and <code>DESeq2</code>,
instead of simple T-test to perform DE analysis of GeoMX DSP data.</p>
<p>In this workshop, we’ll demonstrate the DE analysis using the
<code>limma-voom</code> pipeline.</p>
<div class="section level3">
<h3 id="linear-model">Linear model<a class="anchor" aria-label="anchor" href="#linear-model"></a>
</h3>
<p>Since we’ve shown in previous sections that for this dataset, using
<code>k = 4</code> is able to remove the batch effect and other undsired
technical variations. Here in the linear model, we need to include the
weight matrix generated from the function
<code>geomxBatchCorrection</code> as covariates. The weight matrix can
be found in the <code>colData</code>.</p>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioinf.wehi.edu.au/edgeR" class="external-link">edgeR</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioinf.wehi.edu.au/limma" class="external-link">limma</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">dge</span> <span class="op">&lt;-</span> <span class="fu">edgeR</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/SE2DGEList.html" class="external-link">SE2DGEList</a></span><span class="op">(</span><span class="va">spe_ruv</span><span class="op">)</span></span>
<span></span>
<span><span class="va">design</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="op">~</span><span class="fl">0</span> <span class="op">+</span> <span class="va">biology</span> <span class="op">+</span> <span class="va">ruv_W1</span> <span class="op">+</span> <span class="va">ruv_W2</span> <span class="op">+</span> <span class="va">ruv_W3</span> <span class="op">+</span> <span class="va">ruv_W4</span>, data <span class="op">=</span> <span class="fu">colData</span><span class="op">(</span><span class="va">spe_ruv</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">design</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">design</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">str_replace</span><span class="op">(</span><span class="st">"biology"</span>,<span class="st">""</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">design</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "DKD_glomerulus_abnormal"    "DKD_glomerulus_healthy"    </span></span>
<span><span class="co">##  [3] "DKD_tubule_neg"             "DKD_tubule_PanCK"          </span></span>
<span><span class="co">##  [5] "normal_glomerulus_abnormal" "normal_glomerulus_healthy" </span></span>
<span><span class="co">##  [7] "normal_tubule_neg"          "normal_tubule_PanCK"       </span></span>
<span><span class="co">##  [9] "ruv_W1"                     "ruv_W2"                    </span></span>
<span><span class="co">## [11] "ruv_W3"                     "ruv_W4"</span></span></code></pre>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">keep</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/filterByExpr.html" class="external-link">filterByExpr</a></span><span class="op">(</span><span class="va">dge</span>, <span class="va">design</span><span class="op">)</span></span>
<span><span class="va">dge_all</span> <span class="op">&lt;-</span> <span class="va">dge</span><span class="op">[</span><span class="va">keep</span>, <span class="op">]</span></span>
<span></span>
<span><span class="va">dge_all</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/estimateDisp.html" class="external-link">estimateDisp</a></span><span class="op">(</span><span class="va">dge_all</span>, <span class="va">design</span>, robust<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="biological-coefficient-of-variation">Biological coefficient of variation<a class="anchor" aria-label="anchor" href="#biological-coefficient-of-variation"></a>
</h3>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/plotBCV.html" class="external-link">plotBCV</a></span><span class="op">(</span><span class="va">dge_all</span><span class="op">)</span></span>
<span><span class="va">bcv_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  <span class="st">'BCV'</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="va">dge_all</span><span class="op">$</span><span class="va">tagwise.dispersion</span><span class="op">)</span>,</span>
<span>  <span class="st">'AveLogCPM'</span> <span class="op">=</span> <span class="va">dge_all</span><span class="op">$</span><span class="va">AveLogCPM</span>,</span>
<span>  <span class="st">'gene_id'</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">dge_all</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">highbcv</span> <span class="op">&lt;-</span> <span class="va">bcv_df</span><span class="op">$</span><span class="va">AveLogCPM</span> <span class="op">&gt;</span> <span class="fl">7</span> <span class="op">&amp;</span> <span class="va">bcv_df</span><span class="op">$</span><span class="va">BCV</span> <span class="op">&gt;</span> <span class="fl">0.8</span></span>
<span><span class="va">highbcv_df</span> <span class="op">&lt;-</span> <span class="va">bcv_df</span><span class="op">[</span><span class="va">highbcv</span>, <span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/points.html" class="external-link">points</a></span><span class="op">(</span><span class="va">highbcv_df</span><span class="op">$</span><span class="va">AveLogCPM</span>, <span class="va">highbcv_df</span><span class="op">$</span><span class="va">BCV</span>, col <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/text.html" class="external-link">text</a></span><span class="op">(</span><span class="va">highbcv_df</span><span class="op">$</span><span class="va">AveLogCPM</span>, <span class="va">highbcv_df</span><span class="op">$</span><span class="va">BCV</span>, labels <span class="op">=</span> <span class="va">highbcv_df</span><span class="op">$</span><span class="va">gene_id</span>, pos <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span></code></pre></div>
<p><img src="GeoMXAnalysisWorkflow_files/figure-html/unnamed-chunk-41-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="differential-expression">Differential expression<a class="anchor" aria-label="anchor" href="#differential-expression"></a>
</h3>
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">v</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/voom.html" class="external-link">voom</a></span><span class="op">(</span><span class="va">dge_all</span>, <span class="va">design</span><span class="op">)</span> </span>
<span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/lmFit.html" class="external-link">lmFit</a></span><span class="op">(</span><span class="va">v</span><span class="op">)</span></span>
<span></span>
<span><span class="va">contr.matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/makeContrasts.html" class="external-link">makeContrasts</a></span><span class="op">(</span></span>
<span>   DvCinGab <span class="op">=</span> <span class="va">DKD_glomerulus_abnormal</span> <span class="op">-</span> <span class="va">normal_glomerulus_abnormal</span>,</span>
<span>   levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">design</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">fit_contrast</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/contrasts.fit.html" class="external-link">contrasts.fit</a></span><span class="op">(</span><span class="va">fit</span>, contrasts <span class="op">=</span> <span class="va">contr.matrix</span><span class="op">)</span></span>
<span></span>
<span><span class="va">efit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/ebayes.html" class="external-link">eBayes</a></span><span class="op">(</span><span class="va">fit_contrast</span>, robust <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">results_efit</span><span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/decideTests.html" class="external-link">decideTests</a></span><span class="op">(</span><span class="va">efit</span>, p.value <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span>
<span><span class="va">summary_efit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">results_efit</span><span class="op">)</span></span></code></pre></div>
<div class="section level4">
<h4 id="visualization">Visualization<a class="anchor" aria-label="anchor" href="#visualization"></a>
</h4>
<p>We can then visualise the DE genes using plots such as volcano plot
and MA plot.</p>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/slowkow/ggrepel" class="external-link">ggrepel</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">de_results_DvCinGab</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/toptable.html" class="external-link">topTable</a></span><span class="op">(</span><span class="va">efit</span>, coef <span class="op">=</span> <span class="fl">1</span>, sort.by <span class="op">=</span> <span class="st">"P"</span>, n <span class="op">=</span> <span class="cn">Inf</span><span class="op">)</span></span>
<span></span>
<span><span class="va">de_genes_toptable_DvCinGab</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/toptable.html" class="external-link">topTable</a></span><span class="op">(</span><span class="va">efit</span>, coef <span class="op">=</span> <span class="fl">1</span>, sort.by <span class="op">=</span> <span class="st">"P"</span>, n <span class="op">=</span> <span class="cn">Inf</span>, p.value <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span>
<span></span>
<span></span>
<span></span>
<span><span class="va">de_results_DvCinGab</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>DE <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">logFC</span> <span class="op">&gt;</span> <span class="fl">0</span> <span class="op">&amp;</span> <span class="va">adj.P.Val</span> <span class="op">&lt;</span><span class="fl">0.05</span>, <span class="st">"UP"</span>, </span>
<span>                       <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">logFC</span> <span class="op">&lt;</span><span class="fl">0</span> <span class="op">&amp;</span> <span class="va">adj.P.Val</span><span class="op">&lt;</span><span class="fl">0.05</span>, <span class="st">"DOWN"</span>, <span class="st">"NOT DE"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">AveExpr</span>, <span class="va">logFC</span>, col <span class="op">=</span> <span class="va">DE</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>shape <span class="op">=</span> <span class="fl">1</span>, size <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/ggrepel/man/geom_text_repel.html" class="external-link">geom_text_repel</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">de_genes_toptable_DvCinGab</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>                    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>DE <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">logFC</span> <span class="op">&gt;</span> <span class="fl">0</span> <span class="op">&amp;</span> <span class="va">adj.P.Val</span> <span class="op">&lt;</span><span class="fl">0.05</span>, <span class="st">"UP"</span>, </span>
<span>                       <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">logFC</span> <span class="op">&lt;</span><span class="fl">0</span> <span class="op">&amp;</span> <span class="va">adj.P.Val</span><span class="op">&lt;</span><span class="fl">0.05</span>, <span class="st">"DOWN"</span>, <span class="st">"NOT DE"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>                    <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html" class="external-link">rownames_to_column</a></span><span class="op">(</span><span class="op">)</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>label <span class="op">=</span> <span class="va">rowname</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"Average log-expression"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"Log-fold-change"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"DKD vs Normal in abnormal glomerulus (limma-voom)"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"blue"</span>,<span class="st">"gray"</span>,<span class="st">"red"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size<span class="op">=</span><span class="fl">15</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="GeoMXAnalysisWorkflow_files/figure-html/unnamed-chunk-43-1.png" width="700"></p>
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">de_results_DvCinGab</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>DE <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">logFC</span> <span class="op">&gt;</span> <span class="fl">0</span> <span class="op">&amp;</span> <span class="va">adj.P.Val</span> <span class="op">&lt;</span><span class="fl">0.05</span>, <span class="st">"UP"</span>, </span>
<span>                       <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">logFC</span> <span class="op">&lt;</span><span class="fl">0</span> <span class="op">&amp;</span> <span class="va">adj.P.Val</span><span class="op">&lt;</span><span class="fl">0.05</span>, <span class="st">"DOWN"</span>, <span class="st">"NOT DE"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">logFC</span>, <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="va">adj.P.Val</span><span class="op">)</span>, col <span class="op">=</span> <span class="va">DE</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>shape <span class="op">=</span> <span class="fl">1</span>, size <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/ggrepel/man/geom_text_repel.html" class="external-link">geom_text_repel</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">de_genes_toptable_DvCinGab</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>                    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>DE <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">logFC</span> <span class="op">&gt;</span> <span class="fl">0</span> <span class="op">&amp;</span> <span class="va">adj.P.Val</span> <span class="op">&lt;</span><span class="fl">0.05</span>, <span class="st">"UP"</span>, </span>
<span>                       <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">logFC</span> <span class="op">&lt;</span><span class="fl">0</span> <span class="op">&amp;</span> <span class="va">adj.P.Val</span><span class="op">&lt;</span><span class="fl">0.05</span>, <span class="st">"DOWN"</span>, <span class="st">"NOT DE"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>                    <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html" class="external-link">rownames_to_column</a></span><span class="op">(</span><span class="op">)</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>label <span class="op">=</span> <span class="va">rowname</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"Fold change (log2)"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"-log10(p-value)"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"DKD vs Normal in abnormal glomerulus (limma-voom)"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"blue"</span>,<span class="st">"gray"</span>,<span class="st">"red"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size<span class="op">=</span><span class="fl">15</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="GeoMXAnalysisWorkflow_files/figure-html/unnamed-chunk-43-2.png" width="700"></p>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">updn_cols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu">RColorBrewer</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html" class="external-link">brewer.pal</a></span><span class="op">(</span><span class="fl">6</span>, <span class="st">'Greens'</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>, <span class="fu">RColorBrewer</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html" class="external-link">brewer.pal</a></span><span class="op">(</span><span class="fl">6</span>, <span class="st">'Purples'</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="va">de_genes_toptable_DvCinGab</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"logFC"</span>, <span class="st">"AveExpr"</span>, <span class="st">"P.Value"</span>, <span class="st">"adj.P.Val"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">DT</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/DT/man/datatable.html" class="external-link">datatable</a></span><span class="op">(</span>caption <span class="op">=</span> <span class="st">'Differentially expressed genes. DKD vs Normal in abnormal glomerulus (limma-voom)'</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">DT</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/DT/man/formatCurrency.html" class="external-link">formatStyle</a></span><span class="op">(</span><span class="st">'logFC'</span>,</span>
<span>                valueColumns <span class="op">=</span> <span class="st">'logFC'</span>,</span>
<span>                backgroundColor <span class="op">=</span> <span class="fu">DT</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/DT/man/styleInterval.html" class="external-link">styleInterval</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fu"><a href="https://rdrr.io/r/base/rev.html" class="external-link">rev</a></span><span class="op">(</span><span class="va">updn_cols</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">DT</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/DT/man/formatCurrency.html" class="external-link">formatSignif</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span>, digits <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span></code></pre></div>
<div id="htmlwidget-b3f389d2bfa73dc7ef24" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-b3f389d2bfa73dc7ef24">{"x":{"filter":"none","vertical":false,"caption":"<caption>Differentially expressed genes. DKD vs Normal in abnormal glomerulus (limma-voom)<\/caption>","data":[["KLK6","RPS2","DDN","APLP1","DCN","NDNF","AIF1","FGF1","CALM2","KLK7","DSTN","MME","CDKN1C","MYL9","CLIC5","PCOLCE2","PEA15","SORT1","CHD3","PMEPA1","MSLN","SEMA5A","MPP5","NDRG2","CD68","B4GAT1","TPPP3","ARHGEF3","SMAD6","RPL41","PLS3","SYNPO","SPOCK2","RBMY1J","TJP1","HIPK2","MYO1E","HMOX1","ZDHHC6","FTH1","SLC2A3","GOLIM4","BGN","NPHS1","GBP2","NEBL","RBM3","TCN2","ZBTB4","MAL2","RPL18A","RPS19","RPLP0","DHRS3","PLA2R1","TKT","NPR3","ARHGEF12","WT1","TUSC3","FYN","RIPOR1","CRB2","SOD2","S100A10","SH3BGRL2","ORM2","SEPTIN11","PALLD","ALOX5","SELENOP","FGR","GPM6B","NAMPT","NPHS2","IFI30","DDX3Y","DAAM2","TNS3","TYRO3","SOST","HPGD","PRKAR2B","ST3GAL6","MYO1D","CCNI","NPNT","FLRT3","TCF21","TSC22D3","JUN","RPS28","SLC11A1","STARD7","TSPY1","RPL29","PRY","MAP1B","BHLHE40","C1R","SGCB","CALD1","MTSS1","INMT","TSPAN2","APLN","S100A8","NTNG1","RPL18","CLIP3","HTRA1","SUPT7L","SERF2","SLC6A6","PARVA","ACTN4","ARMH4","VAMP5","RPL28","RPS12","SEMA3G","ITGA8","DYNLL1","CNTN1","FOXC1","CLU","MYLK","THY1","RCN2","PODXL","CERS6","FAM20B","PRCP","ADAMTS1","CSDE1","ITGB1","SERINC2","GATM","S100A9","CAVIN2","TNS2","GAGE1","BAMBI","RPS9","PTH1R","PARD6G","HEXB","ANKRD33B","KLF4","RNF145","VSIG8","TSPAN1","LPAR6","OPRPN","ARRDC4","PDE4DIP","NET1","KRT8","TAF15","TIMP2","HCK","SLC7A8","USP46","CPM","GJA5","ABLIM3","STK17B","RPS27","NES","TNS1","IL1RL1","RASA4","HSFY1","GALNT15","ZBTB16","SRGAP2B","BEX3","CTDSPL","VNN2","AIF1L","FABP5","RPS10","WFDC1","NINL","MFSD6","KIRREL1","MICAL2","TBL1X","LCP1","BIN2","ALS2CL","CMYA5","TNFRSF1B","WIPF3","ATOH8","ESRRG","TAS1R3","STS","KDM5D","RPS8","KLF2","SNAI2","RPS4Y1","ARHGAP19","MXD1","CTSB","DCBLD2","ALDOB","ITGB8","RPS14","IFIT1","ARRDC2","PTPN13","NEDD9","ZNF185","DST","NRIP2","CLDN4","CGNL1","ITGB5","EPHB4","PTPN14","GNAS","MIA2","GTF2A2","SERINC1","CXorf49B","PKM","SGPP2","HLA-A","RPS20","CA10","PDE10A","HNRNPH3","OAZ2","FAM89A","SUN2","SLC38A2","TMEM252","CD9","CORO1A","NCBP2","SSX2","WFDC2","DDIT4","FCER1G","UNC5B","SLK","AKR1B15","CD81","EMC10","TGM2","PDPN","RNF11","CELF5","MYH9","SPI1","DSP","DDX17","MXRA8","N4BP2L2","SLC50A1","LRRC2","PHACTR4","ARMCX3","ZNF664","NKD1","EEF1G","CAPN2","PTPRQ","LILRB2","CHRDL1","RPL13","MIB1","GBP3","DENND2D","RENBP","KANK1","TMEM106C","SLC4A11","KCNG3","TFRC","REG1A","TMEM176A","SPOCK1","UBE2J1","SSB","FAM98A","YAP1","RCAN1","LMX1B","NELL1","RPS5","HEXIM1","OVGP1","RPS15A","ATP5MG","ANAPC16","TGFBR3","CYREN","HLX","FXYD2","SCTR","POSTN","SLC40A1","SEPTIN4","PAQR7","PTPN6","TIPIN","SLC6A8","WNT8A","PRDX3","CHTOP","ATP5MC2","CCN4","SDCBP","FBXO21","LILRB3","SAR1A","PTPRB","RPS3","PILRA","ZFY","AKR1C1","FPR1","ZNF140","IDO1","TNPO2","CIB1","OR10G9","FAT1","MGAT5","CCL14","MILR1","PATL1","CBLC","MORF4L1","PEBP1","TCTN1","HKDC1","CCDC198","TBCEL","FAM47E","CRYAB","THSD7A","RPL37A","PIEZO2","ANKRD13A","FAM156A","SPTAN1","MTRNR2L1","AFAP1L2","TRMT10A","PLK2","CDH11","CALCRL","LARGE1","HSPA4L","NRGN","SNX25","RPL9","USP9Y","CXADR","FAM43A","KRT18","ARPP19","GJA4","CTSH","TNFAIP3","FRZB","CREBL2","RPS17","WTIP","ARL17B","FAM90A1","BEX4","PNPLA2","RPL12","LRG1","RRAS","MPV17L2","SLC35F6","AKT3","MXD4","AGT","CRTAC1","RNF220","PPFIA4","DMTF1","TMSB4X","PIGR","CCDC80","DEDD2","TMCC3","GAS6","NFE2L1","SSX3","RPL7A","SLCO4A1","BPY2","LACTB2","TIA1","RIC3","BANF1","RAB32","KIAA1549L","LGALS3","YWHAE","TMEM98","MUC20","RPL34","DMGDH","PLXNA3","TMEM47","MYC","HGD","TAAR8","KIAA1549","PCMTD2","PPP2CB","S100A11","GMFB","CXCL14","ACTR3B","UACA","PRX","SCOC","SLC2A13","CNTLN","SKP1","CTSL","SLC25A6","F2R","UBP1","SH3BP2","RPL27A","MIDN","SLC16A3","ANAPC13","PRMT6","ETS2","GARNL3","MRFAP1","NOS3","VCL","SRFBP1","NEK9","HYOU1","KIDINS220","SIGLEC14","IL6ST","LILRA5","VNN1","TTC39A","KAT7","PAMR1","FOSL2","SNURF","GBP4","GRN","NFX1","SLC16A5","PAFAH1B1","CT45A1","EXOC3L2","CALM1","HIF1A","SEPTIN2","TRIM22","OR1F1","FRY","GPI","FXYD1","CTNNBIP1","NTN4","ENG","RPL17","SPRYD3","OSMR","FCRL4","ARRB1","DEFA1B","PLD3","WDR41","PALD1","AFG3L2","C11orf58","KCNK5","BACE1","TCIM","TMEM262","NUDT10","RPS23","COPE","DESI1","FERMT2","RUFY3","HNRNPA1","IFITM10","NOP53","PTPN21","HYAL2","MAF","MGST1","ECM2","ALYREF","PREX2","CNPY4","EFHD1","XAF1","MICALL2","DYNLRB1","CYP1B1","SNX16","SLC35D2","APLNR","MYL12B","RACK1","RARRES2","FBRS","MRC2","OAS3","HSPA6","IFI6","CDC42EP4","OR6Y1","BID","MOB3A","GABARAPL1","TAPBP","LINGO1","UNC13B","HEATR1","PDHA2","USP18","EPAS1","KIFC2","HABP2","RPL19","DUX5","PAPOLA","PCMTD1","CTSD","RAPGEF3","CYTIP","PPP1R36","PABPC1","TTYH2","LTBP3","C1orf198","HSPB1","SEMA4B","TUFT1","ANKS1B","CHCHD2","RPL8","KIF13A","THBD","PDLIM2","FNBP1L","AKIRIN1","BBX","AGTR1","PRKCA","RPS6","GSAP","IGFBP2","SLC44A4","ZC3HAV1L","GAS2","RNF149","QPRT","XAGE5","STN1","MYO1B","POR","DYNLL2","CMTR1","FKBP5","APBB2","AMN","PLXNB2","SEMA6A","TMEM132B","TCEAL2","RPS25","PAX8","AK3","PHF10","HSPE1","LGALS3BP","CLDN5","HSPA12A","RAB23","INPP4B","CRYZ","IP6K3","TUBAL3","RIPK4","SLC44A2","RAD21","KCNE5","ITGA6","ROBO3","GSTP1","ST6GALNAC6","MST1","SRSF1","NOTCH3","MS4A4A","FBXO2","CAMK2N2","ACVRL1","RGL1","GCNT3","ART1","SP1","GLUL","CCDC90B","GPHN","YIPF6","NCKIPSD","DDX6","ARF1","EFCAB14","CFI","KYNU","TMED7","SIPA1L1","IQGAP2","OR5M8","CCDC91","PGM1","ADD3","RPS27A","NR5A2","PLCE1","AFM","TNNT2","FAM3B","KIAA0040","SERPINA1","SON","EPS15","MAGT1","RBM4","REG3G","CFB","MPZL2","ATP10D","NNAT","PTK6","EMP3","IFNGR2","CASC4","SLC15A1","PDGFRB","CD47","GOLGA8A","SP7","DIDO1","KHDRBS3","RIBC1","INO80D","DEK","EIF4G2","SMAD2","CCND2","ABTB2","FAXC","SPATA32","LEPR","SORD","CYS1","CALCOCO2","GNRHR","S100A4","TEX13A","GIMAP8","BCL6","AGPAT1","PLAUR","YAE1","BMP7","PLEKHA2","EEF2KMT","CASP12"],[-1.56249852075715,0.874176849054845,-1.47762213117364,-1.30409399277973,-1.83552029050978,-1.60514249277753,-1.20207086663023,-1.2137393337074,-0.805802542971358,-1.30356927540952,-0.852782990212227,-1.4013104469036,-1.64550767328968,-0.956953542611636,-1.086491481723,-1.30169001327672,-0.901470469708215,-0.940006043884212,-1.12137235680564,-1.02374901710439,-1.13141813116568,-0.99123846029529,-0.923867937468478,-1.01056769677636,1.09294503805111,-0.926059079458048,-1.0659359365368,-0.84324656851642,-1.00616776161318,0.651604717376768,-0.797081167279542,-1.04162952487471,-0.88791515592396,1.1024078198614,-0.838797642098771,-0.81400141120858,-0.860889744129239,1.47026872814048,-0.78426630017022,0.964154172166132,1.56117228131789,-0.789868763251719,-1.34130948870015,-1.06919100231578,1.02674710277253,-0.885784795824542,0.920036360469357,0.973822017526702,-0.712829970711383,-0.996720314314767,0.5337311318626,0.612958891301221,0.644772540708809,-0.832611942452358,-1.02216288003772,0.87925105541344,-1.1318973770805,-0.70434664180581,-0.939753418935322,-0.763393664428091,-0.764516010472277,-0.821652457039852,-0.911012527421489,1.1765467435199,0.795304512647088,-0.781651942157289,0.754309238157145,-0.713472448632328,-0.825747679657653,0.892265125657326,-0.741174148903967,0.976880154444707,-0.881970565173999,1.09270126754404,-0.981772304734111,1.09386022681494,1.23660030962494,-0.811234361292339,-0.840310267111859,-0.7744300036272,-1.12081300302865,-1.25582724476088,-0.927186239668069,-0.883073152146615,-0.788758752851216,-0.569166552567057,-0.861770334620775,-0.851369080700845,-0.98124642433439,-1.10325075149014,1.34389337643596,0.497610738087992,0.876742507415166,-0.6584659375817,1.15418612517901,0.520963029937888,0.960928035143677,-0.833736091698073,1.34160522041227,1.0324491671457,-0.709046944463897,-0.742635624334143,-0.788046191989283,-1.37272028424048,-0.926915922487862,-0.900461490305416,1.5622080234406,-0.782747371774866,0.493086201324418,-0.775955580743912,-0.941280762543798,-1.21637060006762,-0.594156387076326,0.77006605675877,-0.613563939191516,-0.55100101377511,-0.7893701829176,-0.679064603794112,0.461918168638402,0.500598236246266,-0.713842403133278,-0.959632586048874,-0.546525121165366,-0.805786612799849,-0.720375698073267,1.11753081799947,-0.717645593839002,-1.13894595028276,-0.67137599627831,-0.867313915771779,-0.721587410724099,-0.715899392170465,0.710160860131742,1.19468013318682,-0.553154399311457,-0.484161238679374,0.678308012229188,-1.08400000568715,1.72209464908707,-0.882613515436784,-0.618751330780871,-0.693498028633983,-0.761486774376982,0.465732474676908,-0.864099637228863,-0.756411861613087,0.66615715257724,-0.74238705987103,0.923151957570105,-0.590746710058681,-0.690504320616308,0.946258846129916,0.733933634326315,0.784866558046845,-0.7361387635295,-0.69554484310248,0.756434298659609,0.877341385746389,-0.616067856159657,-0.684271406811444,0.707161253091701,-0.875934581822635,-0.615236817901651,-0.849736312800719,-0.781859257402857,0.859319760067391,-0.7100927764895,0.384173758658823,-0.708907162129192,-0.642714948285138,-1.36738144055402,-0.699216440857517,0.852642129424124,0.983220411716538,-0.871400627956338,-0.548424283946808,-0.621627381985694,-0.62847161305031,0.779718915430365,-0.561081895248354,-0.600343175179998,0.437634010810804,-0.713642071987151,0.605271643854184,-0.655097662233263,-0.631959989428079,-0.779109502706092,-0.686217518494506,0.892804635927639,0.731032303953564,-0.77974781286807,-0.744402549597145,0.707459278821574,-0.715613614704877,-0.734966725334967,0.66877073009347,0.61201710886786,-0.838169726075789,0.928752304710544,0.45561244965277,0.90452898106753,-0.752050919286924,1.68146883861791,-0.684863020331445,0.764898475260544,0.83204138176392,-0.644785765508747,-1.27619972921325,-0.690028718368109,0.375548001756279,-0.720766956261333,-0.742210228408041,-0.667786223471267,-0.717261537146912,-0.680032789207958,-0.637470935387073,-0.659925737711665,0.941312037465394,-0.688078676590772,-0.655996016937747,0.623685371259747,-0.590475011932329,-0.489571074734423,0.723488398575949,-0.650377056120073,-0.428946603664222,-0.709419112162044,0.504197093838336,0.619224658743255,0.672601619807547,0.430076670473444,-0.712906514889132,0.663038191113759,-0.440778514881783,-0.546068821198189,-0.633050567977538,-0.492585189908457,-0.658775896856767,0.905356666547005,0.696612592046851,0.703549508971586,-0.550537266442928,-0.647980745857819,1.04084602415744,1.20601357991416,0.809349403864667,-0.631890707518405,-0.591699588832158,1.04428202850781,-0.557518546879141,-0.486173796023627,0.815338480976841,-0.699488793858036,-0.523377861232367,-0.703269124487788,-0.433841818140032,0.726165073161055,0.585441058044922,-0.387221273265818,-0.680732479339207,-0.509181972605791,0.561379182911113,-0.685089992686144,-0.529232620213698,-0.571845141465189,-0.547433101096835,-0.682238628022832,0.392581745551735,-0.516786955090886,-0.76386997504184,0.63826795171665,-0.672758903777053,0.363366013754601,-0.526844942454028,0.630685072076731,0.641913299175616,0.680132275057415,-0.545859735748068,-0.69722316541222,-0.70181863420699,0.697212227650304,0.65974751168084,1.14966162270477,0.664475080504354,-0.527031000635695,0.592397897048167,-0.564957429538378,-0.554513083234782,-0.500682739423136,-0.713196509699539,-0.67079443157658,0.82802191074932,0.434603331609038,-0.593770110665045,0.614270146918598,0.381016594264993,-0.622272233618233,-0.517370250254841,-0.59355180736055,-0.578339184499286,0.676322474094283,-0.891132082098413,-0.675109048024451,0.868261638070118,-0.558205419579645,-0.596601710774049,-0.647047652369105,0.589482772657197,0.556585013831659,0.60682501897138,0.576575140419536,-0.4914547843142,-0.486889196838032,-1.30388030119895,-0.705000408060799,0.460442320676906,-0.539463980782894,0.569674287675971,-0.523731409944973,-0.76900727883282,0.384794181652305,0.666300812708116,0.615358628065451,1.33145303301945,0.639195925166984,-0.455547417482489,-0.658677348734267,0.578166583706485,0.617883159900218,0.592359393946417,-0.563506501925077,-0.554659191112375,0.853052198166084,0.747956871840044,0.528479906358664,0.535251114538097,-0.37905732097166,-0.486300730061423,-0.572132002887689,0.819180620506187,-0.607952245113457,-0.507008442855944,-0.652782919613788,0.970049747568078,-0.648334996412562,0.337877787742852,-0.617647247394029,0.551714542405181,-0.568916374639628,-0.413883465314362,0.635250117117307,-0.569446774559509,0.601314830843712,0.647138368303711,-0.606881846425249,0.661685858418099,-0.580306494784816,0.689656748856471,-0.698779593351736,-0.645267423469263,0.349182118685511,0.655678821790262,-0.518748855834277,-0.614160743879719,0.619701004916179,-0.415299578808921,-0.573964362802021,0.89942975801579,0.736585706121343,-0.77858703734892,-0.503757038729696,0.397003634626989,-0.619027721147576,-0.61450994200573,0.65251583625515,-0.491510568727673,0.450683578894109,0.390020913236653,0.922907866649615,-0.483787972781601,0.607061455802247,0.71170005331922,-0.492301772464915,-0.515333861293501,0.91447107731338,-0.609559986500386,-0.55934444370099,-0.561743260243948,-0.47961527800264,-0.505709978416764,0.894683794899518,-0.632877250264944,0.58790858768892,-0.549141444930855,-0.568683561284326,-0.456336245677243,-0.497188503818442,0.30369185120579,0.861650896622878,0.591420052859322,0.652475484014158,-0.519906383277503,-0.5878698224623,-0.419272205217071,0.55590215314303,0.583531126275703,0.717907012854006,-0.314553078328772,-0.536503531369111,0.574208842098428,0.420593181647207,0.587564225745062,-0.613237408527661,-0.610741925878184,0.711452913916705,0.540486460597098,0.597925076714159,-0.571113292122172,-0.488485313139309,-0.422926675893358,0.523960183111378,-0.510455543896246,-1.00327862517998,0.535855668690388,-0.528115697366287,-0.589970194606288,-0.45753704270086,-0.571615739688211,-0.585540598258007,-0.306072375869779,0.75081721052663,0.453301910275029,-0.564754050441239,-0.473466581234135,0.631665799863034,0.354894161149527,0.595272153056499,0.594727417050094,-0.438815893518133,-0.540194550428,0.678997397211778,0.523053953361881,-0.398570207818098,0.618427847141016,-0.517448029364178,-0.590036747672641,-0.506249580881025,-0.518328431262111,-0.501463244882021,0.599967116818071,-0.413904520984138,0.570453010157943,0.731398239161646,0.639400853518763,-0.497010566672212,0.574586849382046,0.522014674513303,-0.481410437768962,-0.581088053275515,0.444905535981552,0.628647857618839,0.559453969718085,-0.435545276376505,-0.528846335551187,-0.577118359756519,-0.418007343557207,0.566212189586739,-0.345876026774068,-0.517555727336937,0.559056069330028,-0.490859003087733,0.490404085459299,-0.607093401853886,-0.486333371064092,-0.641110436419534,-0.709060441208714,0.324589084990648,-0.487904549397583,0.521988852953564,0.555772723522887,-0.445693575677379,1.12723790263518,-0.442756395458214,-0.458973079424373,-0.567404948145118,0.476637354579036,-0.394527678179252,0.59159542867118,-0.522035396899225,1.02556251839776,0.585503643206483,-0.629690962889317,0.405287495658321,0.472645404499087,0.542643068962041,-0.441275410080694,0.478378823608514,0.319216474054276,0.573570099234319,0.384088566833866,-0.509492509501395,0.493287735829067,-0.636036142520464,0.645895270700231,-0.526067960538473,0.468160520298035,-0.504997385161305,-0.525463640158632,-0.592992761706671,-0.535846750707426,0.712042622161186,-0.434540057282118,-0.646152917407207,-0.487924885032408,0.497110165613,-0.583695425839421,-0.370184126042647,0.335552650074036,-0.565925716811994,-0.464760198793671,-0.651870819842384,-0.577589969260389,1.55538571056931,-0.481537848874399,-0.458502439171391,0.532082023854819,0.506702327899102,-0.502606878757105,0.387317029179223,0.458868543718455,-0.512108009148969,-0.521315159977592,-0.497813795659244,0.560070817196607,0.456492660108026,-0.761867617704173,0.491070862867046,0.720079168468414,0.324548872998523,0.452684638881702,-0.421490680956505,-0.447783123037846,0.408082312847376,-0.545407176219342,0.542540922031689,0.585096728248663,0.333952259977047,0.56331863089604,-0.4487331506135,-0.492389463893221,0.773426999001942,0.652090633712106,0.515277145599244,-0.598320447436784,-0.401565215009538,0.330250689691376,-0.464020145433627,0.524429248358027,-0.517296376506337,-0.431535968290726,-0.472077858155113,-0.424040888813133,-0.580330907954721,-0.461416583774399,0.378128797398621,0.508667604690329,-0.610884286756257,0.529309809261507,-0.60022184381673,-0.623853872179907,0.475376814128335,-0.551430829643322,-0.541389769224219,-0.508259580558843,-0.566033435013122,0.491761416293503,-0.408600222276864,-0.485003174507811,-0.57688559703946,-0.467493661345601,-0.568568057528285,-0.382383880756676,0.583741508559084,0.539880148180261,-0.583894376464912,0.301715835359359,0.611585668973117,-0.4379742118889,-0.507288654778204,0.606608771976448,-0.414114171118225,-0.480072460402093,-0.514327133066605,-0.513621436784121,-0.516760571892982,0.508225016909642,-0.546402664057014,0.517629445707849,0.521662604621714,-0.425319273858447,-0.394356744009738,-0.531369566713563,0.443827298346779,0.558260599431485,-0.406279277961116,-0.415256789972567,0.568325611436906,-0.344463200976722,-0.608107373633055,0.642613765494911,-0.567484571751473,0.474720761559885,-0.475495853833589,-0.49119929440435,0.637321517577762,0.486013343217776,-0.473409477429801,-0.402249454332638,-0.426852624043058,-0.449001957165483,-0.44813302735935,0.494359272817762,-0.430118888358222,-0.2860517312005,-0.45926287933573,0.49129656025323,0.63280879195853,-0.453988963499427,0.500586794197375,-0.444101360803993,0.518598271333065,-0.472179780081648,0.453160064556048,-0.410522514466648,0.297450183459008,0.51932731084084,-0.456435213463581,-0.591342372941659,-0.454589258462491,0.500294341332921,-0.488107513328808,1.37984762015011,-0.369986714231628,-0.448832650120032,-0.437918701244286,-0.424730484796453,0.474248631449669,0.599264339379717,0.575175229069012,-0.449524498011955,0.518499809545981,0.478244593689057,0.545280610716371,0.402333505612873,-0.37719150204037,0.699266256458477,-0.580473891084847,-0.442430476358736,-0.430900176663837,0.507867694502529,-0.469785720873992,0.51267961861878,-0.545962362827135,0.473220629957039,-0.401999294080723,-0.323329253285964,-0.383171103190262,-0.50717591706796,0.559148080859677,0.521140807057204,0.51113610466091,-0.501803474676755,0.558958995802002,-0.457564793329646,-0.384575788086477,0.537488882634007,-0.595734264735871,-0.561378555253352,-0.500864027922159,0.497073647606035,-0.428207073295515,0.549465754576472,-0.469990394705542,-0.538118568527603,-0.428026701406046,0.521425903878561,-0.454378798902407],[6.03607587098281,8.7707573607221,6.14540087623084,5.9179147856783,7.79119634294983,7.46118976457571,6.54867918165769,7.67656236396124,8.15029623009923,6.22360174403137,7.99290093255035,8.25854461209406,8.00795308418871,8.11005262952887,7.91591467162477,6.14341387973405,6.90149662779019,6.8856246397754,6.16622143807882,5.79863845140217,5.45361983066653,6.70326945781006,6.65352200186479,6.78822800610747,6.25790525283165,6.19010698384545,6.58066046043538,6.69972382193374,5.82121287884198,9.37655874563303,6.70290350680832,7.09377351943063,7.56997885659717,5.16355786023101,6.74932819377596,7.04002884254463,6.54840743994476,6.14617918244257,7.65549381264792,8.22787221861891,6.92582673431591,6.85651139548215,8.19644625520374,6.89301173074599,6.18303980678231,6.73082514547056,6.74149520633951,6.11120334200024,7.01042026109081,7.19989267577478,8.49236839148033,8.41885309355579,7.88647913102731,6.93816272058297,7.57196584545573,6.29047660733209,6.18609798022319,7.19834880719141,6.30963701780881,6.52310383245158,6.5043282760375,7.1851876813566,6.07255259025187,9.00698017621485,7.5712089715285,6.27924740092712,6.24673951740996,6.80032524256874,6.55446333699292,5.6249225921377,7.38525264645698,5.27825743572647,5.46226337988513,7.23125966764693,7.69258357073369,6.56569256195797,5.45748324402768,6.06955671251616,6.65419361938611,7.08850278284607,6.38137350177381,6.48287766487596,5.86917639257976,5.99416095932326,6.18441997624373,7.69433866932469,8.3296720555673,6.36655708794748,6.54645421930914,6.43831721653961,6.52897377930419,9.45083332330382,5.50101036662579,6.94035346167033,5.13268090498531,8.47994183357567,5.06636439612412,5.94143460375673,6.06454452252434,7.2048000931803,6.82993084112248,6.84164316824573,6.13032200800464,6.62498141324244,5.74256954516255,5.34443611185949,6.40603741136833,6.41691736345046,8.38515341210794,6.32068181672877,7.98560743307625,9.47752914808076,7.55676411548209,6.24563798517122,7.1652991369027,8.59311358918091,7.20059578342461,6.83585788226192,8.80705072799663,9.56327235568418,6.42465982988462,5.67284719558644,7.46750993976911,5.53122534168935,6.33914127584101,7.4155885140732,6.73867222844993,5.40459428485918,6.36440151762197,8.58570880142001,5.82029247112244,5.7836370453235,6.37142549992075,6.83835026219415,7.92770580154915,8.31144494167078,5.85952214321574,6.38642907109122,7.13366155092816,5.81823690779061,7.44610530535766,6.89711719819346,6.1021529033768,8.57493280381496,7.211572116285,5.80563203298529,6.3619251779724,5.83428342738772,5.96174777492098,6.47939459018656,5.8159441544578,6.1391961200179,5.5947222509015,4.89347096679076,6.29194050053125,6.45633439519715,5.74169398147101,7.33920767569676,6.43740435851212,7.41507162600446,5.81571592370387,6.35494710058494,6.37802524384991,6.18139735840878,6.05628344429213,5.37352365943499,6.06670931858046,9.52326201245695,6.57509885670338,6.8505347986913,7.4860767483247,6.54678530508565,4.61347889094207,5.27447788819006,6.57029139086977,8.09399485191145,6.82485927367352,6.45999949001789,5.07251515422558,7.27350202429976,6.13429548664266,7.92710047508201,5.7011598755985,5.44985006364349,5.77796374239821,6.40328390787182,5.1208195652882,5.488668004254,5.5971646604844,5.13751143933812,7.57288382264347,5.12700970162295,5.56410549772395,6.30092773779456,5.48810148306166,5.36071326211072,6.08681270057418,5.04201696439427,5.19190857824587,9.04882786258449,7.54317976695539,5.46795376274696,6.0000378424704,5.74504549055471,5.91669312987683,7.84188298429205,5.98472044948411,7.91527876527223,7.40247003230204,9.16154592988948,5.65130115025368,5.52938021133496,6.34997318148344,6.17277705840123,5.74207255308778,6.49718589487399,5.77835410037342,5.9411234035401,6.71525293553393,6.35400558473409,5.86407990069879,6.1699098388575,8.01255337828039,5.41164491303214,5.8946361455819,7.90152072056732,4.8779757246418,7.29366336062101,5.62957629686238,8.72746008990771,9.27280462618927,5.45562567103743,5.48942368194976,7.21898787679795,6.33436146864397,5.7727554369833,7.12278286172251,6.0980720269832,5.62421580795818,8.17993356029216,5.32667464874545,6.18375350404666,5.2071231282769,7.39097822379191,6.49353548460142,5.50633121681524,5.67290280091485,6.27200873548044,5.66429441003133,8.93874112591227,7.05708549357033,6.61267778942672,5.34917590288672,6.87728908700116,4.79283394367755,9.13890928982582,5.60541101467725,5.73112519553153,8.79308524904162,6.26862350686822,6.57802464483799,6.02501231786828,5.75910861729092,6.30985696131203,6.52200267298482,6.08465613011506,5.49030340003227,8.25762372295154,7.85911154968459,6.29963766366895,5.72864790404865,5.86419655759037,8.83349663586259,6.60994336443725,5.61619683353547,5.28596900460224,5.49026943112095,6.38107515485116,9.99888546495783,5.10428588521726,4.86159964343092,5.88926567020948,5.52165961502927,5.87721304168308,7.09831998466349,6.39894013192796,6.09872688366353,6.01598932072302,6.71869277793797,7.01703310421846,5.68507752192831,5.06821767122462,7.54953754124572,5.97495512827015,5.1493864107258,7.94872969723034,7.66936456610411,6.43615846571429,7.09377125096786,5.67676972571767,5.39164911650864,8.62993089149249,7.04822021300573,7.06764273881839,6.13999690962817,5.57081365143456,5.94984840482295,5.55484182118313,5.73637999439922,6.05503699591524,5.935942124857,6.44508795448256,6.22530097019368,9.61892698066348,5.30392209079124,7.04108714612801,6.5243605817677,6.03099355884044,6.32087371291453,6.50469568021567,8.54784490688459,5.27954353069672,4.89206555898874,4.67862777734478,5.4660568188217,6.98462371854093,5.28421494006849,5.32892683244569,6.18515669377247,5.20216744004156,6.55099686067465,6.46767079782342,5.72806700176126,4.35347289779404,5.9246486048226,5.73470411174028,7.93452284907912,7.42835750280966,5.48638471102257,5.27924211185161,6.06908643353845,5.90695719135751,4.87962641568506,8.95967761392666,6.42963052963221,9.54601059066181,5.30434075517956,5.99047497507581,5.62204262442518,7.43122387408258,5.84986625195594,5.9937996953076,5.22315732855155,5.47577720995368,5.47730909691774,5.61314425377121,5.69431599687282,5.1956282912452,5.83268315064414,5.6143217467899,9.02812025256439,4.98831133046312,6.66243997393174,5.54622027101982,7.66143088659506,7.31874146049961,5.90982724026586,6.11644941892117,6.12126273508465,5.32576447516733,6.00624499551216,9.37012559793141,5.32260408727654,5.47959930361248,6.42529135218448,6.73059982217017,7.19709277946591,8.72407585194751,5.67565625462134,6.06134330784872,5.32174536642434,5.66510183916846,6.26514394597421,6.34775412954685,5.1433659543691,5.29441023838547,5.7735103181419,5.82756431294225,5.94814225006659,10.3614424965579,6.65534348532986,5.24275578077823,5.62692738843988,5.61797901390954,6.96494936408684,7.38207642029948,6.15587525386453,9.54734181059211,5.61279511422591,5.08811454643235,5.5104346069541,5.9715080903868,5.30360695622743,6.81060492422844,5.40633420763284,4.85385278032963,6.114058980501,8.33019490396982,5.98164310827907,6.3120920902925,9.55957322997032,5.26375767104972,5.09973599014311,5.68801106247899,5.82168995004695,5.90433302957662,5.00239282052603,5.40543359969177,6.31320457530493,7.01927983311116,7.98612868978865,5.79194712036169,6.28466658410423,5.87964024493205,6.165469868416,5.59666505241361,6.38706965275638,5.55022049762354,5.36818558196139,8.83323661325463,6.78405010173655,7.28266267340107,5.48309437447208,5.7332359641869,6.04485033928024,8.64167602121623,6.36918540782266,6.21630969237399,6.31408495476752,5.60100877788321,6.85353087075935,5.39105697132016,7.33183474357209,5.41953795122717,5.92023480822217,5.35723763035532,5.7113166347309,6.05896029832044,5.76814556553389,4.97153308400397,7.94931472717008,5.47178872468009,5.42545946127598,5.25198211893409,6.71906331310818,5.36000580422667,7.63921884239931,6.41464463788148,5.41244095672695,6.93728427737107,4.87904015950147,5.69592900584769,6.68872984860684,5.67895093481243,5.41006325217904,6.69209023941614,6.83216350047318,7.53744729817044,6.0382698762097,4.97225751244764,6.24900383414608,6.97730384332886,5.15960762259431,6.07514510487749,7.04758216243775,7.28729044880891,8.93695354966684,5.49382836391322,6.09605108064304,5.11344349626765,6.2912520192317,5.98130809211401,7.11917334041287,6.07203400121587,5.18712820269655,6.18680043717208,7.30936157509039,5.30567847174157,5.95952485628334,7.56569411701088,4.89747380072726,4.74482527151253,8.96567077029545,5.97875267024736,5.61997926390242,7.09301890939376,6.1513010707946,8.87021712387692,5.60412536715775,7.3766913689224,5.87741049499465,6.89565031266459,6.79917279232123,5.48916347020603,5.47270295769294,5.84079365964669,6.06071860236742,5.30861810497,6.55632983502499,5.87755971956512,6.86657665952027,6.69030993825104,6.40196140490455,5.48402564860391,5.5916333666957,5.01688501436132,8.28583212734588,8.34070500716283,6.41554527056504,6.30252203540517,5.28708375400007,5.13572051077289,5.78301498523182,6.20170981208331,6.68370130857774,5.03380678328807,5.35885857993735,5.58932594123866,7.28385491302233,6.93206612484305,5.43995429183194,5.41411637767952,5.26542455344142,4.79529336310817,6.65969950775052,8.0922470528481,5.58914344563667,5.3005900780228,9.32672479283972,6.11463773000795,6.46338045201077,6.59286722652993,9.60726578104658,5.82885717873464,5.42261878074265,4.76530646193938,7.98178493719145,5.18450916919969,6.36037214332538,5.70548476894463,7.32897874274616,6.45489024845236,5.23857505738361,5.23171857640441,7.08468017531347,7.84744578364221,6.28420446974832,5.65922349479165,6.17227755855382,6.85534339535504,6.16605062722331,6.04129399115056,6.11593512941914,6.06416309041489,9.56715694938387,5.44442158946774,7.32586164080511,5.37679415178635,5.20466081264722,5.27197422649637,6.18184624448636,5.54107280280549,5.01487977851432,5.91827058954583,6.90840334300954,6.22794140521968,6.72822513219308,5.67849712510364,7.00229877582202,6.35547733508485,5.04241808273083,7.0531828036951,6.21299576782671,4.77564440618554,5.41201122764256,8.69996945251332,6.42952139948781,6.2918531527239,5.75947713889277,7.45463261849476,6.98484862679576,6.46094632577601,5.69711116883294,5.00016211786779,5.4150511249529,5.756357503442,5.2278956357748,5.11953610570329,5.33793693540725,7.37960895802508,6.46603766041006,4.71578861027324,6.12428593081703,4.85904265767107,7.44290035052601,6.71650874315535,5.99497754485396,7.46416665129665,6.30676766880817,4.88727520456092,4.99006690484233,5.61054192327708,6.04746976906338,6.09585506961157,5.03548513660818,5.34347924579858,6.02468640130678,7.56952286050204,6.11447032162511,6.38358227064244,5.87186254224395,5.6124863971687,6.4455636581417,8.31613711815625,5.82292834394081,5.97758768411526,4.76103983929938,5.74440608929386,5.52248917117522,6.81699622847192,5.18408706057333,5.8010198345585,5.68132611273835,6.54193285551738,8.66272744585687,4.93583760062925,6.92924548859851,5.29531703333269,6.60484629170268,5.09642131769293,5.75007861854727,7.63836062916417,6.65741675293386,6.16938102227187,6.3345058702986,6.21408076115439,5.2408705494164,5.49350718795067,5.65905467390549,5.89821564878794,5.23088772064603,5.75712929252331,6.4711338815851,6.42621825792424,6.73467657897053,5.34836623448461,6.78383943424952,6.34934664947994,6.1757968036372,5.74180805886824,5.84673710478085,5.75666014620147,4.88056904036325,5.43264313681425,6.51848848084525,8.03459007093852,6.76465762209759,5.58212984486092,5.3510593662939,4.84442048277158,5.11462568617255,5.48852635583184,5.70548184763339,6.20793125565286,6.46580764072258,4.97992147690934,6.7670512832505,4.48387843147266,5.93239785775622,5.86951469802172,6.09070752389455,5.03366662926713,5.54749417181223,5.50782191921367,6.34379279667361,5.34628621644484,5.55474521053061],[1.15133049921644e-18,1.07606699164099e-17,5.56997564129475e-17,5.93006471191753e-16,1.10269343863768e-14,1.13673995607639e-14,1.59240641455345e-14,2.04453206106056e-14,8.4654573824991e-14,1.41295569414561e-13,1.51947326608596e-13,5.37296079580523e-13,2.37908588103294e-12,2.80716262393301e-12,4.0082350070076e-12,4.10870919668355e-12,4.22172392798328e-12,4.72339932277132e-12,4.78535011324186e-12,1.30050519432819e-11,2.60654883313141e-11,3.17667755386378e-11,3.66879550818864e-11,4.62155003366351e-11,5.26934389342915e-11,2.19897741442199e-10,2.47394771171153e-10,2.74181990576645e-10,6.32864541113879e-10,6.53948587620719e-10,8.17124016162609e-10,8.75715481638775e-10,9.36861317664302e-10,9.40157608209875e-10,1.17881004284113e-09,1.36108561765779e-09,1.53023038791628e-09,1.64223224050972e-09,1.73637412188076e-09,1.77813741954084e-09,3.0900129713237e-09,3.16287932554634e-09,3.32158472673342e-09,3.50692170087597e-09,3.66357319762981e-09,3.91828839844243e-09,3.94142603517182e-09,5.25993767898659e-09,8.16181944240207e-09,8.6092790331402e-09,9.47418141974681e-09,9.77347550426033e-09,1.05094458872244e-08,1.05400911207054e-08,1.08085805648533e-08,1.2390343464156e-08,1.30580748615507e-08,1.37368614539641e-08,1.67019704826922e-08,1.7314008011269e-08,1.8776830435649e-08,2.07186413665971e-08,2.19272486763479e-08,2.32431317089117e-08,2.39596709132937e-08,2.64829015492079e-08,3.10872022452606e-08,3.11766197361411e-08,3.25114058310591e-08,3.26831263503811e-08,3.44952454377255e-08,4.21735014096068e-08,4.40528303072097e-08,4.82771652700756e-08,5.17063366285929e-08,5.44941445428264e-08,5.48309983879101e-08,5.50187560746129e-08,5.70196636455685e-08,6.65198269294457e-08,6.8220199744825e-08,7.04506376526001e-08,7.60727467380178e-08,8.51385588112579e-08,9.51667702583723e-08,9.81250046394975e-08,1.04668564743553e-07,1.17401323948515e-07,1.30918242177916e-07,1.63175086429792e-07,1.71252729748073e-07,1.76269636323841e-07,1.77900138317496e-07,1.8567076584593e-07,1.89093301372413e-07,1.94438875658645e-07,2.42433608880015e-07,2.47854351394278e-07,2.54482530962725e-07,2.66802763627497e-07,2.83010731541936e-07,2.86037731141198e-07,2.96478438362166e-07,2.96705976860339e-07,3.12243955708979e-07,3.21841829798184e-07,3.40130124346847e-07,3.53620639243087e-07,3.57196770916137e-07,3.95672627708095e-07,4.38893293126639e-07,4.51937470682035e-07,4.7489322629201e-07,4.77742306473207e-07,4.9470064767768e-07,5.65083548245589e-07,5.65526455021082e-07,6.01622051211684e-07,7.15055498045791e-07,7.28406814146597e-07,7.65697951982673e-07,9.16145175666561e-07,9.24032312737305e-07,9.44868685677201e-07,9.71897805524488e-07,9.97020592150319e-07,1.02350615801977e-06,1.10043788424269e-06,1.11468696630797e-06,1.14129737655782e-06,1.20360018664003e-06,1.22434045841729e-06,1.26814347815531e-06,1.51232209395283e-06,1.60940080112311e-06,1.63680222940508e-06,1.6494815523777e-06,1.75260857317314e-06,1.86095602067738e-06,2.0908821384967e-06,2.14558542499683e-06,2.19044122012513e-06,2.22425559798401e-06,2.23621226277216e-06,2.25668235906579e-06,2.36388172446011e-06,2.42582260538693e-06,2.48066555868203e-06,2.89541353845529e-06,2.92666567034798e-06,3.00708833893841e-06,3.02575286350929e-06,3.78105809917921e-06,4.10453417453701e-06,4.18660631098189e-06,4.44433378908219e-06,4.51680267692344e-06,4.53766525370653e-06,4.8419314518851e-06,4.87695285666168e-06,4.94390317667725e-06,4.95600887483386e-06,5.02276407571109e-06,5.53064168474296e-06,5.6270546396681e-06,5.86248369483348e-06,6.16816070524244e-06,6.29706226739861e-06,6.71746014761707e-06,7.12311077485401e-06,7.29348772444985e-06,7.59757423601779e-06,8.42827144399864e-06,8.44733728767776e-06,9.05052673458874e-06,9.06549206510187e-06,9.95285990007819e-06,1.02827714854821e-05,1.07165845197019e-05,1.10725456739984e-05,1.16426732382143e-05,1.23834362004248e-05,1.24451391257553e-05,1.25706609503745e-05,1.3532828902174e-05,1.37572608485828e-05,1.37577766793328e-05,1.43442376030922e-05,1.43904648022705e-05,1.43910394134642e-05,1.44094461940511e-05,1.48066646091029e-05,1.49225750672771e-05,1.51983022440932e-05,1.56468768810191e-05,1.57389870004969e-05,1.58187332451073e-05,1.75001155644783e-05,1.78537926757685e-05,1.83457145935516e-05,1.86958687030924e-05,1.9061681679236e-05,1.90710308384079e-05,1.92524579893038e-05,1.96684522903235e-05,1.98017204385213e-05,2.00922470163023e-05,2.01745769560528e-05,2.02406188368791e-05,2.07222037379218e-05,2.07226313032056e-05,2.1552460527116e-05,2.25107384070193e-05,2.33854065907597e-05,2.39308982846666e-05,2.39629819014984e-05,2.42360013234943e-05,2.55508014080942e-05,2.63802939888719e-05,2.63870984102808e-05,2.69033503548967e-05,2.70072041046472e-05,2.78922076333923e-05,2.79148824644402e-05,2.79247839428607e-05,2.79602301342598e-05,2.89712100623928e-05,3.14340880808212e-05,3.171261677675e-05,3.30619857937944e-05,3.34791880679522e-05,3.37585336579562e-05,3.39636717077439e-05,3.41054081189413e-05,3.46955128432217e-05,3.48543917208642e-05,3.52521707802733e-05,3.54503999873971e-05,3.5614524513613e-05,3.6133814695015e-05,3.76297900416481e-05,3.96039361943479e-05,3.98771893589919e-05,4.05162278626009e-05,4.06410952125705e-05,4.08035836889425e-05,4.2072344915414e-05,4.23836644030398e-05,4.37576415990447e-05,4.4058657693518e-05,4.42932112890141e-05,4.47908862570615e-05,4.53454138053082e-05,4.64534685090311e-05,4.65838701656893e-05,4.76683794604766e-05,4.77822533186835e-05,5.03717716108071e-05,5.19615338855661e-05,5.48657570324922e-05,5.49677964386129e-05,5.55338771944963e-05,5.66934269488252e-05,5.7184800959905e-05,5.73971991947118e-05,5.77726917223151e-05,5.91161910271804e-05,5.97252005246534e-05,6.02422103750666e-05,6.05204127896873e-05,6.21964225105881e-05,6.29134290314046e-05,6.38970575746703e-05,6.4379013027297e-05,6.59876109164219e-05,6.62437621002816e-05,6.9523225892729e-05,7.406136807087e-05,7.5044837806488e-05,7.70830246041055e-05,7.84201486096327e-05,7.88726246816439e-05,8.19953287717976e-05,8.36512426660716e-05,8.43653372924888e-05,8.44552545954782e-05,8.45196119749345e-05,8.52746688074115e-05,8.82120625249233e-05,8.86772836586039e-05,8.93861979811195e-05,9.19025973088395e-05,9.38813866902379e-05,9.5002526912387e-05,9.50285064433582e-05,9.58582028023482e-05,0.000100407043030718,0.000102089740057892,0.000104053021286556,0.000104177235404526,0.00010851396596293,0.000109391694874948,0.00011001773782014,0.000111124542965947,0.00011260936285742,0.000113076957002151,0.000115931680233975,0.000117058193305616,0.000117104126679885,0.000117468258120936,0.000119900789317566,0.000121441717374414,0.000122463811566874,0.000122509297941322,0.000123212889044417,0.000123257552527829,0.0001256402963967,0.000126062650585762,0.000127557595709312,0.000128349664012178,0.000128499721538756,0.000128787334980637,0.000130179665896343,0.000130570810624218,0.000131793438179725,0.000133077342971718,0.000134857399719727,0.000134989906917087,0.000136950485108509,0.000139249587148821,0.000141567769751865,0.000141818180639884,0.000142099718059347,0.000143323963381995,0.000145108060117041,0.000145430464917795,0.000147585239482686,0.000147670204425354,0.000148355667157351,0.000149802298971378,0.000152137951334441,0.000155999730555735,0.000156371613503147,0.000157577241978869,0.000164915579714468,0.000166661393137935,0.000168234199346486,0.000168387257419733,0.000175697344392364,0.000175898604320844,0.000177137372878799,0.000178449717176379,0.000180604916351035,0.000180662834323528,0.000190871290818742,0.000191872971385995,0.000193928349902506,0.000200817506800886,0.000201847007912703,0.000203140346248178,0.000203354768735836,0.000206024472757078,0.000206365126573533,0.000209760078299852,0.000211967289339438,0.00021344639043104,0.000215594705174755,0.000218237209825125,0.000221496152416517,0.000223019285393971,0.000224681494933626,0.000225769739903548,0.000232894759630488,0.00023673721540152,0.000244605492143278,0.000246895295910464,0.000249102334731448,0.000249334940778898,0.000249939155985272,0.000251446958770553,0.000253021055854616,0.00025559428965114,0.000263597322445556,0.000263661773388989,0.000266143383905928,0.000270195960436487,0.000272292796431251,0.000273164384580234,0.000274057026537306,0.000275643361448825,0.000277070263551781,0.000286047035623237,0.000286475066290798,0.000287348375260979,0.000288996112433977,0.000289343302044388,0.000295751000503429,0.00029699423758708,0.000299513612597086,0.000300831918203654,0.000302459917713153,0.000305256714458093,0.00030847768918447,0.000310105797072107,0.000310304085734471,0.000320338004045172,0.000322534677622117,0.00032603757513976,0.000328849424992389,0.000329655834936324,0.000340901340607867,0.000341811643627655,0.000344034319863861,0.000348538004879776,0.000348951708727446,0.000353724916177757,0.000360243697717624,0.000362828860107682,0.000365329907042854,0.000377291680046949,0.000382420029402356,0.000382506731434889,0.000384227485441828,0.00039614805642475,0.000400892260050051,0.000408287992726636,0.00041357714283578,0.000418783168190046,0.00042195315035314,0.000434462301461477,0.000435876654684554,0.000437724550585036,0.000440304998825346,0.000445590273041403,0.000451057624376017,0.000451903114061267,0.000452446537270342,0.00045684211649086,0.000465521640906662,0.000471166361428657,0.000474852804183746,0.000486391357948579,0.000493469286055668,0.000496139028677345,0.000507565717237274,0.000523754607071532,0.000524122486012493,0.000529735403717654,0.000537585093019682,0.000543249796490972,0.000543983196708132,0.000544400819366953,0.000545919326431078,0.000546215612002907,0.000557483579887785,0.00056210699998023,0.000573665424711759,0.000574898859527321,0.000606697273779602,0.000612894315794471,0.000617734167206474,0.000618429677572555,0.000620311447745061,0.000623666072488976,0.000626946429588134,0.000631494492009396,0.000631628650179715,0.000641172851948204,0.00064345783006174,0.000644978461053121,0.000649787668687561,0.000649988882361548,0.000650357773211704,0.000651465246314255,0.000653574207999297,0.000656378714625519,0.000677051067645793,0.000677191361390587,0.000677360763582759,0.000680790239686297,0.000687516266386129,0.000695036339060187,0.000695537441870655,0.00069574777377033,0.000696733777518808,0.000701355744639837,0.000701398342332555,0.000701709366114507,0.000705849906921122,0.000711032029478262,0.000712636660219204,0.000714140473533509,0.000720999407280109,0.000721705792415582,0.000721787355144414,0.00072664353144917,0.000730030523652496,0.000767721820998902,0.000783430531749416,0.000793995118551717,0.000795479524063202,0.000799836598924688,0.000800599210735626,0.000802660618488485,0.000805974375845801,0.000806769818141919,0.000817372570902061,0.000822699334396398,0.000823762254163497,0.000825545627955953,0.000830488595137415,0.000849383793631356,0.000853484012613636,0.000855989391054182,0.000856598272897875,0.000862549987651809,0.000867586028214727,0.000868849809407172,0.000870802598891358,0.00087243210421126,0.000872668311629958,0.00088213088502134,0.000885703237277134,0.000890906260064526,0.000891553517812017,0.000892244641491564,0.000893030020284366,0.000893080911382643,0.000898337608291849,0.000911898634623931,0.000916657735268232,0.000917726988230534,0.000921531866905828,0.000922354637434692,0.000926240909879122,0.000926254847151148,0.000929854130859896,0.000933879904160834,0.000934633220801182,0.000939238114436281,0.00094027739945676,0.000941387609425782,0.000945110149038202,0.000946957432239789,0.000950359492711136,0.000952314557910227,0.00095426419409846,0.000956153586636311,0.000962758169693453,0.000964694544768801,0.000975178612713236,0.000983974008281146,0.000990492328111326,0.000991863666809167,0.00099189136538657,0.00100185613631749,0.00100668464650643,0.00101209970131364,0.00101280052311814,0.0010184466431749,0.00103165551041532,0.00104039200977363,0.0010472248789137,0.00105794641120171,0.00105891273495104,0.00106017717656627,0.00107320054370717,0.00107641950064806,0.00107677206308351,0.00108537165532854,0.00108629907566885,0.00109010598011374,0.00109238022161711,0.0010961254165991,0.00110478360128337,0.00111228664640138,0.00111684802471455,0.00113823997847214,0.00115674417248723,0.00115794474295223,0.00116061439815418,0.00117353987490994,0.00117478304859816,0.00117597408718329,0.00117605828565252,0.00117963869852581,0.0011810592261429,0.00118822495686983,0.00119973588958982,0.00120082947704112,0.00121065040201464,0.00121585623588391,0.00121704062543767,0.00123234346167566,0.00124608668280508,0.00125394684612056,0.00125571703238776,0.00125913731517001,0.00125922710493561,0.0012595760167269,0.00128100271091155,0.00128446633500101,0.00129747859566884,0.00129889668771019,0.00129903776290008,0.00130314294779504,0.00130880683439435,0.00131019627758351,0.00131213443781442,0.00132287329690671,0.00132333630270506,0.00133160534044442,0.00133488296697811,0.00135026479535963,0.00135798733949182,0.0013588535793286,0.00136784552731917,0.00137963853901772,0.00140093742547589,0.00141032788825263,0.00141943551727461,0.00142555428711298,0.00142788898377073,0.00143721100844,0.00144558014324099,0.00144862297420055,0.00144952105646248,0.00146535046769303,0.00146818259220979,0.0014868691030932,0.00148796913229573,0.00149213420431763,0.00149473498902222,0.00149821736648811,0.00150127991915979,0.00151286102600454,0.00151430986181432,0.00152833935010856,0.00152847746612933,0.00152883711125896,0.00153063072260806,0.0015312650189156,0.00153305793649304,0.00154238015970355,0.00155734786619564,0.00155836350444132,0.00156647805281182,0.00157353367826428,0.0015928320888456,0.00161464553714615,0.00161647765276187,0.00162543307038109,0.00163411366624768,0.00164131185114965,0.00164554894736983,0.0016583119872134,0.00166213352008328,0.00166623151567715,0.00167661461331146,0.00168023977251535,0.00168637803050239,0.00169853085590525,0.00170948660026744,0.00172127681008508,0.00172668571789904,0.00173014251454786,0.00173173778168178,0.00173745818552515,0.00174147489216673,0.00174345063065167,0.00174609360371449,0.00175313070129038,0.00175808560331727,0.00176106768669374,0.00176811814414172,0.0017916524603117,0.00179955321169292,0.00180791325150327,0.00181725538192791,0.00182375837375597,0.0018288036758479,0.00183026102770874,0.00183721236759692,0.00184417156900428,0.00184517322675801,0.0018520677664487,0.00185682016836673],[2.11626059060973e-14,9.88959368667653e-14,3.41272407542129e-13,2.7250129867439e-12,3.48240285544003e-11,3.48240285544003e-11,4.18143175798672e-11,4.69756797679428e-11,1.72892857941907e-10,2.53903982762963e-10,2.53903982762963e-10,8.23003269897466e-10,3.36384442917434e-09,3.68560401360805e-09,4.56467691295651e-09,4.56467691295651e-09,4.56467691295651e-09,4.62944844376308e-09,4.62944844376308e-09,1.19522929884732e-08,2.28147495722802e-08,2.65411409625319e-08,2.93200566243545e-08,3.53952963203204e-08,3.87423240420485e-08,1.55459245594195e-07,1.68420862551739e-07,1.79990684599619e-07,4.00674299635215e-07,4.00674299635215e-07,4.84501823905965e-07,5.03016445875073e-07,5.08265794014874e-07,5.08265794014874e-07,6.19077354213222e-07,6.94947631615773e-07,7.60193642169975e-07,7.94365021389713e-07,8.17098597714504e-07,8.17098597714504e-07,1.38421154483017e-06,1.38421154483017e-06,1.41986160144388e-06,1.46501654054094e-06,1.49644753212519e-06,1.54143302026581e-06,1.54143302026581e-06,2.01422738494693e-06,3.06168169736311e-06,3.164943158163e-06,3.41460644463463e-06,3.45473563930402e-06,3.58772990536456e-06,3.58772990536456e-06,3.61222762477397e-06,4.0669089859759e-06,4.21088550930113e-06,4.35340086871232e-06,5.20337151597228e-06,5.30414635425225e-06,5.6579822989781e-06,6.14240882192615e-06,6.39753583999921e-06,6.67550006158603e-06,6.77542632395772e-06,7.37548808145441e-06,8.42731540250015e-06,8.42731540250015e-06,8.58212207780508e-06,8.58212207780508e-06,8.93038178015256e-06,1.07665434640275e-05,1.10922612859839e-05,1.19916564166116e-05,1.26721889809356e-05,1.29653814795828e-05,1.29653814795828e-05,1.29653814795828e-05,1.3266815664167e-05,1.52837617348768e-05,1.5480931993946e-05,1.57921118377127e-05,1.68469055155603e-05,1.8630141065592e-05,2.05795341661076e-05,2.09725082590535e-05,2.21139412477155e-05,2.45222015397461e-05,2.70382944884525e-05,3.33257918185113e-05,3.45911695109816e-05,3.51611015313321e-05,3.51611015313321e-05,3.63065356065323e-05,3.6586568131856e-05,3.72289684737662e-05,4.59399192249851e-05,4.6487865642635e-05,4.72489232487459e-05,4.90410159823701e-05,5.15051510541815e-05,5.1545681726533e-05,5.24399284679798e-05,5.24399284679798e-05,5.46605347608262e-05,5.58091950332116e-05,5.84292693048541e-05,6.01842682400665e-05,6.02351729010047e-05,6.61168960900227e-05,7.26783569455923e-05,7.41702022197008e-05,7.70296608358247e-05,7.70296608358247e-05,7.90703704779429e-05,8.88456561516453e-05,8.88456561516453e-05,9.37153807061183e-05,0.000110449034534283,0.000111573713756905,0.000116316479796641,0.000138030036671533,0.000138086487320524,0.000140061542834134,0.000142915628506765,0.00014544631352631,0.000148134383390246,0.000158024599611444,0.000158829931222533,0.000161370669834687,0.00016888072542466,0.000170489408834607,0.000175261242646411,0.00020744770454438,0.000219128860188473,0.000221221042490402,0.00022130744827923,0.000233439841909388,0.000246088004432165,0.000274517889912199,0.000279702168062885,0.000283538732867043,0.000285443177791771,0.000285443177791771,0.00028606950649647,0.000297606232721242,0.000303326838840934,0.000308088605636043,0.00035718520973387,0.000358633611244442,0.000365897127527396,0.000365897127527396,0.000454245940660216,0.000489905471832239,0.000496477487755859,0.000523662175494357,0.000527891297647973,0.000527891297647973,0.000559745547277359,0.000560270440364365,0.000562323451409391,0.000562323451409391,0.000566401389421138,0.000619870273215003,0.000626853886859026,0.00064914646261888,0.000678903963611145,0.000688966080577702,0.000730613224694375,0.000770175877368185,0.000783985952415863,0.000811924488559553,0.000892359233820718,0.000892359233820718,0.000946777327549077,0.000946777327549077,0.00103357919674202,0.00106184057682385,0.00110045553104268,0.00113069145574313,0.00118234241321335,0.00125002241677874,0.00125002241677874,0.0012557680376567,0.00134457798946411,0.0013523085194803,0.0013523085194803,0.00138670173032907,0.00138670173032907,0.00138670173032907,0.00138670173032907,0.00141750678218709,0.00142120130731409,0.00143999996674576,0.00147489868692314,0.00147596008009298,0.00147596008009298,0.00162459406156907,0.00164909830740353,0.00168606289972036,0.00170969533647533,0.00172682077754077,0.00172682077754077,0.00173470308971271,0.0017635405929192,0.00176687098728379,0.0017801091619171,0.0017801091619171,0.0017801091619171,0.00180522599992522,0.00180522599992522,0.00186865932523075,0.00194258160872968,0.00200863158198483,0.0020391831959789,0.0020391831959789,0.00205291216740621,0.00215435449854211,0.00220464207217896,0.00220464207217896,0.00223612350742126,0.00223612350742126,0.00227405747830898,0.00227405747830898,0.00227405747830898,0.00227405747830898,0.00234590225619754,0.00253416654830515,0.00254545680774429,0.00264222765598146,0.00266398682197848,0.00267463623778834,0.00267902353262505,0.00267902353262505,0.00271377966626068,0.00271465497551358,0.00273405127051563,0.00273787311835439,0.00273904006311598,0.00276739853295446,0.00287001315666197,0.00300809897185252,0.0030163893728709,0.00304882386905062,0.00304882386905062,0.00304882386905062,0.00313089786190375,0.00314134732013014,0.00323015747081141,0.00323936874825821,0.00324363950877836,0.00326706857258352,0.00329444288994218,0.00335787497064916,0.00335787497064916,0.00341745368969152,0.00341745368969152,0.00358869586813273,0.00368766391641154,0.00387112285953312,0.00387112285953312,0.00389606181951159,0.00396228851994812,0.00398119969206792,0.00398119969206792,0.00399217987424012,0.00406971800475881,0.00409630190613304,0.00411640174313792,0.00412009521291571,0.00421856989729565,0.00425151374642003,0.00430216782153851,0.00431879795056477,0.00441061191365364,0.00441169054770028,0.00461338056005145,0.00489684175003835,0.0049440830240898,0.00506022526874308,0.00512968239001302,0.00514098480238757,0.00532564006414987,0.00541308358087551,0.00541308358087551,0.00541308358087551,0.00541308358087551,0.00544247808107303,0.00561047031581528,0.00562061086527172,0.00564607458794143,0.00578514260662253,0.00588953504693946,0.005921081277747,0.005921081277747,0.00595260008685798,0.00621408032978998,0.0062970184966581,0.00638293921323531,0.00638293921323531,0.00662656215403528,0.00665804219700803,0.00667404633291085,0.00671901389558248,0.00678646786453192,0.00679237760345271,0.00694117333674494,0.00696510984684172,0.00696510984684172,0.00696510984684172,0.0070864836284443,0.00715455194570228,0.00716961099055069,0.00716961099055069,0.00716961099055069,0.00716961099055069,0.00728515548286355,0.00728665905791477,0.00734995663552624,0.00735167703192263,0.00735167703192263,0.00735167703192263,0.00740747552494982,0.00740747552494982,0.00745383134517394,0.007503357794979,0.00756478499708224,0.00756478499708224,0.00765132786255168,0.00775620200418935,0.00784364840134791,0.00784364840134791,0.00784364840134791,0.00788753823630075,0.00795582552277975,0.00795582552277975,0.00803055037734445,0.00803055037734445,0.00804402807675299,0.0080985766393909,0.00820072634451132,0.00837978608688439,0.00837978608688439,0.00841984675817905,0.00878641527748301,0.00885376609037103,0.00889404074319573,0.00889404074319573,0.00923769213148981,0.00923769213148981,0.00927624515921708,0.00931842116880402,0.00938068801610386,0.00938068801610386,0.0098828315395473,0.00990678957035385,0.00998486554498029,0.0103106888058857,0.010334679254717,0.010354193917267,0.010354193917267,0.0104495795910416,0.0104495795910416,0.0105923076901911,0.0106744403982143,0.0107195576571392,0.0107979462556326,0.0109005928092272,0.0110333896411057,0.0110792364454772,0.0111317265724393,0.0111555741644277,0.0114767790261877,0.0116349378510571,0.0119895828028949,0.0120696341333251,0.0121217193302514,0.0121217193302514,0.0121217193302514,0.0121627540767409,0.0122067717261514,0.0122986351782136,0.0126207475433933,0.0126207475433933,0.0127064455573373,0.0128665076393344,0.0129328524320486,0.0129408107035291,0.012949722891471,0.0129912836584381,0.0130251368653332,0.0133987231386544,0.0133987231386544,0.0134054580854621,0.013430351603227,0.013430351603227,0.0136931968268351,0.0137162087464526,0.0137978940179124,0.0138239787212534,0.0138641290460984,0.0139575215633189,0.0140697975307686,0.0140832083947785,0.0140832083947785,0.0145027902767347,0.0145663634136907,0.0146884722270685,0.0147789273368829,0.0147790339072306,0.015246003751127,0.0152496112172814,0.0153116097661444,0.0154556177304077,0.0154556177304077,0.0156293694333254,0.0158792311936394,0.0159549217168404,0.0160265609101544,0.0165119008831975,0.0166607967547504,0.0166607967547504,0.0166961830021424,0.0171735788328852,0.0173383544281882,0.017616764305888,0.017803188436685,0.0179851715292085,0.0180790695958999,0.0185717478213102,0.0185889763103406,0.0186245716766286,0.018691099730736,0.0188718774395715,0.019030708928069,0.019030708928069,0.019030708928069,0.0191717236146541,0.0194914653337252,0.0196829747486821,0.0197919940900259,0.0202270578064544,0.0204750766297725,0.0205394853290952,0.0209653156146929,0.021552338736903,0.021552338736903,0.0217345233386924,0.0220074645763804,0.0221145135775891,0.0221145135775891,0.0221145135775891,0.0221145135775891,0.0221145135775891,0.0225211113888294,0.0226580894005189,0.0230725238798508,0.0230725238798508,0.0242956483427949,0.0244904574317786,0.0246046664577081,0.0246046664577081,0.0246262304989243,0.0247060475828014,0.0247825856392677,0.0248607413682084,0.0248607413682084,0.0251824747685041,0.0252183334208206,0.025224147005569,0.025262832684604,0.025262832684604,0.025262832684604,0.025262832684604,0.0252912579310212,0.0253464225914531,0.0259928354810328,0.0259928354810328,0.0259928354810328,0.0260700112409872,0.0262728409406308,0.026405491885718,0.026405491885718,0.026405491885718,0.026405491885718,0.0264305734806368,0.0264305734806368,0.0264305734806368,0.0265321618386854,0.0266724076200815,0.0266781557056806,0.0266801139106086,0.0268023704543626,0.0268023704543626,0.0268023704543626,0.0269282958700951,0.0269993783807978,0.0283363349232546,0.0288581895873467,0.0291850481672769,0.0291850481672769,0.0292560916352516,0.0292560916352516,0.0292732238659461,0.0293067905677206,0.0293067905677206,0.0296333830882658,0.0297476895752048,0.0297476895752048,0.0297536356616831,0.0298732110904517,0.0304932099819101,0.0305730735031764,0.0305730735031764,0.0305730735031764,0.0307258358973409,0.0307879390327644,0.0307879390327644,0.0307879390327644,0.0307879390327644,0.0307879390327644,0.0310621605317572,0.0310903792275083,0.0310903792275083,0.0310903792275083,0.0310903792275083,0.0310903792275083,0.0310903792275083,0.0312142600718572,0.0316256769868348,0.0317081574636568,0.0317081574636568,0.0317486902447323,0.0317486902447323,0.0317639745251591,0.0317639745251591,0.0318280237976457,0.0318729002440566,0.0318729002440566,0.0319255454775928,0.0319255454775928,0.0319255454775928,0.0319927617853981,0.0319963686801462,0.032052399698208,0.0320595126171207,0.0320664170963872,0.0320712756860621,0.0322339852771136,0.0322400916861733,0.0325313213798221,0.0327652649387966,0.0328503697066136,0.0328503697066136,0.0328503697066136,0.0331207151828271,0.033220593334712,0.0333028379524766,0.0333028379524766,0.0334286924074963,0.0338018893706665,0.0340274831524004,0.0341901252208043,0.0344295347746725,0.0344295347746725,0.0344295347746725,0.0347840901433006,0.0347840901433006,0.0347840901433006,0.0349689374953925,0.0349689374953925,0.0350301363994243,0.0350419561143876,0.0351008384712683,0.0353165693481557,0.0354946889713608,0.0355784810091475,0.0361972128794056,0.0366968660693188,0.0366968660693188,0.0367181639457349,0.0369523544420154,0.0369523544420154,0.0369523544420154,0.0369523544420154,0.0369830487831903,0.0369830487831903,0.03714415464664,0.0374109264703268,0.0374109264703268,0.0376530711327091,0.0377241546984315,0.0377241546984315,0.0381341837862967,0.038494654313681,0.0385871112724285,0.0385871112724285,0.0385871112724285,0.0385871112724285,0.0385871112724285,0.0391782210137523,0.039218896517697,0.0394671291237462,0.0394671291237462,0.0394671291237462,0.0395265190155455,0.0396031906428026,0.0396031906428026,0.0396031906428026,0.0398105475941436,0.0398105475941436,0.039993852553446,0.0400268904013453,0.0404221778558718,0.0405472201974659,0.0405472201974659,0.0407493819086769,0.0410342006240852,0.0416003728880004,0.0418116724418898,0.0420139198760461,0.0421271918833178,0.0421284549128246,0.0423355377341918,0.0424938541289263,0.0424938541289263,0.0424938541289263,0.0428895015074292,0.0429040766731448,0.0433444700803928,0.0433444700803928,0.0433970234328518,0.0434039870982898,0.0434364880337824,0.0434567341638995,0.0436962787598259,0.0436962787598259,0.0438244757864364,0.0438244757864364,0.0438244757864364,0.0438244757864364,0.0438244757864364,0.0438244757864364,0.0440224995582467,0.0443409900543899,0.0443409900543899,0.0445029877723864,0.0446344483644687,0.0451122444145933,0.0456412837717602,0.0456412837717602,0.0458237504090106,0.045997922357272,0.0461298977614398,0.0461783743535951,0.0464655985319658,0.0465017903084488,0.0465455949690908,0.0467645723934415,0.0467946776645525,0.0468945757619734,0.0471611716954599,0.0473937755648806,0.0476487786839969,0.0477227468742021,0.0477227468742021,0.0477227468742021,0.0478087109403261,0.0478303970776243,0.0478303970776243,0.0478315149476544,0.047952820566099,0.0480168966932759,0.0480269809334088,0.0481478216406949,0.0487165146050139,0.0488590658554321,0.0490136481945156,0.0491943610827937,0.049297798041189,0.0493284867306661,0.0493284867306661,0.0494433389879925,0.049512597198597,0.049512597198597,0.0496251568733143,0.0496800749850783]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>logFC<\/th>\n      <th>AveExpr<\/th>\n      <th>P.Value<\/th>\n      <th>adj.P.Val<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"targets":1,"render":"function(data, type, row, meta) {\n    return type !== 'display' ? data : DTWidget.formatSignif(data, 4, 3, \",\", \".\", null);\n  }"},{"targets":2,"render":"function(data, type, row, meta) {\n    return type !== 'display' ? data : DTWidget.formatSignif(data, 4, 3, \",\", \".\", null);\n  }"},{"targets":3,"render":"function(data, type, row, meta) {\n    return type !== 'display' ? data : DTWidget.formatSignif(data, 4, 3, \",\", \".\", null);\n  }"},{"targets":4,"render":"function(data, type, row, meta) {\n    return type !== 'display' ? data : DTWidget.formatSignif(data, 4, 3, \",\", \".\", null);\n  }"},{"className":"dt-right","targets":[1,2,3,4]},{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false,"rowCallback":"function(row, data, displayNum, displayIndex, dataIndex) {\nvar value=data[1]; $(this.api().cell(row, 1).node()).css({'background-color':isNaN(parseFloat(value)) ? '' : value <= 0 ? \"#DADAEB\" : \"#C7E9C0\"});\n}"}},"evals":["options.columnDefs.0.render","options.columnDefs.1.render","options.columnDefs.2.render","options.columnDefs.3.render","options.rowCallback"],"jsHooks":[]}</script>
</div>
</div>
</div>
<div class="section level2">
<h2 id="gsea-and-visualisation-with-visse">GSEA and visualisation with vissE<a class="anchor" aria-label="anchor" href="#gsea-and-visualisation-with-visse"></a>
</h2>
<p>Here we try to do GSEA with the DE genes using limma fry. We select
the following gene sets to conduct gene set enrichment analysis:</p>
<ul>
<li>MSigDB Hallmarks - genesets from the hallmarks collection of
MSigDB</li>
<li>MSigDB C2 - genesets from the C2 collection of MSigDB which contains
curated genesets such as those obtained from databases such as BioCarta,
KEGG, PID, and Reactome, and from chemical or genetic perturbation
experiments</li>
<li>GO BP - biological processes from the gene ontology database</li>
<li>GO MF - molecular functions from the gene ontology database</li>
<li>GO CC - cellular component from the gene ontolgoy database</li>
</ul>
<p>FDR &lt; 0.05 indicates significantly enriched gene set.</p>
<div class="section level3">
<h3 id="load-gene-sets">Load Gene sets<a class="anchor" aria-label="anchor" href="#load-gene-sets"></a>
</h3>
<div class="sourceCode" id="cb67"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://davislaboratory.github.io/msigdb" class="external-link">msigdb</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">GSEABase</span><span class="op">)</span></span>
<span></span>
<span><span class="va">msigdb_hs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/msigdb/man/getMsigdb.html" class="external-link">getMsigdb</a></span><span class="op">(</span>version <span class="op">=</span> <span class="st">'7.2'</span><span class="op">)</span></span>
<span><span class="va">msigdb_hs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/msigdb/man/appendKEGG.html" class="external-link">appendKEGG</a></span><span class="op">(</span><span class="va">msigdb_hs</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/msigdb/man/listSubCollections.html" class="external-link">listSubCollections</a></span><span class="op">(</span><span class="va">msigdb_hs</span><span class="op">)</span></span>
<span></span>
<span><span class="va">gsc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/msigdb/man/subsetCollection.html" class="external-link">subsetCollection</a></span><span class="op">(</span><span class="va">msigdb_hs</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'h'</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/msigdb/man/subsetCollection.html" class="external-link">subsetCollection</a></span><span class="op">(</span><span class="va">msigdb_hs</span>, <span class="st">'c2'</span>, <span class="va">sc</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"^CP:"</span>,<span class="va">sc</span><span class="op">)</span><span class="op">]</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/msigdb/man/subsetCollection.html" class="external-link">subsetCollection</a></span><span class="op">(</span><span class="va">msigdb_hs</span>, <span class="st">'c5'</span>, <span class="va">sc</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"^GO:"</span>,<span class="va">sc</span><span class="op">)</span><span class="op">]</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/GSEABase/man/GeneSetCollection-methods.html" class="external-link">GeneSetCollection</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="enrichment-analysis">Enrichment analysis<a class="anchor" aria-label="anchor" href="#enrichment-analysis"></a>
</h3>
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fry_indices</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/ids2indices.html" class="external-link">ids2indices</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">gsc</span>, <span class="va">geneIds</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">v</span><span class="op">)</span>, remove.empty <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">fry_indices</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">gsc</span>, <span class="va">setName</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">gsc_category</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">gsc</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/pkg/GSEABase/man/CollectionType-class.html" class="external-link">bcCategory</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/GSEABase/man/GeneSet-class.html" class="external-link">collectionType</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">gsc_category</span> <span class="op">&lt;-</span> <span class="va">gsc_category</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">fry_indices</span>, <span class="va">length</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">5</span><span class="op">]</span></span>
<span></span>
<span></span>
<span><span class="va">gsc_subcategory</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">gsc</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/pkg/GSEABase/man/CollectionType-class.html" class="external-link">bcSubCategory</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/GSEABase/man/GeneSet-class.html" class="external-link">collectionType</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">gsc_subcategory</span> <span class="op">&lt;-</span> <span class="va">gsc_subcategory</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">fry_indices</span>, <span class="va">length</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">5</span><span class="op">]</span></span>
<span></span>
<span><span class="va">fry_indices</span> <span class="op">&lt;-</span> <span class="va">fry_indices</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">fry_indices</span>, <span class="va">length</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">5</span><span class="op">]</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">gsc_category</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">gsc_subcategory</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">fry_indices</span><span class="op">)</span></span>
<span></span>
<span><span class="va">fry_indices_cat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/split.html" class="external-link">split</a></span><span class="op">(</span><span class="va">fry_indices</span>, <span class="va">gsc_category</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">fry_indices</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">fry_res_out</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">fry_indices_cat</span>, <span class="kw">function</span> <span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu">limma</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/limma/man/roast.html" class="external-link">fry</a></span><span class="op">(</span><span class="va">v</span>, index <span class="op">=</span> <span class="va">x</span>, design <span class="op">=</span> <span class="va">design</span>, contrast <span class="op">=</span> <span class="va">contr.matrix</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>, robust <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span></span>
<span><span class="va">post_fry_format</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">fry_output</span>, <span class="va">gsc_category</span>, <span class="va">gsc_subcategory</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">fry_output</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span>
<span>  <span class="va">fry_output</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">rbind</span>, <span class="va">fry_output</span><span class="op">)</span></span>
<span>  <span class="va">fry_output</span><span class="op">$</span><span class="va">GenesetName</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">fry_output</span><span class="op">)</span></span>
<span>  <span class="va">fry_output</span><span class="op">$</span><span class="va">GenesetCat</span> <span class="op">&lt;-</span> <span class="va">gsc_category</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">fry_output</span><span class="op">)</span><span class="op">]</span></span>
<span>  <span class="va">fry_output</span><span class="op">$</span><span class="va">GenesetSubCat</span> <span class="op">&lt;-</span> <span class="va">gsc_subcategory</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">fry_output</span><span class="op">)</span><span class="op">]</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">fry_output</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">fry_res_sig</span> <span class="op">&lt;-</span> <span class="fu">post_fry_format</span><span class="op">(</span><span class="va">fry_res_out</span>, <span class="va">gsc_category</span>, <span class="va">gsc_subcategory</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">FDR</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span> </span></code></pre></div>
</div>
<div class="section level3">
<h3 id="visualization-1">Visualization<a class="anchor" aria-label="anchor" href="#visualization-1"></a>
</h3>
<p>To interpret the GSEA output, we can use the <code>vissE</code>
package to visualise the important enriched gene sets.</p>
<div class="sourceCode" id="cb69"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://davislaboratory.github.io/vissE" class="external-link">vissE</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://igraph.org" class="external-link">igraph</a></span><span class="op">)</span></span>
<span><span class="va">dovissE</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">fry_out</span>, <span class="va">de_table</span>, <span class="va">topN</span> <span class="op">=</span> <span class="fl">6</span>, <span class="va">title</span> <span class="op">=</span> <span class="st">""</span>, <span class="va">specific_clusters</span> <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span><span class="op">{</span></span>
<span>  </span>
<span>  <span class="va">n_row</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="fl">1000</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">fry_out</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">gs_sig_name</span> <span class="op">&lt;-</span> <span class="va">fry_out</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">FDR</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="va">FDR</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="va">.</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="va">n_row</span>,<span class="op">]</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="op">)</span></span>
<span>  <span class="va">gsc_sig</span> <span class="op">&lt;-</span> <span class="va">gsc</span><span class="op">[</span><span class="va">gs_sig_name</span>,<span class="op">]</span></span>
<span>  </span>
<span>  <span class="va">gs_ovlap</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://DavisLaboratory.github.io/vissE/reference/computeMsigOverlap.html" class="external-link">computeMsigOverlap</a></span><span class="op">(</span><span class="va">gsc_sig</span>, thresh <span class="op">=</span> <span class="fl">0.15</span><span class="op">)</span></span>
<span>  <span class="va">gs_ovnet</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://DavisLaboratory.github.io/vissE/reference/computeMsigNetwork.html" class="external-link">computeMsigNetwork</a></span><span class="op">(</span><span class="va">gs_ovlap</span>, <span class="va">gsc</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="va">gs_stats</span> <span class="op">&lt;-</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="va">fry_out</span><span class="op">[</span><span class="va">gs_sig_name</span>,<span class="op">]</span><span class="op">$</span><span class="va">FDR</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">gs_stats</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">gs_sig_name</span></span>
<span>  </span>
<span>  <span class="co">#identify clusters</span></span>
<span>  <span class="va">grps</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/igraph/man/cluster_walktrap.html" class="external-link">cluster_walktrap</a></span><span class="op">(</span><span class="va">gs_ovnet</span><span class="op">)</span></span>
<span>  <span class="co">#extract clustering results</span></span>
<span>  <span class="va">grps</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/igraph/man/groups.html" class="external-link">groups</a></span><span class="op">(</span><span class="va">grps</span><span class="op">)</span></span>
<span>  <span class="co">#sort by cluster size</span></span>
<span>  <span class="va">grps</span> <span class="op">=</span> <span class="va">grps</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/order.html" class="external-link">order</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">grps</span>, <span class="va">length</span><span class="op">)</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">]</span></span>
<span>  </span>
<span>  <span class="co"># write output</span></span>
<span>  <span class="va">output_clusters</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span>  <span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">grps</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="va">output_clusters</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>geneset <span class="op">=</span> <span class="va">grps</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>, cluster <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"cluster"</span>,<span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">grps</span><span class="op">)</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span>  <span class="va">output_clusters</span> <span class="op">&lt;&lt;-</span> <span class="va">output_clusters</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind.html" class="external-link">bind_rows</a></span><span class="op">(</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="kw">if</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">specific_clusters</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="va">grps</span> <span class="op">&lt;-</span> <span class="va">grps</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="va">topN</span><span class="op">]</span></span>
<span>  <span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>    <span class="va">grps</span> <span class="op">&lt;-</span> <span class="va">grps</span><span class="op">[</span><span class="va">specific_clusters</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="op">)</span><span class="op">]</span></span>
<span>  <span class="op">}</span></span>
<span>  </span>
<span>  <span class="co">#plot the top 12 clusters</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">36</span><span class="op">)</span> <span class="co">#set seed for reproducible layout</span></span>
<span>  <span class="va">p1</span> <span class="op">&lt;&lt;-</span> <span class="fu"><a href="https://DavisLaboratory.github.io/vissE/reference/plotMsigNetwork.html" class="external-link">plotMsigNetwork</a></span><span class="op">(</span><span class="va">gs_ovnet</span>, markGroups <span class="op">=</span> <span class="va">grps</span>, </span>
<span>                        genesetStat <span class="op">=</span> <span class="va">gs_stats</span>, rmUnmarkedGroups <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">scico</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scico/man/ggplot2-scales.html" class="external-link">scale_fill_scico</a></span><span class="op">(</span>name <span class="op">=</span> <span class="st">"-log10(FDR)"</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="va">p2</span> <span class="op">&lt;&lt;-</span> <span class="fu"><a href="https://DavisLaboratory.github.io/vissE/reference/plotMsigWordcloud.html" class="external-link">plotMsigWordcloud</a></span><span class="op">(</span><span class="va">gsc</span>, <span class="va">grps</span>, type <span class="op">=</span> <span class="st">'Name'</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/GSEABase/man/GeneSet-class.html" class="external-link">geneIds</a></span><span class="op">(</span><span class="va">gsc_sig</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="va">genes_logfc</span> <span class="op">&lt;-</span> <span class="va">de_table</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html" class="external-link">rownames_to_column</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">rowname</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="va">genes</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="va">.</span><span class="op">$</span><span class="va">logFC</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">genes_logfc</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">de_table</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html" class="external-link">rownames_to_column</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">rowname</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="va">genes</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="va">.</span><span class="op">$</span><span class="va">rowname</span></span>
<span>  </span>
<span>  <span class="va">p3</span> <span class="op">&lt;&lt;-</span> <span class="fu"><a href="https://DavisLaboratory.github.io/vissE/reference/plotGeneStats.html" class="external-link">plotGeneStats</a></span><span class="op">(</span><span class="va">genes_logfc</span>, <span class="va">gsc</span>, <span class="va">grps</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="fl">0</span>, colour <span class="op">=</span> <span class="fl">2</span>, lty <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"logFC"</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="co">#p4 &lt;- plotMsigPPI(ppi, gsc, grps[1:topN], geneStat = genes_logfc) +</span></span>
<span>  <span class="co">#  guides(col=guide_legend(title="logFC"))</span></span>
<span>  </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">p2</span> <span class="op">+</span> <span class="va">p1</span> <span class="op">+</span> <span class="va">p3</span> <span class="op">+</span> <span class="fu">patchwork</span><span class="fu">::</span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html" class="external-link">plot_layout</a></span><span class="op">(</span>ncol <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="op">+</span></span>
<span>          <span class="fu">patchwork</span><span class="fu">::</span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_annotation.html" class="external-link">plot_annotation</a></span><span class="op">(</span>title <span class="op">=</span> <span class="va">title</span><span class="op">)</span><span class="op">)</span>  </span>
<span>    </span>
<span><span class="op">}</span></span></code></pre></div>
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">dovissE</span><span class="op">(</span><span class="va">fry_res_sig</span>, <span class="va">de_genes_toptable_DvCinGab</span>, title <span class="op">=</span> <span class="st">"DKD vs Normal in abnormal glomerulus"</span> <span class="op">)</span></span></code></pre></div>
<p><img src="GeoMXAnalysisWorkflow_files/figure-html/unnamed-chunk-47-1.png" width="1440"></p>
</div>
</div>
<div class="section level2">
<h2 id="summary">Summary<a class="anchor" aria-label="anchor" href="#summary"></a>
</h2>
<p>To write.</p>
</div>
<div class="section level2">
<h2 class="unnumbered" id="packages-used">Packages used<a class="anchor" aria-label="anchor" href="#packages-used"></a>
</h2>
<p>This workflow depends on various packages from version 3.16 of the
Bioconductor project, running on R version 4.2.0 (2022-04-22) or higher.
The complete list of the packages used for this workflow are shown
below:</p>
<div class="sourceCode" id="cb71"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.2.0 (2022-04-22)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">## Running under: Ubuntu 20.04.4 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3</span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/liblapack.so.3</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] igraph_1.3.2                ggrepel_0.9.1              </span></span>
<span><span class="co">##  [3] ExperimentHub_2.5.0         AnnotationHub_3.5.0        </span></span>
<span><span class="co">##  [5] BiocFileCache_2.5.0         dbplyr_2.2.1               </span></span>
<span><span class="co">##  [7] ggalluvial_0.12.3           msigdb_1.5.0               </span></span>
<span><span class="co">##  [9] GSEABase_1.59.0             graph_1.75.0               </span></span>
<span><span class="co">## [11] annotate_1.75.0             XML_3.99-0.10              </span></span>
<span><span class="co">## [13] AnnotationDbi_1.59.1        vissE_1.5.0                </span></span>
<span><span class="co">## [15] forcats_0.5.1               stringr_1.4.0              </span></span>
<span><span class="co">## [17] dplyr_1.0.9                 purrr_0.3.4                </span></span>
<span><span class="co">## [19] readr_2.1.2                 tidyr_1.2.0                </span></span>
<span><span class="co">## [21] tibble_3.1.7                ggplot2_3.3.6              </span></span>
<span><span class="co">## [23] tidyverse_1.3.1             edgeR_3.39.1               </span></span>
<span><span class="co">## [25] limma_3.53.3                SpatialExperiment_1.7.0    </span></span>
<span><span class="co">## [27] SingleCellExperiment_1.19.0 SummarizedExperiment_1.27.1</span></span>
<span><span class="co">## [29] Biobase_2.57.1              GenomicRanges_1.49.0       </span></span>
<span><span class="co">## [31] GenomeInfoDb_1.33.3         IRanges_2.31.0             </span></span>
<span><span class="co">## [33] S4Vectors_0.35.1            BiocGenerics_0.43.0        </span></span>
<span><span class="co">## [35] MatrixGenerics_1.9.1        matrixStats_0.62.0         </span></span>
<span><span class="co">## [37] standR_1.1.0               </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] rappdirs_0.3.3                R.methodsS3_1.8.2            </span></span>
<span><span class="co">##   [3] ragg_1.2.2                    bit64_4.0.5                  </span></span>
<span><span class="co">##   [5] knitr_1.39                    irlba_2.3.5                  </span></span>
<span><span class="co">##   [7] DelayedArray_0.23.0           R.utils_2.11.0               </span></span>
<span><span class="co">##   [9] data.table_1.14.2             KEGGREST_1.37.2              </span></span>
<span><span class="co">##  [11] RCurl_1.98-1.7                generics_0.1.2               </span></span>
<span><span class="co">##  [13] ScaledMatrix_1.5.0            qdapRegex_0.7.5              </span></span>
<span><span class="co">##  [15] cowplot_1.1.1                 RSQLite_2.2.14               </span></span>
<span><span class="co">##  [17] bit_4.0.4                     tzdb_0.3.0                   </span></span>
<span><span class="co">##  [19] textstem_0.1.4                xml2_1.3.3                   </span></span>
<span><span class="co">##  [21] lubridate_1.8.0               httpuv_1.6.5                 </span></span>
<span><span class="co">##  [23] assertthat_0.2.1              viridis_0.6.2                </span></span>
<span><span class="co">##  [25] xfun_0.31                     hms_1.1.1                    </span></span>
<span><span class="co">##  [27] jquerylib_0.1.4               evaluate_0.15                </span></span>
<span><span class="co">##  [29] promises_1.2.0.1              fansi_1.0.3                  </span></span>
<span><span class="co">##  [31] readxl_1.4.0                  DBI_1.1.3                    </span></span>
<span><span class="co">##  [33] htmlwidgets_1.5.4             ellipsis_0.3.2               </span></span>
<span><span class="co">##  [35] RSpectra_0.16-1               crosstalk_1.2.0              </span></span>
<span><span class="co">##  [37] ggpubr_0.4.0                  backports_1.4.1              </span></span>
<span><span class="co">##  [39] sparseMatrixStats_1.9.0       vctrs_0.4.1                  </span></span>
<span><span class="co">##  [41] ggwordcloud_0.5.0             prettydoc_0.4.1              </span></span>
<span><span class="co">##  [43] abind_1.4-5                   cachem_1.0.6                 </span></span>
<span><span class="co">##  [45] withr_2.5.0                   ggforce_0.3.3                </span></span>
<span><span class="co">##  [47] vroom_1.5.7                   cluster_2.1.3                </span></span>
<span><span class="co">##  [49] crayon_1.5.1                  koRpus_0.13-8                </span></span>
<span><span class="co">##  [51] slam_0.1-50                   pkgconfig_2.0.3              </span></span>
<span><span class="co">##  [53] labeling_0.4.2                sylly_0.1-6                  </span></span>
<span><span class="co">##  [55] tweenr_1.0.2                  nlme_3.1-158                 </span></span>
<span><span class="co">##  [57] vipor_0.4.5                   rlang_1.0.3                  </span></span>
<span><span class="co">##  [59] lifecycle_1.0.1               textshape_1.7.3              </span></span>
<span><span class="co">##  [61] filelock_1.0.2                lexicon_1.2.1                </span></span>
<span><span class="co">##  [63] modelr_0.1.8                  rsvd_1.0.5                   </span></span>
<span><span class="co">##  [65] cellranger_1.1.0              rprojroot_2.0.3              </span></span>
<span><span class="co">##  [67] polyclip_1.10-0               Matrix_1.4-1                 </span></span>
<span><span class="co">##  [69] carData_3.0-5                 Rhdf5lib_1.19.2              </span></span>
<span><span class="co">##  [71] reprex_2.0.1                  beeswarm_0.4.0               </span></span>
<span><span class="co">##  [73] png_0.1-7                     viridisLite_0.4.0            </span></span>
<span><span class="co">##  [75] rjson_0.2.21                  bitops_1.0-7                 </span></span>
<span><span class="co">##  [77] R.oo_1.25.0                   rhdf5filters_1.9.0           </span></span>
<span><span class="co">##  [79] Biostrings_2.65.1             blob_1.2.3                   </span></span>
<span><span class="co">##  [81] DelayedMatrixStats_1.19.0     mclustcomp_0.3.3             </span></span>
<span><span class="co">##  [83] rstatix_0.7.0                 ggsignif_0.6.3               </span></span>
<span><span class="co">##  [85] beachmat_2.13.4               scales_1.2.0                 </span></span>
<span><span class="co">##  [87] memoise_2.0.1                 magrittr_2.0.3               </span></span>
<span><span class="co">##  [89] plyr_1.8.7                    zlibbioc_1.43.0              </span></span>
<span><span class="co">##  [91] compiler_4.2.0                dqrng_0.3.0                  </span></span>
<span><span class="co">##  [93] RColorBrewer_1.1-3            cli_3.3.0                    </span></span>
<span><span class="co">##  [95] XVector_0.37.0                sylly.en_0.1-3               </span></span>
<span><span class="co">##  [97] patchwork_1.1.1               MASS_7.3-57                  </span></span>
<span><span class="co">##  [99] mgcv_1.8-40                   tidyselect_1.1.2             </span></span>
<span><span class="co">## [101] stringi_1.7.6                 textshaping_0.3.6            </span></span>
<span><span class="co">## [103] highr_0.9                     koRpus.lang.en_0.1-4         </span></span>
<span><span class="co">## [105] yaml_2.3.5                    BiocSingular_1.13.0          </span></span>
<span><span class="co">## [107] locfit_1.5-9.5                grid_4.2.0                   </span></span>
<span><span class="co">## [109] sass_0.4.1                    tools_4.2.0                  </span></span>
<span><span class="co">## [111] ruv_0.9.7.1                   parallel_4.2.0               </span></span>
<span><span class="co">## [113] rstudioapi_0.13               gridExtra_2.3                </span></span>
<span><span class="co">## [115] farver_2.1.0                  ggraph_2.0.5                 </span></span>
<span><span class="co">## [117] DropletUtils_1.17.0           digest_0.6.29                </span></span>
<span><span class="co">## [119] BiocManager_1.30.18           FNN_1.1.3.1                  </span></span>
<span><span class="co">## [121] shiny_1.7.1                   Rcpp_1.0.8.3                 </span></span>
<span><span class="co">## [123] car_3.1-0                     broom_0.8.0                  </span></span>
<span><span class="co">## [125] scuttle_1.7.2                 BiocVersion_3.16.0           </span></span>
<span><span class="co">## [127] later_1.3.0                   org.Hs.eg.db_3.15.0          </span></span>
<span><span class="co">## [129] httr_1.4.3                    Rdpack_2.3.1                 </span></span>
<span><span class="co">## [131] colorspace_2.0-3              rvest_1.0.2                  </span></span>
<span><span class="co">## [133] fs_1.5.2                      splines_4.2.0                </span></span>
<span><span class="co">## [135] uwot_0.1.11                   statmod_1.4.36               </span></span>
<span><span class="co">## [137] pkgdown_2.0.5                 scater_1.25.1                </span></span>
<span><span class="co">## [139] graphlayouts_0.8.0            systemfonts_1.0.4            </span></span>
<span><span class="co">## [141] xtable_1.8-4                  jsonlite_1.8.0               </span></span>
<span><span class="co">## [143] tidygraph_1.2.1               NLP_0.2-1                    </span></span>
<span><span class="co">## [145] scico_1.3.0                   R6_2.5.1                     </span></span>
<span><span class="co">## [147] tm_0.7-8                      pillar_1.7.0                 </span></span>
<span><span class="co">## [149] htmltools_0.5.2               mime_0.12                    </span></span>
<span><span class="co">## [151] glue_1.6.2                    fastmap_1.1.0                </span></span>
<span><span class="co">## [153] DT_0.23                       BiocParallel_1.31.8          </span></span>
<span><span class="co">## [155] BiocNeighbors_1.15.1          interactiveDisplayBase_1.35.0</span></span>
<span><span class="co">## [157] codetools_0.2-18              utf8_1.2.2                   </span></span>
<span><span class="co">## [159] lattice_0.20-45               bslib_0.3.1                  </span></span>
<span><span class="co">## [161] curl_4.3.2                    ggbeeswarm_0.6.0             </span></span>
<span><span class="co">## [163] magick_2.7.3                  textclean_0.9.3              </span></span>
<span><span class="co">## [165] syuzhet_1.0.6                 rmarkdown_2.14               </span></span>
<span><span class="co">## [167] desc_1.4.1                    munsell_0.5.0                </span></span>
<span><span class="co">## [169] rhdf5_2.41.1                  GenomeInfoDbData_1.2.8       </span></span>
<span><span class="co">## [171] HDF5Array_1.25.1              haven_2.5.0                  </span></span>
<span><span class="co">## [173] reshape2_1.4.4                gtable_0.3.0                 </span></span>
<span><span class="co">## [175] rbibutils_2.2.8</span></span></code></pre>
</div>
<div class="section level2">
<h2 class="unnumbered" id="acknowledgments">Acknowledgments<a class="anchor" aria-label="anchor" href="#acknowledgments"></a>
</h2>
<p>To write.</p>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-R-rmarkdown" class="csl-entry">
Allaire, JJ, Yihui Xie, Jonathan McPherson, Javier Luraschi, Kevin
Ushey, Aron Atkins, Hadley Wickham, Joe Cheng, Winston Chang, and
Richard Iannone. 2022. <em>Rmarkdown: Dynamic Documents for r</em>. <a href="https://CRAN.R-project.org/package=rmarkdown" class="external-link">https://CRAN.R-project.org/package=rmarkdown</a>.
</div>
<div id="ref-R-prettydoc" class="csl-entry">
Qiu, Yixuan. 2021. <em>Prettydoc: Creating Pretty Documents from r
Markdown</em>. <a href="https://github.com/yixuan/prettydoc" class="external-link">https://github.com/yixuan/prettydoc</a>.
</div>
<div id="ref-R-ggplot2" class="csl-entry">
Wickham, Hadley, Winston Chang, Lionel Henry, Thomas Lin Pedersen,
Kohske Takahashi, Claus Wilke, Kara Woo, Hiroaki Yutani, and Dewey
Dunnington. 2021. <em>Ggplot2: Create Elegant Data Visualisations Using
the Grammar of Graphics</em>. <a href="https://CRAN.R-project.org/package=ggplot2" class="external-link">https://CRAN.R-project.org/package=ggplot2</a>.
</div>
<div id="ref-R-knitr" class="csl-entry">
Xie, Yihui. 2022. <em>Knitr: A General-Purpose Package for Dynamic
Report Generation in r</em>. <a href="https://yihui.org/knitr/" class="external-link">https://yihui.org/knitr/</a>.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Ning Liu, Chin Wee Tan, Melissa J. Davis.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.5.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
